// @theia/monaco-editor-core/esm/vs/editor/contrib/colorPicker/browser/colorPickerParticipantUtils.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.createColorHover = createColorHover;
exports.updateEditorModel = updateEditorModel;
exports.updateColorPresentations = updateColorPresentations;
const cancellation_js_1 = require("../../../../base/common/cancellation.js");
const color_js_1 = require("../../../../base/common/color.js");
const color_js_2 = require("./color.js");
const colorPickerModel_js_1 = require("./colorPickerModel.js");
const range_js_1 = require("../../../common/core/range.js");
async function createColorHover(editorModel, colorInfo, provider) {
    const originalText = editorModel.getValueInRange(colorInfo.range);
    const { red, green, blue, alpha } = colorInfo.color;
    const rgba = new color_js_1.RGBA(Math.round(red * 255), Math.round(green * 255), Math.round(blue * 255), alpha);
    const color = new color_js_1.Color(rgba);
    const colorPresentations = await (0, color_js_2.getColorPresentations)(editorModel, colorInfo, provider, cancellation_js_1.CancellationToken.None);
    const model = new colorPickerModel_js_1.ColorPickerModel(color, [], 0);
    model.colorPresentations = colorPresentations || [];
    model.guessColorPresentation(color, originalText);
    return {
        range: range_js_1.Range.lift(colorInfo.range),
        model,
        provider
    };
}
function updateEditorModel(editor, range, model) {
    const textEdits = [];
    const edit = model.presentation.textEdit ?? { range, text: model.presentation.label, forceMoveMarkers: false };
    textEdits.push(edit);
    if (model.presentation.additionalTextEdits) {
        textEdits.push(...model.presentation.additionalTextEdits);
    }
    const replaceRange = range_js_1.Range.lift(edit.range);
    const trackedRange = editor.getModel()._setTrackedRange(null, replaceRange, 3 /* TrackedRangeStickiness.GrowsOnlyWhenTypingAfter */);
    editor.executeEdits('colorpicker', textEdits);
    editor.pushUndoStop();
    return editor.getModel()._getTrackedRange(trackedRange) ?? replaceRange;
}
async function updateColorPresentations(editorModel, colorPickerModel, color, range, colorHover) {
    const colorPresentations = await (0, color_js_2.getColorPresentations)(editorModel, {
        range: range,
        color: {
            red: color.rgba.r / 255,
            green: color.rgba.g / 255,
            blue: color.rgba.b / 255,
            alpha: color.rgba.a
        }
    }, colorHover.provider, cancellation_js_1.CancellationToken.None);
    colorPickerModel.colorPresentations = colorPresentations || [];
}
