// @theia/monaco-editor-core/esm/vs/base/common/ime.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.IME = exports.IMEImpl = void 0;
const event_js_1 = require("./event.js");
class IMEImpl {
    constructor() {
        this._onDidChange = new event_js_1.Emitter();
        this.onDidChange = this._onDidChange.event;
        this._enabled = true;
    }
    get enabled() {
        return this._enabled;
    }
    /**
     * Enable IME
     */
    enable() {
        this._enabled = true;
        this._onDidChange.fire();
    }
    /**
     * Disable IME
     */
    disable() {
        this._enabled = false;
        this._onDidChange.fire();
    }
}
exports.IMEImpl = IMEImpl;
exports.IME = new IMEImpl();
