// @theia/monaco-editor-core/esm/vs/editor/browser/widget/diffEditor/diffEditorOptions.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiffEditorOptions = void 0;
const observable_js_1 = require("../../../../base/common/observable.js");
const accessibility_js_1 = require("../../../../platform/accessibility/common/accessibility.js");
const diffEditor_js_1 = require("../../../common/config/diffEditor.js");
const editorOptions_js_1 = require("../../../common/config/editorOptions.js");
const diffEditorViewZones_js_1 = require("./components/diffEditorViewZones/diffEditorViewZones.js");
let DiffEditorOptions = class DiffEditorOptions {
    get editorOptions() { return this._options; }
    constructor(options, _accessibilityService) {
        this._accessibilityService = _accessibilityService;
        this._diffEditorWidth = (0, observable_js_1.observableValue)(this, 0);
        this._screenReaderMode = (0, observable_js_1.observableFromEvent)(this, this._accessibilityService.onDidChangeScreenReaderOptimized, () => this._accessibilityService.isScreenReaderOptimized());
        this.couldShowInlineViewBecauseOfSize = (0, observable_js_1.derived)(this, reader => this._options.read(reader).renderSideBySide && this._diffEditorWidth.read(reader) <= this._options.read(reader).renderSideBySideInlineBreakpoint);
        this.renderOverviewRuler = (0, observable_js_1.derived)(this, reader => this._options.read(reader).renderOverviewRuler);
        this.renderSideBySide = (0, observable_js_1.derived)(this, reader => {
            if (this.compactMode.read(reader)) {
                if (this.shouldRenderInlineViewInSmartMode.read(reader)) {
                    return false;
                }
            }
            return this._options.read(reader).renderSideBySide
                && !(this._options.read(reader).useInlineViewWhenSpaceIsLimited && this.couldShowInlineViewBecauseOfSize.read(reader) && !this._screenReaderMode.read(reader));
        });
        this.readOnly = (0, observable_js_1.derived)(this, reader => this._options.read(reader).readOnly);
        this.shouldRenderOldRevertArrows = (0, observable_js_1.derived)(this, reader => {
            if (!this._options.read(reader).renderMarginRevertIcon) {
                return false;
            }
            if (!this.renderSideBySide.read(reader)) {
                return false;
            }
            if (this.readOnly.read(reader)) {
                return false;
            }
            if (this.shouldRenderGutterMenu.read(reader)) {
                return false;
            }
            return true;
        });
        this.shouldRenderGutterMenu = (0, observable_js_1.derived)(this, reader => this._options.read(reader).renderGutterMenu);
        this.renderIndicators = (0, observable_js_1.derived)(this, reader => this._options.read(reader).renderIndicators);
        this.enableSplitViewResizing = (0, observable_js_1.derived)(this, reader => this._options.read(reader).enableSplitViewResizing);
        this.splitViewDefaultRatio = (0, observable_js_1.derived)(this, reader => this._options.read(reader).splitViewDefaultRatio);
        this.ignoreTrimWhitespace = (0, observable_js_1.derived)(this, reader => this._options.read(reader).ignoreTrimWhitespace);
        this.maxComputationTimeMs = (0, observable_js_1.derived)(this, reader => this._options.read(reader).maxComputationTime);
        this.showMoves = (0, observable_js_1.derived)(this, reader => this._options.read(reader).experimental.showMoves && this.renderSideBySide.read(reader));
        this.isInEmbeddedEditor = (0, observable_js_1.derived)(this, reader => this._options.read(reader).isInEmbeddedEditor);
        this.diffWordWrap = (0, observable_js_1.derived)(this, reader => this._options.read(reader).diffWordWrap);
        this.originalEditable = (0, observable_js_1.derived)(this, reader => this._options.read(reader).originalEditable);
        this.diffCodeLens = (0, observable_js_1.derived)(this, reader => this._options.read(reader).diffCodeLens);
        this.accessibilityVerbose = (0, observable_js_1.derived)(this, reader => this._options.read(reader).accessibilityVerbose);
        this.diffAlgorithm = (0, observable_js_1.derived)(this, reader => this._options.read(reader).diffAlgorithm);
        this.showEmptyDecorations = (0, observable_js_1.derived)(this, reader => this._options.read(reader).experimental.showEmptyDecorations);
        this.onlyShowAccessibleDiffViewer = (0, observable_js_1.derived)(this, reader => this._options.read(reader).onlyShowAccessibleDiffViewer);
        this.compactMode = (0, observable_js_1.derived)(this, reader => this._options.read(reader).compactMode);
        this.trueInlineDiffRenderingEnabled = (0, observable_js_1.derived)(this, reader => this._options.read(reader).experimental.useTrueInlineView);
        this.useTrueInlineDiffRendering = (0, observable_js_1.derived)(this, reader => !this.renderSideBySide.read(reader) && this.trueInlineDiffRenderingEnabled.read(reader));
        this.hideUnchangedRegions = (0, observable_js_1.derived)(this, reader => this._options.read(reader).hideUnchangedRegions.enabled);
        this.hideUnchangedRegionsRevealLineCount = (0, observable_js_1.derived)(this, reader => this._options.read(reader).hideUnchangedRegions.revealLineCount);
        this.hideUnchangedRegionsContextLineCount = (0, observable_js_1.derived)(this, reader => this._options.read(reader).hideUnchangedRegions.contextLineCount);
        this.hideUnchangedRegionsMinimumLineCount = (0, observable_js_1.derived)(this, reader => this._options.read(reader).hideUnchangedRegions.minimumLineCount);
        this._model = (0, observable_js_1.observableValue)(this, undefined);
        this.shouldRenderInlineViewInSmartMode = this._model
            .map(this, model => (0, observable_js_1.derivedConstOnceDefined)(this, reader => {
            const diffs = model?.diff.read(reader);
            return diffs ? isSimpleDiff(diffs, this.trueInlineDiffRenderingEnabled.read(reader)) : undefined;
        }))
            .flatten()
            .map(this, v => !!v);
        this.inlineViewHideOriginalLineNumbers = this.compactMode;
        const optionsCopy = { ...options, ...validateDiffEditorOptions(options, diffEditor_js_1.diffEditorDefaultOptions) };
        this._options = (0, observable_js_1.observableValue)(this, optionsCopy);
    }
    updateOptions(changedOptions) {
        const newDiffEditorOptions = validateDiffEditorOptions(changedOptions, this._options.get());
        const newOptions = { ...this._options.get(), ...changedOptions, ...newDiffEditorOptions };
        this._options.set(newOptions, undefined, { changedOptions: changedOptions });
    }
    setWidth(width) {
        this._diffEditorWidth.set(width, undefined);
    }
    setModel(model) {
        this._model.set(model, undefined);
    }
};
exports.DiffEditorOptions = DiffEditorOptions;
exports.DiffEditorOptions = DiffEditorOptions = __decorate([
    __param(1, accessibility_js_1.IAccessibilityService)
], DiffEditorOptions);
function isSimpleDiff(diff, supportsTrueDiffRendering) {
    return diff.mappings.every(m => isInsertion(m.lineRangeMapping) || isDeletion(m.lineRangeMapping) || (supportsTrueDiffRendering && (0, diffEditorViewZones_js_1.allowsTrueInlineDiffRendering)(m.lineRangeMapping)));
}
function isInsertion(mapping) {
    return mapping.original.length === 0;
}
function isDeletion(mapping) {
    return mapping.modified.length === 0;
}
function validateDiffEditorOptions(options, defaults) {
    return {
        enableSplitViewResizing: (0, editorOptions_js_1.boolean)(options.enableSplitViewResizing, defaults.enableSplitViewResizing),
        splitViewDefaultRatio: (0, editorOptions_js_1.clampedFloat)(options.splitViewDefaultRatio, 0.5, 0.1, 0.9),
        renderSideBySide: (0, editorOptions_js_1.boolean)(options.renderSideBySide, defaults.renderSideBySide),
        renderMarginRevertIcon: (0, editorOptions_js_1.boolean)(options.renderMarginRevertIcon, defaults.renderMarginRevertIcon),
        maxComputationTime: (0, editorOptions_js_1.clampedInt)(options.maxComputationTime, defaults.maxComputationTime, 0, 1073741824 /* Constants.MAX_SAFE_SMALL_INTEGER */),
        maxFileSize: (0, editorOptions_js_1.clampedInt)(options.maxFileSize, defaults.maxFileSize, 0, 1073741824 /* Constants.MAX_SAFE_SMALL_INTEGER */),
        ignoreTrimWhitespace: (0, editorOptions_js_1.boolean)(options.ignoreTrimWhitespace, defaults.ignoreTrimWhitespace),
        renderIndicators: (0, editorOptions_js_1.boolean)(options.renderIndicators, defaults.renderIndicators),
        originalEditable: (0, editorOptions_js_1.boolean)(options.originalEditable, defaults.originalEditable),
        diffCodeLens: (0, editorOptions_js_1.boolean)(options.diffCodeLens, defaults.diffCodeLens),
        renderOverviewRuler: (0, editorOptions_js_1.boolean)(options.renderOverviewRuler, defaults.renderOverviewRuler),
        diffWordWrap: (0, editorOptions_js_1.stringSet)(options.diffWordWrap, defaults.diffWordWrap, ['off', 'on', 'inherit']),
        diffAlgorithm: (0, editorOptions_js_1.stringSet)(options.diffAlgorithm, defaults.diffAlgorithm, ['legacy', 'advanced'], { 'smart': 'legacy', 'experimental': 'advanced' }),
        accessibilityVerbose: (0, editorOptions_js_1.boolean)(options.accessibilityVerbose, defaults.accessibilityVerbose),
        experimental: {
            showMoves: (0, editorOptions_js_1.boolean)(options.experimental?.showMoves, defaults.experimental.showMoves),
            showEmptyDecorations: (0, editorOptions_js_1.boolean)(options.experimental?.showEmptyDecorations, defaults.experimental.showEmptyDecorations),
            useTrueInlineView: (0, editorOptions_js_1.boolean)(options.experimental?.useTrueInlineView, defaults.experimental.useTrueInlineView),
        },
        hideUnchangedRegions: {
            enabled: (0, editorOptions_js_1.boolean)(options.hideUnchangedRegions?.enabled ?? options.experimental?.collapseUnchangedRegions, defaults.hideUnchangedRegions.enabled),
            contextLineCount: (0, editorOptions_js_1.clampedInt)(options.hideUnchangedRegions?.contextLineCount, defaults.hideUnchangedRegions.contextLineCount, 0, 1073741824 /* Constants.MAX_SAFE_SMALL_INTEGER */),
            minimumLineCount: (0, editorOptions_js_1.clampedInt)(options.hideUnchangedRegions?.minimumLineCount, defaults.hideUnchangedRegions.minimumLineCount, 0, 1073741824 /* Constants.MAX_SAFE_SMALL_INTEGER */),
            revealLineCount: (0, editorOptions_js_1.clampedInt)(options.hideUnchangedRegions?.revealLineCount, defaults.hideUnchangedRegions.revealLineCount, 0, 1073741824 /* Constants.MAX_SAFE_SMALL_INTEGER */),
        },
        isInEmbeddedEditor: (0, editorOptions_js_1.boolean)(options.isInEmbeddedEditor, defaults.isInEmbeddedEditor),
        onlyShowAccessibleDiffViewer: (0, editorOptions_js_1.boolean)(options.onlyShowAccessibleDiffViewer, defaults.onlyShowAccessibleDiffViewer),
        renderSideBySideInlineBreakpoint: (0, editorOptions_js_1.clampedInt)(options.renderSideBySideInlineBreakpoint, defaults.renderSideBySideInlineBreakpoint, 0, 1073741824 /* Constants.MAX_SAFE_SMALL_INTEGER */),
        useInlineViewWhenSpaceIsLimited: (0, editorOptions_js_1.boolean)(options.useInlineViewWhenSpaceIsLimited, defaults.useInlineViewWhenSpaceIsLimited),
        renderGutterMenu: (0, editorOptions_js_1.boolean)(options.renderGutterMenu, defaults.renderGutterMenu),
        compactMode: (0, editorOptions_js_1.boolean)(options.compactMode, defaults.compactMode),
    };
}
