// @theia/monaco-editor-core/esm/vs/editor/browser/widget/diffEditor/registrations.contribution.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.diffDeleteDecorationEmpty = exports.diffWholeLineDeleteDecoration = exports.diffDeleteDecoration = exports.diffAddDecorationEmpty = exports.diffWholeLineAddDecoration = exports.diffAddDecoration = exports.diffLineDeleteDecorationBackground = exports.diffLineAddDecorationBackground = exports.diffLineDeleteDecorationBackgroundWithIndicator = exports.diffLineAddDecorationBackgroundWithIndicator = exports.diffRemoveIcon = exports.diffInsertIcon = exports.diffEditorUnchangedRegionShadow = exports.diffMoveBorderActive = exports.diffMoveBorder = void 0;
const codicons_js_1 = require("../../../../base/common/codicons.js");
const themables_js_1 = require("../../../../base/common/themables.js");
const textModel_js_1 = require("../../../common/model/textModel.js");
const nls_js_1 = require("../../../../nls.js");
const colorRegistry_js_1 = require("../../../../platform/theme/common/colorRegistry.js");
const iconRegistry_js_1 = require("../../../../platform/theme/common/iconRegistry.js");
exports.diffMoveBorder = (0, colorRegistry_js_1.registerColor)('diffEditor.move.border', '#8b8b8b9c', (0, nls_js_1.localize)('diffEditor.move.border', 'The border color for text that got moved in the diff editor.'));
exports.diffMoveBorderActive = (0, colorRegistry_js_1.registerColor)('diffEditor.moveActive.border', '#FFA500', (0, nls_js_1.localize)('diffEditor.moveActive.border', 'The active border color for text that got moved in the diff editor.'));
exports.diffEditorUnchangedRegionShadow = (0, colorRegistry_js_1.registerColor)('diffEditor.unchangedRegionShadow', { dark: '#000000', light: '#737373BF', hcDark: '#000000', hcLight: '#737373BF', }, (0, nls_js_1.localize)('diffEditor.unchangedRegionShadow', 'The color of the shadow around unchanged region widgets.'));
exports.diffInsertIcon = (0, iconRegistry_js_1.registerIcon)('diff-insert', codicons_js_1.Codicon.add, (0, nls_js_1.localize)('diffInsertIcon', 'Line decoration for inserts in the diff editor.'));
exports.diffRemoveIcon = (0, iconRegistry_js_1.registerIcon)('diff-remove', codicons_js_1.Codicon.remove, (0, nls_js_1.localize)('diffRemoveIcon', 'Line decoration for removals in the diff editor.'));
exports.diffLineAddDecorationBackgroundWithIndicator = textModel_js_1.ModelDecorationOptions.register({
    className: 'line-insert',
    description: 'line-insert',
    isWholeLine: true,
    linesDecorationsClassName: 'insert-sign ' + themables_js_1.ThemeIcon.asClassName(exports.diffInsertIcon),
    marginClassName: 'gutter-insert',
});
exports.diffLineDeleteDecorationBackgroundWithIndicator = textModel_js_1.ModelDecorationOptions.register({
    className: 'line-delete',
    description: 'line-delete',
    isWholeLine: true,
    linesDecorationsClassName: 'delete-sign ' + themables_js_1.ThemeIcon.asClassName(exports.diffRemoveIcon),
    marginClassName: 'gutter-delete',
});
exports.diffLineAddDecorationBackground = textModel_js_1.ModelDecorationOptions.register({
    className: 'line-insert',
    description: 'line-insert',
    isWholeLine: true,
    marginClassName: 'gutter-insert',
});
exports.diffLineDeleteDecorationBackground = textModel_js_1.ModelDecorationOptions.register({
    className: 'line-delete',
    description: 'line-delete',
    isWholeLine: true,
    marginClassName: 'gutter-delete',
});
exports.diffAddDecoration = textModel_js_1.ModelDecorationOptions.register({
    className: 'char-insert',
    description: 'char-insert',
    shouldFillLineOnLineBreak: true,
});
exports.diffWholeLineAddDecoration = textModel_js_1.ModelDecorationOptions.register({
    className: 'char-insert',
    description: 'char-insert',
    isWholeLine: true,
});
exports.diffAddDecorationEmpty = textModel_js_1.ModelDecorationOptions.register({
    className: 'char-insert diff-range-empty',
    description: 'char-insert diff-range-empty',
});
exports.diffDeleteDecoration = textModel_js_1.ModelDecorationOptions.register({
    className: 'char-delete',
    description: 'char-delete',
    shouldFillLineOnLineBreak: true,
});
exports.diffWholeLineDeleteDecoration = textModel_js_1.ModelDecorationOptions.register({
    className: 'char-delete',
    description: 'char-delete',
    isWholeLine: true,
});
exports.diffDeleteDecorationEmpty = textModel_js_1.ModelDecorationOptions.register({
    className: 'char-delete diff-range-empty',
    description: 'char-delete diff-range-empty',
});
