// @theia/monaco-editor-core/esm/vs/editor/contrib/inlineCompletions/browser/model/inlineEdit.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.InlineEdit = void 0;
class InlineEdit {
    constructor(edit, isCollapsed, renderExplicitly, commands, inlineCompletion) {
        this.edit = edit;
        this.isCollapsed = isCollapsed;
        this.renderExplicitly = renderExplicitly;
        this.commands = commands;
        this.inlineCompletion = inlineCompletion;
    }
    get range() {
        return this.edit.range;
    }
    get text() {
        return this.edit.text;
    }
    equals(other) {
        return this.edit.equals(other.edit)
            && this.isCollapsed === other.isCollapsed
            && this.renderExplicitly === other.renderExplicitly
            && this.inlineCompletion === other.inlineCompletion;
    }
}
exports.InlineEdit = InlineEdit;
