// @theia/monaco-editor-core/esm/vs/editor/common/services/editorBaseApi.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyMod = void 0;
exports.createMonacoBaseAPI = createMonacoBaseAPI;
const cancellation_js_1 = require("../../../base/common/cancellation.js");
const event_js_1 = require("../../../base/common/event.js");
const keyCodes_js_1 = require("../../../base/common/keyCodes.js");
const uri_js_1 = require("../../../base/common/uri.js");
const position_js_1 = require("../core/position.js");
const range_js_1 = require("../core/range.js");
const selection_js_1 = require("../core/selection.js");
const languages_js_1 = require("../languages.js");
const standaloneEnums = require("../standalone/standaloneEnums.js");
class KeyMod {
    static { this.CtrlCmd = 2048 /* ConstKeyMod.CtrlCmd */; }
    static { this.Shift = 1024 /* ConstKeyMod.Shift */; }
    static { this.Alt = 512 /* ConstKeyMod.Alt */; }
    static { this.WinCtrl = 256 /* ConstKeyMod.WinCtrl */; }
    static chord(firstPart, secondPart) {
        return (0, keyCodes_js_1.KeyChord)(firstPart, secondPart);
    }
}
exports.KeyMod = KeyMod;
function createMonacoBaseAPI() {
    return {
        editor: undefined, // undefined override expected here
        languages: undefined, // undefined override expected here
        CancellationTokenSource: cancellation_js_1.CancellationTokenSource,
        Emitter: event_js_1.Emitter,
        KeyCode: standaloneEnums.KeyCode,
        KeyMod: KeyMod,
        Position: position_js_1.Position,
        Range: range_js_1.Range,
        Selection: selection_js_1.Selection,
        SelectionDirection: standaloneEnums.SelectionDirection,
        MarkerSeverity: standaloneEnums.MarkerSeverity,
        MarkerTag: standaloneEnums.MarkerTag,
        Uri: uri_js_1.URI,
        Token: languages_js_1.Token
    };
}
