// @theia/monaco-editor-core/esm/vs/editor/common/services/languageFeaturesService.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageFeaturesService = void 0;
const languageFeatureRegistry_js_1 = require("../languageFeatureRegistry.js");
const languageFeatures_js_1 = require("./languageFeatures.js");
const extensions_js_1 = require("../../../platform/instantiation/common/extensions.js");
class LanguageFeaturesService {
    constructor() {
        this.referenceProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.renameProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.newSymbolNamesProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.codeActionProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.definitionProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.typeDefinitionProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.declarationProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.implementationProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.documentSymbolProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.inlayHintsProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.colorProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.codeLensProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.documentFormattingEditProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.documentRangeFormattingEditProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.onTypeFormattingEditProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.signatureHelpProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.hoverProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.documentHighlightProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.multiDocumentHighlightProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.selectionRangeProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.foldingRangeProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.linkProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.inlineCompletionsProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.inlineEditProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.completionProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.linkedEditingRangeProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.inlineValuesProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.evaluatableExpressionProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.documentRangeSemanticTokensProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.documentSemanticTokensProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.documentDropEditProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.documentPasteEditProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
        this.mappedEditsProvider = new languageFeatureRegistry_js_1.LanguageFeatureRegistry(this._score.bind(this));
    }
    setNotebookTypeResolver(resolver) {
        this._notebookTypeResolver = resolver;
    }
    _score(uri) {
        return this._notebookTypeResolver?.(uri);
    }
}
exports.LanguageFeaturesService = LanguageFeaturesService;
(0, extensions_js_1.registerSingleton)(languageFeatures_js_1.ILanguageFeaturesService, LanguageFeaturesService, 1 /* InstantiationType.Delayed */);
