// @theia/monaco-editor-core/esm/vs/base/browser/ui/selectBox/selectBox.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectBox = exports.unthemedSelectBoxStyles = void 0;
const listWidget_js_1 = require("../list/listWidget.js");
const selectBoxCustom_js_1 = require("./selectBoxCustom.js");
const selectBoxNative_js_1 = require("./selectBoxNative.js");
const widget_js_1 = require("../widget.js");
const platform_js_1 = require("../../../common/platform.js");
require("./selectBox.css");
exports.unthemedSelectBoxStyles = {
    ...listWidget_js_1.unthemedListStyles,
    selectBackground: '#3C3C3C',
    selectForeground: '#F0F0F0',
    selectBorder: '#3C3C3C',
    decoratorRightForeground: undefined,
    selectListBackground: undefined,
    selectListBorder: undefined,
    focusBorder: undefined,
};
class SelectBox extends widget_js_1.Widget {
    constructor(options, selected, contextViewProvider, styles, selectBoxOptions) {
        super();
        // Default to native SelectBox for OSX unless overridden
        if (platform_js_1.isMacintosh && !selectBoxOptions?.useCustomDrawn) {
            this.selectBoxDelegate = new selectBoxNative_js_1.SelectBoxNative(options, selected, styles, selectBoxOptions);
        }
        else {
            this.selectBoxDelegate = new selectBoxCustom_js_1.SelectBoxList(options, selected, contextViewProvider, styles, selectBoxOptions);
        }
        this._register(this.selectBoxDelegate);
    }
    // Public SelectBox Methods - routed through delegate interface
    get onDidSelect() {
        return this.selectBoxDelegate.onDidSelect;
    }
    setOptions(options, selected) {
        this.selectBoxDelegate.setOptions(options, selected);
    }
    select(index) {
        this.selectBoxDelegate.select(index);
    }
    setAriaLabel(label) {
        this.selectBoxDelegate.setAriaLabel(label);
    }
    focus() {
        this.selectBoxDelegate.focus();
    }
    blur() {
        this.selectBoxDelegate.blur();
    }
    setFocusable(focusable) {
        this.selectBoxDelegate.setFocusable(focusable);
    }
    setEnabled(enabled) {
        this.selectBoxDelegate.setEnabled(enabled);
    }
    render(container) {
        this.selectBoxDelegate.render(container);
    }
}
exports.SelectBox = SelectBox;
