// @theia/notebook/lib/browser/contributions/notebook-preferences.js
"use strict";
// *****************************************************************************
// Copyright (C) 2024 TypeFox and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0.
//
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License v. 2.0 are satisfied: GNU General Public License, version 2
// with the GNU Classpath Exception which is available at
// https://www.gnu.org/software/classpath/license.html.
//
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
// *****************************************************************************
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.bindNotebookPreferences = exports.NotebookPreferenceContribution = exports.notebookPreferenceSchema = exports.NotebookPreferences = void 0;
const core_1 = require("@theia/core");
const browser_1 = require("@theia/core/lib/browser");
var NotebookPreferences;
(function (NotebookPreferences) {
    NotebookPreferences.NOTEBOOK_LINE_NUMBERS = 'notebook.lineNumbers';
    NotebookPreferences.OUTPUT_LINE_HEIGHT = 'notebook.output.lineHeight';
    NotebookPreferences.OUTPUT_FONT_SIZE = 'notebook.output.fontSize';
    NotebookPreferences.OUTPUT_FONT_FAMILY = 'notebook.output.fontFamily';
    NotebookPreferences.OUTPUT_SCROLLING = 'notebook.output.scrolling';
    NotebookPreferences.OUTPUT_WORD_WRAP = 'notebook.output.wordWrap';
    NotebookPreferences.OUTPUT_LINE_LIMIT = 'notebook.output.textLineLimit';
})(NotebookPreferences || (exports.NotebookPreferences = NotebookPreferences = {}));
exports.notebookPreferenceSchema = {
    properties: {
        [NotebookPreferences.NOTEBOOK_LINE_NUMBERS]: {
            type: 'string',
            enum: ['on', 'off'],
            default: 'off',
            description: core_1.nls.localizeByDefault('Controls the display of line numbers in the cell editor.')
        },
        [NotebookPreferences.OUTPUT_LINE_HEIGHT]: {
            // eslint-disable-next-line max-len
            markdownDescription: core_1.nls.localizeByDefault('Line height of the output text within notebook cells.\n - When set to 0, editor line height is used.\n - Values between 0 and 8 will be used as a multiplier with the font size.\n - Values greater than or equal to 8 will be used as effective values.'),
            type: 'number',
            default: 0,
            tags: ['notebookLayout', 'notebookOutputLayout']
        },
        [NotebookPreferences.OUTPUT_FONT_SIZE]: {
            markdownDescription: core_1.nls.localizeByDefault('Font size for the output text within notebook cells. When set to 0, {0} is used.', '`#editor.fontSize#`'),
            type: 'number',
            default: 0,
            tags: ['notebookLayout', 'notebookOutputLayout']
        },
        [NotebookPreferences.OUTPUT_FONT_FAMILY]: {
            markdownDescription: core_1.nls.localizeByDefault('The font family of the output text within notebook cells. When set to empty, the {0} is used.', '`#editor.fontFamily#`'),
            type: 'string',
            tags: ['notebookLayout', 'notebookOutputLayout']
        },
        [NotebookPreferences.OUTPUT_SCROLLING]: {
            markdownDescription: core_1.nls.localizeByDefault('Initially render notebook outputs in a scrollable region when longer than the limit.'),
            type: 'boolean',
            tags: ['notebookLayout', 'notebookOutputLayout'],
            default: false
        },
        [NotebookPreferences.OUTPUT_WORD_WRAP]: {
            markdownDescription: core_1.nls.localizeByDefault('Controls whether the lines in output should wrap.'),
            type: 'boolean',
            tags: ['notebookLayout', 'notebookOutputLayout'],
            default: false
        },
        [NotebookPreferences.OUTPUT_LINE_LIMIT]: {
            markdownDescription: core_1.nls.localizeByDefault('Controls how many lines of text are displayed in a text output. If {0} is enabled, this setting is used to determine the scroll height of the output.', '`#notebook.output.scrolling#`'),
            type: 'number',
            default: 30,
            tags: ['notebookLayout', 'notebookOutputLayout'],
            minimum: 1,
        },
    }
};
exports.NotebookPreferenceContribution = Symbol('NotebookPreferenceContribution');
function bindNotebookPreferences(bind) {
    // We don't need a NotebookPreferenceConfiguration class, so there's no preference proxy to bind
    bind(exports.NotebookPreferenceContribution).toConstantValue({ schema: exports.notebookPreferenceSchema });
    bind(browser_1.PreferenceContribution).toService(exports.NotebookPreferenceContribution);
}
exports.bindNotebookPreferences = bindNotebookPreferences;
