// @theia/monaco-editor-core/esm/vs/editor/contrib/anchorSelect/browser/anchorSelect.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var SelectionAnchorController_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectionAnchorSet = void 0;
const aria_js_1 = require("../../../../base/browser/ui/aria/aria.js");
const htmlContent_js_1 = require("../../../../base/common/htmlContent.js");
const keyCodes_js_1 = require("../../../../base/common/keyCodes.js");
require("./anchorSelect.css");
const editorExtensions_js_1 = require("../../../browser/editorExtensions.js");
const selection_js_1 = require("../../../common/core/selection.js");
const editorContextKeys_js_1 = require("../../../common/editorContextKeys.js");
const nls_js_1 = require("../../../../nls.js");
const contextkey_js_1 = require("../../../../platform/contextkey/common/contextkey.js");
exports.SelectionAnchorSet = new contextkey_js_1.RawContextKey('selectionAnchorSet', false);
let SelectionAnchorController = class SelectionAnchorController {
    static { SelectionAnchorController_1 = this; }
    static { this.ID = 'editor.contrib.selectionAnchorController'; }
    static get(editor) {
        return editor.getContribution(SelectionAnchorController_1.ID);
    }
    constructor(editor, contextKeyService) {
        this.editor = editor;
        this.selectionAnchorSetContextKey = exports.SelectionAnchorSet.bindTo(contextKeyService);
        this.modelChangeListener = editor.onDidChangeModel(() => this.selectionAnchorSetContextKey.reset());
    }
    setSelectionAnchor() {
        if (this.editor.hasModel()) {
            const position = this.editor.getPosition();
            this.editor.changeDecorations((accessor) => {
                if (this.decorationId) {
                    accessor.removeDecoration(this.decorationId);
                }
                this.decorationId = accessor.addDecoration(selection_js_1.Selection.fromPositions(position, position), {
                    description: 'selection-anchor',
                    stickiness: 1 /* TrackedRangeStickiness.NeverGrowsWhenTypingAtEdges */,
                    hoverMessage: new htmlContent_js_1.MarkdownString().appendText((0, nls_js_1.localize)('selectionAnchor', "Selection Anchor")),
                    className: 'selection-anchor'
                });
            });
            this.selectionAnchorSetContextKey.set(!!this.decorationId);
            (0, aria_js_1.alert)((0, nls_js_1.localize)('anchorSet', "Anchor set at {0}:{1}", position.lineNumber, position.column));
        }
    }
    goToSelectionAnchor() {
        if (this.editor.hasModel() && this.decorationId) {
            const anchorPosition = this.editor.getModel().getDecorationRange(this.decorationId);
            if (anchorPosition) {
                this.editor.setPosition(anchorPosition.getStartPosition());
            }
        }
    }
    selectFromAnchorToCursor() {
        if (this.editor.hasModel() && this.decorationId) {
            const start = this.editor.getModel().getDecorationRange(this.decorationId);
            if (start) {
                const end = this.editor.getPosition();
                this.editor.setSelection(selection_js_1.Selection.fromPositions(start.getStartPosition(), end));
                this.cancelSelectionAnchor();
            }
        }
    }
    cancelSelectionAnchor() {
        if (this.decorationId) {
            const decorationId = this.decorationId;
            this.editor.changeDecorations((accessor) => {
                accessor.removeDecoration(decorationId);
                this.decorationId = undefined;
            });
            this.selectionAnchorSetContextKey.set(false);
        }
    }
    dispose() {
        this.cancelSelectionAnchor();
        this.modelChangeListener.dispose();
    }
};
SelectionAnchorController = SelectionAnchorController_1 = __decorate([
    __param(1, contextkey_js_1.IContextKeyService)
], SelectionAnchorController);
class SetSelectionAnchor extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.setSelectionAnchor',
            label: (0, nls_js_1.localize2)('setSelectionAnchor', "Set Selection Anchor"),
            precondition: undefined,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.editorTextFocus,
                primary: (0, keyCodes_js_1.KeyChord)(2048 /* KeyMod.CtrlCmd */ | 41 /* KeyCode.KeyK */, 2048 /* KeyMod.CtrlCmd */ | 32 /* KeyCode.KeyB */),
                weight: 100 /* KeybindingWeight.EditorContrib */
            }
        });
    }
    async run(_accessor, editor) {
        SelectionAnchorController.get(editor)?.setSelectionAnchor();
    }
}
class GoToSelectionAnchor extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.goToSelectionAnchor',
            label: (0, nls_js_1.localize2)('goToSelectionAnchor', "Go to Selection Anchor"),
            precondition: exports.SelectionAnchorSet,
        });
    }
    async run(_accessor, editor) {
        SelectionAnchorController.get(editor)?.goToSelectionAnchor();
    }
}
class SelectFromAnchorToCursor extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.selectFromAnchorToCursor',
            label: (0, nls_js_1.localize2)('selectFromAnchorToCursor', "Select from Anchor to Cursor"),
            precondition: exports.SelectionAnchorSet,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.editorTextFocus,
                primary: (0, keyCodes_js_1.KeyChord)(2048 /* KeyMod.CtrlCmd */ | 41 /* KeyCode.KeyK */, 2048 /* KeyMod.CtrlCmd */ | 41 /* KeyCode.KeyK */),
                weight: 100 /* KeybindingWeight.EditorContrib */
            }
        });
    }
    async run(_accessor, editor) {
        SelectionAnchorController.get(editor)?.selectFromAnchorToCursor();
    }
}
class CancelSelectionAnchor extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.cancelSelectionAnchor',
            label: (0, nls_js_1.localize2)('cancelSelectionAnchor', "Cancel Selection Anchor"),
            precondition: exports.SelectionAnchorSet,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.editorTextFocus,
                primary: 9 /* KeyCode.Escape */,
                weight: 100 /* KeybindingWeight.EditorContrib */
            }
        });
    }
    async run(_accessor, editor) {
        SelectionAnchorController.get(editor)?.cancelSelectionAnchor();
    }
}
(0, editorExtensions_js_1.registerEditorContribution)(SelectionAnchorController.ID, SelectionAnchorController, 4 /* EditorContributionInstantiation.Lazy */);
(0, editorExtensions_js_1.registerEditorAction)(SetSelectionAnchor);
(0, editorExtensions_js_1.registerEditorAction)(GoToSelectionAnchor);
(0, editorExtensions_js_1.registerEditorAction)(SelectFromAnchorToCursor);
(0, editorExtensions_js_1.registerEditorAction)(CancelSelectionAnchor);
