// @theia/core/lib/browser/widgets/react-widget.js
"use strict";
// *****************************************************************************
// Copyright (C) 2018 TypeFox and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0.
//
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License v. 2.0 are satisfied: GNU General Public License, version 2
// with the GNU Classpath Exception which is available at
// https://www.gnu.org/software/classpath/license.html.
//
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
// *****************************************************************************
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactWidget = void 0;
const tslib_1 = require("tslib");
const React = require("react");
const inversify_1 = require("inversify");
const common_1 = require("../../common");
const widget_1 = require("./widget");
const widgets_1 = require("@phosphor/widgets");
const client_1 = require("react-dom/client");
let ReactWidget = class ReactWidget extends widget_1.BaseWidget {
    constructor(options) {
        super(options);
        this.scrollOptions = {
            suppressScrollX: true,
            minScrollbarLength: 35,
        };
        this.nodeRoot = (0, client_1.createRoot)(this.node);
        this.toDispose.push(common_1.Disposable.create(() => this.nodeRoot.unmount()));
    }
    onUpdateRequest(msg) {
        super.onUpdateRequest(msg);
        if (!this.isDisposed) {
            this.nodeRoot.render(React.createElement(React.Fragment, null, this.render()));
        }
    }
};
exports.ReactWidget = ReactWidget;
exports.ReactWidget = ReactWidget = tslib_1.__decorate([
    (0, inversify_1.injectable)(),
    tslib_1.__param(0, (0, inversify_1.unmanaged)()),
    tslib_1.__metadata("design:paramtypes", [Object])
], ReactWidget);
