// @theia/monaco-editor-core/esm/vs/editor/browser/widget/multiDiffEditor/multiDiffEditorWidget.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiDiffEditorWidget = void 0;
const event_js_1 = require("../../../../base/common/event.js");
const hotReloadHelpers_js_1 = require("../../../../base/common/hotReloadHelpers.js");
const lifecycle_js_1 = require("../../../../base/common/lifecycle.js");
const observable_js_1 = require("../../../../base/common/observable.js");
const instantiation_js_1 = require("../../../../platform/instantiation/common/instantiation.js");
require("./colors.js");
const diffEditorItemTemplate_js_1 = require("./diffEditorItemTemplate.js");
const multiDiffEditorViewModel_js_1 = require("./multiDiffEditorViewModel.js");
const multiDiffEditorWidgetImpl_js_1 = require("./multiDiffEditorWidgetImpl.js");
let MultiDiffEditorWidget = class MultiDiffEditorWidget extends lifecycle_js_1.Disposable {
    constructor(_element, _workbenchUIElementFactory, _instantiationService) {
        super();
        this._element = _element;
        this._workbenchUIElementFactory = _workbenchUIElementFactory;
        this._instantiationService = _instantiationService;
        this._dimension = (0, observable_js_1.observableValue)(this, undefined);
        this._viewModel = (0, observable_js_1.observableValue)(this, undefined);
        this._widgetImpl = (0, observable_js_1.derivedWithStore)(this, (reader, store) => {
            (0, hotReloadHelpers_js_1.readHotReloadableExport)(diffEditorItemTemplate_js_1.DiffEditorItemTemplate, reader);
            return store.add(this._instantiationService.createInstance(((0, hotReloadHelpers_js_1.readHotReloadableExport)(multiDiffEditorWidgetImpl_js_1.MultiDiffEditorWidgetImpl, reader)), this._element, this._dimension, this._viewModel, this._workbenchUIElementFactory));
        });
        this._activeControl = (0, observable_js_1.derived)(this, (reader) => this._widgetImpl.read(reader).activeControl.read(reader));
        this.onDidChangeActiveControl = event_js_1.Event.fromObservableLight(this._activeControl);
        this._register((0, observable_js_1.recomputeInitiallyAndOnChange)(this._widgetImpl));
    }
    reveal(resource, options) {
        this._widgetImpl.get().reveal(resource, options);
    }
    createViewModel(model) {
        return new multiDiffEditorViewModel_js_1.MultiDiffEditorViewModel(model, this._instantiationService);
    }
    setViewModel(viewModel) {
        this._viewModel.set(viewModel, undefined);
    }
    layout(dimension) {
        this._dimension.set(dimension, undefined);
    }
    getActiveControl() {
        return this._activeControl.get();
    }
    getViewState() {
        return this._widgetImpl.get().getViewState();
    }
    setViewState(viewState) {
        this._widgetImpl.get().setViewState(viewState);
    }
    tryGetCodeEditor(resource) {
        return this._widgetImpl.get().tryGetCodeEditor(resource);
    }
    findDocumentDiffItem(resource) {
        return this._widgetImpl.get().findDocumentDiffItem(resource);
    }
};
exports.MultiDiffEditorWidget = MultiDiffEditorWidget;
exports.MultiDiffEditorWidget = MultiDiffEditorWidget = __decorate([
    __param(2, instantiation_js_1.IInstantiationService)
], MultiDiffEditorWidget);
