// @theia/monaco-editor-core/esm/vs/editor/browser/widget/multiDiffEditor/multiDiffEditorViewModel.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentDiffItemViewModel = exports.MultiDiffEditorViewModel = void 0;
const lifecycle_js_1 = require("../../../../base/common/lifecycle.js");
const observable_js_1 = require("../../../../base/common/observable.js");
const instantiation_js_1 = require("../../../../platform/instantiation/common/instantiation.js");
const model_js_1 = require("../../../common/services/model.js");
const diffEditorOptions_js_1 = require("../diffEditor/diffEditorOptions.js");
const diffEditorViewModel_js_1 = require("../diffEditor/diffEditorViewModel.js");
const utils_js_1 = require("../diffEditor/utils.js");
class MultiDiffEditorViewModel extends lifecycle_js_1.Disposable {
    async waitForDiffs() {
        for (const d of this.items.get()) {
            await d.diffEditorViewModel.waitForDiff();
        }
    }
    collapseAll() {
        (0, observable_js_1.transaction)(tx => {
            for (const d of this.items.get()) {
                d.collapsed.set(true, tx);
            }
        });
    }
    expandAll() {
        (0, observable_js_1.transaction)(tx => {
            for (const d of this.items.get()) {
                d.collapsed.set(false, tx);
            }
        });
    }
    get contextKeys() {
        return this.model.contextKeys;
    }
    constructor(model, _instantiationService) {
        super();
        this.model = model;
        this._instantiationService = _instantiationService;
        this._documents = (0, observable_js_1.observableFromValueWithChangeEvent)(this.model, this.model.documents);
        this._documentsArr = (0, observable_js_1.derived)(this, reader => {
            const result = this._documents.read(reader);
            if (result === 'loading') {
                return [];
            }
            return result;
        });
        this.isLoading = (0, observable_js_1.derived)(this, reader => this._documents.read(reader) === 'loading');
        this.items = (0, observable_js_1.mapObservableArrayCached)(this, this._documentsArr, (d, store) => store.add(this._instantiationService.createInstance(DocumentDiffItemViewModel, d, this))).recomputeInitiallyAndOnChange(this._store);
        this.focusedDiffItem = (0, observable_js_1.derived)(this, reader => this.items.read(reader).find(i => i.isFocused.read(reader)));
        this.activeDiffItem = (0, observable_js_1.derivedObservableWithWritableCache)(this, (reader, lastValue) => this.focusedDiffItem.read(reader) ?? (lastValue && this.items.read(reader).indexOf(lastValue) !== -1) ? lastValue : undefined);
    }
}
exports.MultiDiffEditorViewModel = MultiDiffEditorViewModel;
let DocumentDiffItemViewModel = class DocumentDiffItemViewModel extends lifecycle_js_1.Disposable {
    get diffEditorViewModel() {
        return this.diffEditorViewModelRef.object;
    }
    get originalUri() { return this.documentDiffItem.original?.uri; }
    get modifiedUri() { return this.documentDiffItem.modified?.uri; }
    setIsFocused(source, tx) {
        this._isFocusedSource.set(source, tx);
    }
    get documentDiffItem() {
        return this.documentDiffItemRef.object;
    }
    constructor(documentDiffItem, _editorViewModel, _instantiationService, _modelService) {
        super();
        this._editorViewModel = _editorViewModel;
        this._instantiationService = _instantiationService;
        this._modelService = _modelService;
        this.collapsed = (0, observable_js_1.observableValue)(this, false);
        this.lastTemplateData = (0, observable_js_1.observableValue)(this, { contentHeight: 500, selections: undefined, });
        this.isActive = (0, observable_js_1.derived)(this, reader => this._editorViewModel.activeDiffItem.read(reader) === this);
        this._isFocusedSource = (0, observable_js_1.observableValue)(this, (0, observable_js_1.constObservable)(false));
        this.isFocused = (0, observable_js_1.derived)(this, reader => this._isFocusedSource.read(reader).read(reader));
        this.isAlive = (0, observable_js_1.observableValue)(this, true);
        this._register((0, lifecycle_js_1.toDisposable)(() => {
            this.isAlive.set(false, undefined);
        }));
        this.documentDiffItemRef = this._register(documentDiffItem.createNewRef(this));
        function updateOptions(options) {
            return {
                ...options,
                hideUnchangedRegions: {
                    enabled: true,
                },
            };
        }
        const options = this._instantiationService.createInstance(diffEditorOptions_js_1.DiffEditorOptions, updateOptions(this.documentDiffItem.options || {}));
        if (this.documentDiffItem.onOptionsDidChange) {
            this._register(this.documentDiffItem.onOptionsDidChange(() => {
                options.updateOptions(updateOptions(this.documentDiffItem.options || {}));
            }));
        }
        const diffEditorViewModelStore = new lifecycle_js_1.DisposableStore();
        const originalTextModel = this.documentDiffItem.original ?? diffEditorViewModelStore.add(this._modelService.createModel('', null));
        const modifiedTextModel = this.documentDiffItem.modified ?? diffEditorViewModelStore.add(this._modelService.createModel('', null));
        diffEditorViewModelStore.add(this.documentDiffItemRef.createNewRef(this));
        this.diffEditorViewModelRef = this._register(utils_js_1.RefCounted.createWithDisposable(this._instantiationService.createInstance(diffEditorViewModel_js_1.DiffEditorViewModel, {
            original: originalTextModel,
            modified: modifiedTextModel,
        }, options), diffEditorViewModelStore, this));
    }
    getKey() {
        return JSON.stringify([
            this.originalUri?.toString(),
            this.modifiedUri?.toString()
        ]);
    }
};
exports.DocumentDiffItemViewModel = DocumentDiffItemViewModel;
exports.DocumentDiffItemViewModel = DocumentDiffItemViewModel = __decorate([
    __param(2, instantiation_js_1.IInstantiationService),
    __param(3, model_js_1.IModelService)
], DocumentDiffItemViewModel);
