// @theia/git/lib/browser/diff/git-diff-tree-model.js
"use strict";
// *****************************************************************************
// Copyright (C) 2020 Arm and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0.
//
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License v. 2.0 are satisfied: GNU General Public License, version 2
// with the GNU Classpath Exception which is available at
// https://www.gnu.org/software/classpath/license.html.
//
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
// *****************************************************************************
Object.defineProperty(exports, "__esModule", { value: true });
exports.GitDiffTreeModel = void 0;
const tslib_1 = require("tslib");
const inversify_1 = require("@theia/core/shared/inversify");
const common_1 = require("@theia/core/lib/common");
const uri_1 = tslib_1.__importDefault(require("@theia/core/lib/common/uri"));
const scm_tree_model_1 = require("@theia/scm/lib/browser/scm-tree-model");
const common_2 = require("../../common");
const scm_service_1 = require("@theia/scm/lib/browser/scm-service");
const git_scm_provider_1 = require("../git-scm-provider");
const git_resource_opener_1 = require("./git-resource-opener");
let GitDiffTreeModel = class GitDiffTreeModel extends scm_tree_model_1.ScmTreeModel {
    constructor() {
        super();
        this._groups = [];
        this.toDisposeOnContentChange = new common_1.DisposableCollection();
        this.toDispose.push(this.toDisposeOnContentChange);
    }
    async setContent(options) {
        const { rootUri, diffOptions } = options;
        this.toDisposeOnContentChange.dispose();
        const scmRepository = this.scmService.findRepository(new uri_1.default(rootUri));
        if (scmRepository && scmRepository.provider.id === 'git') {
            const provider = scmRepository.provider;
            this.provider = provider;
            this.diffOptions = diffOptions;
            this.refreshRepository(provider);
            this.toDisposeOnContentChange.push(provider.onDidChange(() => {
                this.refreshRepository(provider);
            }));
        }
    }
    async refreshRepository(provider) {
        const repository = { localUri: provider.rootUri };
        const gitFileChanges = await this.git.diff(repository, this.diffOptions);
        const group = { id: 'changes', label: 'Files Changed', resources: [], provider, dispose: () => { } };
        const resources = gitFileChanges
            .map(change => new git_scm_provider_1.GitScmFileChange(change, provider, this.diffOptions.range))
            .map(change => ({
            sourceUri: new uri_1.default(change.uri),
            decorations: {
                letter: common_2.GitFileStatus.toAbbreviation(change.gitFileChange.status, true),
                color: common_2.GitFileStatus.getColor(change.gitFileChange.status, true),
                tooltip: common_2.GitFileStatus.toString(change.gitFileChange.status, true)
            },
            open: async () => this.open(change),
            group,
        }));
        const changesGroup = { ...group, resources };
        this._groups = [changesGroup];
        this.root = this.createTree();
    }
    get rootUri() {
        if (this.provider) {
            return this.provider.rootUri;
        }
    }
    ;
    canTabToWidget() {
        return true;
    }
    get groups() {
        return this._groups;
    }
    ;
    async open(change) {
        const uriToOpen = change.getUriToOpen();
        await this.resourceOpener.open(uriToOpen);
    }
    storeState() {
        if (this.provider) {
            return {
                ...super.storeState(),
                rootUri: this.provider.rootUri,
                diffOptions: this.diffOptions,
            };
        }
        else {
            return super.storeState();
        }
    }
    restoreState(oldState) {
        super.restoreState(oldState);
        if (oldState.rootUri && oldState.diffOptions) {
            this.setContent(oldState);
        }
    }
};
exports.GitDiffTreeModel = GitDiffTreeModel;
tslib_1.__decorate([
    (0, inversify_1.inject)(common_2.Git),
    tslib_1.__metadata("design:type", Object)
], GitDiffTreeModel.prototype, "git", void 0);
tslib_1.__decorate([
    (0, inversify_1.inject)(scm_service_1.ScmService),
    tslib_1.__metadata("design:type", scm_service_1.ScmService)
], GitDiffTreeModel.prototype, "scmService", void 0);
tslib_1.__decorate([
    (0, inversify_1.inject)(git_resource_opener_1.GitResourceOpener),
    tslib_1.__metadata("design:type", Object)
], GitDiffTreeModel.prototype, "resourceOpener", void 0);
exports.GitDiffTreeModel = GitDiffTreeModel = tslib_1.__decorate([
    (0, inversify_1.injectable)(),
    tslib_1.__metadata("design:paramtypes", [])
], GitDiffTreeModel);
(function (GitDiffTreeModel) {
    ;
})(GitDiffTreeModel || (exports.GitDiffTreeModel = GitDiffTreeModel = {}));
