// @theia/monaco-editor-core/esm/vs/platform/progress/common/progress.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IEditorProgressService = exports.LongRunningOperation = exports.UnmanagedProgress = exports.AsyncProgress = exports.Progress = exports.emptyProgressRunner = exports.IProgressService = void 0;
const async_js_1 = require("../../../base/common/async.js");
const cancellation_js_1 = require("../../../base/common/cancellation.js");
const lifecycle_js_1 = require("../../../base/common/lifecycle.js");
const instantiation_js_1 = require("../../instantiation/common/instantiation.js");
exports.IProgressService = (0, instantiation_js_1.createDecorator)('progressService');
exports.emptyProgressRunner = Object.freeze({
    total() { },
    worked() { },
    done() { }
});
class Progress {
    static { this.None = Object.freeze({ report() { } }); }
    get value() { return this._value; }
    constructor(callback) {
        this.callback = callback;
    }
    report(item) {
        this._value = item;
        this.callback(this._value);
    }
}
exports.Progress = Progress;
class AsyncProgress {
    get value() { return this._value; }
    constructor(callback) {
        this.callback = callback;
    }
    report(item) {
        if (!this._asyncQueue) {
            this._asyncQueue = [item];
        }
        else {
            this._asyncQueue.push(item);
        }
        this._processAsyncQueue();
    }
    async _processAsyncQueue() {
        if (this._processingAsyncQueue) {
            return;
        }
        try {
            this._processingAsyncQueue = true;
            while (this._asyncQueue && this._asyncQueue.length) {
                const item = this._asyncQueue.shift();
                this._value = item;
                await this.callback(this._value);
            }
        }
        finally {
            this._processingAsyncQueue = false;
            const drainListener = this._drainListener;
            this._drainListener = undefined;
            drainListener?.();
        }
    }
    drain() {
        if (this._processingAsyncQueue) {
            return new Promise(resolve => {
                const prevListener = this._drainListener;
                this._drainListener = () => {
                    prevListener?.();
                    resolve();
                };
            });
        }
        return Promise.resolve();
    }
}
exports.AsyncProgress = AsyncProgress;
/**
 * RAII-style progress instance that allows imperative reporting and hides
 * once `dispose()` is called.
 */
let UnmanagedProgress = class UnmanagedProgress extends lifecycle_js_1.Disposable {
    constructor(options, progressService) {
        super();
        this.deferred = new async_js_1.DeferredPromise();
        progressService.withProgress(options, reporter => {
            this.reporter = reporter;
            if (this.lastStep) {
                reporter.report(this.lastStep);
            }
            return this.deferred.p;
        });
        this._register((0, lifecycle_js_1.toDisposable)(() => this.deferred.complete()));
    }
    report(step) {
        if (this.reporter) {
            this.reporter.report(step);
        }
        else {
            this.lastStep = step;
        }
    }
};
exports.UnmanagedProgress = UnmanagedProgress;
exports.UnmanagedProgress = UnmanagedProgress = __decorate([
    __param(1, exports.IProgressService)
], UnmanagedProgress);
class LongRunningOperation extends lifecycle_js_1.Disposable {
    constructor(progressIndicator) {
        super();
        this.progressIndicator = progressIndicator;
        this.currentOperationId = 0;
        this.currentOperationDisposables = this._register(new lifecycle_js_1.DisposableStore());
    }
    start(progressDelay) {
        // Stop any previous operation
        this.stop();
        // Start new
        const newOperationId = ++this.currentOperationId;
        const newOperationToken = new cancellation_js_1.CancellationTokenSource();
        this.currentProgressTimeout = setTimeout(() => {
            if (newOperationId === this.currentOperationId) {
                this.currentProgressRunner = this.progressIndicator.show(true);
            }
        }, progressDelay);
        this.currentOperationDisposables.add((0, lifecycle_js_1.toDisposable)(() => clearTimeout(this.currentProgressTimeout)));
        this.currentOperationDisposables.add((0, lifecycle_js_1.toDisposable)(() => newOperationToken.cancel()));
        this.currentOperationDisposables.add((0, lifecycle_js_1.toDisposable)(() => this.currentProgressRunner ? this.currentProgressRunner.done() : undefined));
        return {
            id: newOperationId,
            token: newOperationToken.token,
            stop: () => this.doStop(newOperationId),
            isCurrent: () => this.currentOperationId === newOperationId
        };
    }
    stop() {
        this.doStop(this.currentOperationId);
    }
    doStop(operationId) {
        if (this.currentOperationId === operationId) {
            this.currentOperationDisposables.clear();
        }
    }
}
exports.LongRunningOperation = LongRunningOperation;
exports.IEditorProgressService = (0, instantiation_js_1.createDecorator)('editorProgressService');
