// @theia/git/lib/browser/history/git-commit-detail-open-handler.js
"use strict";
// *****************************************************************************
// Copyright (C) 2018 TypeFox and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0.
//
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License v. 2.0 are satisfied: GNU General Public License, version 2
// with the GNU Classpath Exception which is available at
// https://www.gnu.org/software/classpath/license.html.
//
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
// *****************************************************************************
Object.defineProperty(exports, "__esModule", { value: true });
exports.GitCommitDetailOpenHandler = exports.GitCommitDetailUri = void 0;
const tslib_1 = require("tslib");
const inversify_1 = require("@theia/core/shared/inversify");
const browser_1 = require("@theia/core/lib/browser");
const git_scm_provider_1 = require("../git-scm-provider");
var GitCommitDetailUri;
(function (GitCommitDetailUri) {
    GitCommitDetailUri.scheme = git_scm_provider_1.GitScmProvider.GIT_COMMIT_DETAIL;
    function toCommitSha(uri) {
        if (uri.scheme === GitCommitDetailUri.scheme) {
            return uri.fragment;
        }
        throw new Error('The given uri is not an commit detail URI, uri: ' + uri);
    }
    GitCommitDetailUri.toCommitSha = toCommitSha;
})(GitCommitDetailUri || (exports.GitCommitDetailUri = GitCommitDetailUri = {}));
let GitCommitDetailOpenHandler = class GitCommitDetailOpenHandler extends browser_1.WidgetOpenHandler {
    constructor() {
        super(...arguments);
        this.id = git_scm_provider_1.GitScmProvider.GIT_COMMIT_DETAIL;
    }
    canHandle(uri) {
        try {
            GitCommitDetailUri.toCommitSha(uri);
            return 200;
        }
        catch {
            return 0;
        }
    }
    async doOpen(widget, options) {
        await super.doOpen(widget, options);
    }
    createWidgetOptions(uri, commit) {
        return commit;
    }
};
exports.GitCommitDetailOpenHandler = GitCommitDetailOpenHandler;
exports.GitCommitDetailOpenHandler = GitCommitDetailOpenHandler = tslib_1.__decorate([
    (0, inversify_1.injectable)()
], GitCommitDetailOpenHandler);
