// @theia/monaco-editor-core/esm/vs/platform/contextview/browser/contextMenuService.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextMenuMenuDelegate = exports.ContextMenuService = void 0;
const dom_js_1 = require("../../../base/browser/dom.js");
const actions_js_1 = require("../../../base/common/actions.js");
const event_js_1 = require("../../../base/common/event.js");
const lifecycle_js_1 = require("../../../base/common/lifecycle.js");
const menuEntryActionViewItem_js_1 = require("../../actions/browser/menuEntryActionViewItem.js");
const actions_js_2 = require("../../actions/common/actions.js");
const contextkey_js_1 = require("../../contextkey/common/contextkey.js");
const keybinding_js_1 = require("../../keybinding/common/keybinding.js");
const notification_js_1 = require("../../notification/common/notification.js");
const telemetry_js_1 = require("../../telemetry/common/telemetry.js");
const contextMenuHandler_js_1 = require("./contextMenuHandler.js");
const contextView_js_1 = require("./contextView.js");
let ContextMenuService = class ContextMenuService extends lifecycle_js_1.Disposable {
    get contextMenuHandler() {
        if (!this._contextMenuHandler) {
            this._contextMenuHandler = new contextMenuHandler_js_1.ContextMenuHandler(this.contextViewService, this.telemetryService, this.notificationService, this.keybindingService);
        }
        return this._contextMenuHandler;
    }
    constructor(telemetryService, notificationService, contextViewService, keybindingService, menuService, contextKeyService) {
        super();
        this.telemetryService = telemetryService;
        this.notificationService = notificationService;
        this.contextViewService = contextViewService;
        this.keybindingService = keybindingService;
        this.menuService = menuService;
        this.contextKeyService = contextKeyService;
        this._contextMenuHandler = undefined;
        this._onDidShowContextMenu = this._store.add(new event_js_1.Emitter());
        this.onDidShowContextMenu = this._onDidShowContextMenu.event;
        this._onDidHideContextMenu = this._store.add(new event_js_1.Emitter());
        this.onDidHideContextMenu = this._onDidHideContextMenu.event;
    }
    configure(options) {
        this.contextMenuHandler.configure(options);
    }
    // ContextMenu
    showContextMenu(delegate) {
        delegate = ContextMenuMenuDelegate.transform(delegate, this.menuService, this.contextKeyService);
        this.contextMenuHandler.showContextMenu({
            ...delegate,
            onHide: (didCancel) => {
                delegate.onHide?.(didCancel);
                this._onDidHideContextMenu.fire();
            }
        });
        dom_js_1.ModifierKeyEmitter.getInstance().resetKeyStatus();
        this._onDidShowContextMenu.fire();
    }
};
exports.ContextMenuService = ContextMenuService;
exports.ContextMenuService = ContextMenuService = __decorate([
    __param(0, telemetry_js_1.ITelemetryService),
    __param(1, notification_js_1.INotificationService),
    __param(2, contextView_js_1.IContextViewService),
    __param(3, keybinding_js_1.IKeybindingService),
    __param(4, actions_js_2.IMenuService),
    __param(5, contextkey_js_1.IContextKeyService)
], ContextMenuService);
var ContextMenuMenuDelegate;
(function (ContextMenuMenuDelegate) {
    function is(thing) {
        return thing && thing.menuId instanceof actions_js_2.MenuId;
    }
    function transform(delegate, menuService, globalContextKeyService) {
        if (!is(delegate)) {
            return delegate;
        }
        const { menuId, menuActionOptions, contextKeyService } = delegate;
        return {
            ...delegate,
            getActions: () => {
                let target = [];
                if (menuId) {
                    const menu = menuService.getMenuActions(menuId, contextKeyService ?? globalContextKeyService, menuActionOptions);
                    target = (0, menuEntryActionViewItem_js_1.getFlatContextMenuActions)(menu);
                }
                if (!delegate.getActions) {
                    return target;
                }
                else {
                    return actions_js_1.Separator.join(delegate.getActions(), target);
                }
            }
        };
    }
    ContextMenuMenuDelegate.transform = transform;
})(ContextMenuMenuDelegate || (exports.ContextMenuMenuDelegate = ContextMenuMenuDelegate = {}));
