// @theia/monaco-editor-core/esm/vs/editor/contrib/suggest/browser/suggestWidgetRenderer.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ItemRenderer = void 0;
exports.getAriaId = getAriaId;
const dom_js_1 = require("../../../../base/browser/dom.js");
const iconLabel_js_1 = require("../../../../base/browser/ui/iconLabel/iconLabel.js");
const codicons_js_1 = require("../../../../base/common/codicons.js");
const themables_js_1 = require("../../../../base/common/themables.js");
const event_js_1 = require("../../../../base/common/event.js");
const filters_js_1 = require("../../../../base/common/filters.js");
const lifecycle_js_1 = require("../../../../base/common/lifecycle.js");
const uri_js_1 = require("../../../../base/common/uri.js");
const languages_js_1 = require("../../../common/languages.js");
const getIconClasses_js_1 = require("../../../common/services/getIconClasses.js");
const model_js_1 = require("../../../common/services/model.js");
const language_js_1 = require("../../../common/languages/language.js");
const nls = require("../../../../nls.js");
const files_js_1 = require("../../../../platform/files/common/files.js");
const iconRegistry_js_1 = require("../../../../platform/theme/common/iconRegistry.js");
const themeService_js_1 = require("../../../../platform/theme/common/themeService.js");
const suggestWidgetDetails_js_1 = require("./suggestWidgetDetails.js");
function getAriaId(index) {
    return `suggest-aria-id:${index}`;
}
const suggestMoreInfoIcon = (0, iconRegistry_js_1.registerIcon)('suggest-more-info', codicons_js_1.Codicon.chevronRight, nls.localize('suggestMoreInfoIcon', 'Icon for more information in the suggest widget.'));
const _completionItemColor = new class ColorExtractor {
    static { this._regexRelaxed = /(#([\da-fA-F]{3}){1,2}|(rgb|hsl)a\(\s*(\d{1,3}%?\s*,\s*){3}(1|0?\.\d+)\)|(rgb|hsl)\(\s*\d{1,3}%?(\s*,\s*\d{1,3}%?){2}\s*\))/; }
    static { this._regexStrict = new RegExp(`^${ColorExtractor._regexRelaxed.source}$`, 'i'); }
    extract(item, out) {
        if (item.textLabel.match(ColorExtractor._regexStrict)) {
            out[0] = item.textLabel;
            return true;
        }
        if (item.completion.detail && item.completion.detail.match(ColorExtractor._regexStrict)) {
            out[0] = item.completion.detail;
            return true;
        }
        if (item.completion.documentation) {
            const value = typeof item.completion.documentation === 'string'
                ? item.completion.documentation
                : item.completion.documentation.value;
            const match = ColorExtractor._regexRelaxed.exec(value);
            if (match && (match.index === 0 || match.index + match[0].length === value.length)) {
                out[0] = match[0];
                return true;
            }
        }
        return false;
    }
};
let ItemRenderer = class ItemRenderer {
    constructor(_editor, _modelService, _languageService, _themeService) {
        this._editor = _editor;
        this._modelService = _modelService;
        this._languageService = _languageService;
        this._themeService = _themeService;
        this._onDidToggleDetails = new event_js_1.Emitter();
        this.onDidToggleDetails = this._onDidToggleDetails.event;
        this.templateId = 'suggestion';
    }
    dispose() {
        this._onDidToggleDetails.dispose();
    }
    renderTemplate(container) {
        const disposables = new lifecycle_js_1.DisposableStore();
        const root = container;
        root.classList.add('show-file-icons');
        const icon = (0, dom_js_1.append)(container, (0, dom_js_1.$)('.icon'));
        const colorspan = (0, dom_js_1.append)(icon, (0, dom_js_1.$)('span.colorspan'));
        const text = (0, dom_js_1.append)(container, (0, dom_js_1.$)('.contents'));
        const main = (0, dom_js_1.append)(text, (0, dom_js_1.$)('.main'));
        const iconContainer = (0, dom_js_1.append)(main, (0, dom_js_1.$)('.icon-label.codicon'));
        const left = (0, dom_js_1.append)(main, (0, dom_js_1.$)('span.left'));
        const right = (0, dom_js_1.append)(main, (0, dom_js_1.$)('span.right'));
        const iconLabel = new iconLabel_js_1.IconLabel(left, { supportHighlights: true, supportIcons: true });
        disposables.add(iconLabel);
        const parametersLabel = (0, dom_js_1.append)(left, (0, dom_js_1.$)('span.signature-label'));
        const qualifierLabel = (0, dom_js_1.append)(left, (0, dom_js_1.$)('span.qualifier-label'));
        const detailsLabel = (0, dom_js_1.append)(right, (0, dom_js_1.$)('span.details-label'));
        const readMore = (0, dom_js_1.append)(right, (0, dom_js_1.$)('span.readMore' + themables_js_1.ThemeIcon.asCSSSelector(suggestMoreInfoIcon)));
        readMore.title = nls.localize('readMore', "Read More");
        const configureFont = () => {
            const options = this._editor.getOptions();
            const fontInfo = options.get(52 /* EditorOption.fontInfo */);
            const fontFamily = fontInfo.getMassagedFontFamily();
            const fontFeatureSettings = fontInfo.fontFeatureSettings;
            const fontSize = options.get(124 /* EditorOption.suggestFontSize */) || fontInfo.fontSize;
            const lineHeight = options.get(125 /* EditorOption.suggestLineHeight */) || fontInfo.lineHeight;
            const fontWeight = fontInfo.fontWeight;
            const letterSpacing = fontInfo.letterSpacing;
            const fontSizePx = `${fontSize}px`;
            const lineHeightPx = `${lineHeight}px`;
            const letterSpacingPx = `${letterSpacing}px`;
            root.style.fontSize = fontSizePx;
            root.style.fontWeight = fontWeight;
            root.style.letterSpacing = letterSpacingPx;
            main.style.fontFamily = fontFamily;
            main.style.fontFeatureSettings = fontFeatureSettings;
            main.style.lineHeight = lineHeightPx;
            icon.style.height = lineHeightPx;
            icon.style.width = lineHeightPx;
            readMore.style.height = lineHeightPx;
            readMore.style.width = lineHeightPx;
        };
        return { root, left, right, icon, colorspan, iconLabel, iconContainer, parametersLabel, qualifierLabel, detailsLabel, readMore, disposables, configureFont };
    }
    renderElement(element, index, data) {
        data.configureFont();
        const { completion } = element;
        data.root.id = getAriaId(index);
        data.colorspan.style.backgroundColor = '';
        const labelOptions = {
            labelEscapeNewLines: true,
            matches: (0, filters_js_1.createMatches)(element.score)
        };
        const color = [];
        if (completion.kind === 19 /* CompletionItemKind.Color */ && _completionItemColor.extract(element, color)) {
            // special logic for 'color' completion items
            data.icon.className = 'icon customcolor';
            data.iconContainer.className = 'icon hide';
            data.colorspan.style.backgroundColor = color[0];
        }
        else if (completion.kind === 20 /* CompletionItemKind.File */ && this._themeService.getFileIconTheme().hasFileIcons) {
            // special logic for 'file' completion items
            data.icon.className = 'icon hide';
            data.iconContainer.className = 'icon hide';
            const labelClasses = (0, getIconClasses_js_1.getIconClasses)(this._modelService, this._languageService, uri_js_1.URI.from({ scheme: 'fake', path: element.textLabel }), files_js_1.FileKind.FILE);
            const detailClasses = (0, getIconClasses_js_1.getIconClasses)(this._modelService, this._languageService, uri_js_1.URI.from({ scheme: 'fake', path: completion.detail }), files_js_1.FileKind.FILE);
            labelOptions.extraClasses = labelClasses.length > detailClasses.length ? labelClasses : detailClasses;
        }
        else if (completion.kind === 23 /* CompletionItemKind.Folder */ && this._themeService.getFileIconTheme().hasFolderIcons) {
            // special logic for 'folder' completion items
            data.icon.className = 'icon hide';
            data.iconContainer.className = 'icon hide';
            labelOptions.extraClasses = [
                (0, getIconClasses_js_1.getIconClasses)(this._modelService, this._languageService, uri_js_1.URI.from({ scheme: 'fake', path: element.textLabel }), files_js_1.FileKind.FOLDER),
                (0, getIconClasses_js_1.getIconClasses)(this._modelService, this._languageService, uri_js_1.URI.from({ scheme: 'fake', path: completion.detail }), files_js_1.FileKind.FOLDER)
            ].flat();
        }
        else {
            // normal icon
            data.icon.className = 'icon hide';
            data.iconContainer.className = '';
            data.iconContainer.classList.add('suggest-icon', ...themables_js_1.ThemeIcon.asClassNameArray(languages_js_1.CompletionItemKinds.toIcon(completion.kind)));
        }
        if (completion.tags && completion.tags.indexOf(1 /* CompletionItemTag.Deprecated */) >= 0) {
            labelOptions.extraClasses = (labelOptions.extraClasses || []).concat(['deprecated']);
            labelOptions.matches = [];
        }
        data.iconLabel.setLabel(element.textLabel, undefined, labelOptions);
        if (typeof completion.label === 'string') {
            data.parametersLabel.textContent = '';
            data.detailsLabel.textContent = stripNewLines(completion.detail || '');
            data.root.classList.add('string-label');
        }
        else {
            data.parametersLabel.textContent = stripNewLines(completion.label.detail || '');
            data.detailsLabel.textContent = stripNewLines(completion.label.description || '');
            data.root.classList.remove('string-label');
        }
        if (this._editor.getOption(123 /* EditorOption.suggest */).showInlineDetails) {
            (0, dom_js_1.show)(data.detailsLabel);
        }
        else {
            (0, dom_js_1.hide)(data.detailsLabel);
        }
        if ((0, suggestWidgetDetails_js_1.canExpandCompletionItem)(element)) {
            data.right.classList.add('can-expand-details');
            (0, dom_js_1.show)(data.readMore);
            data.readMore.onmousedown = e => {
                e.stopPropagation();
                e.preventDefault();
            };
            data.readMore.onclick = e => {
                e.stopPropagation();
                e.preventDefault();
                this._onDidToggleDetails.fire();
            };
        }
        else {
            data.right.classList.remove('can-expand-details');
            (0, dom_js_1.hide)(data.readMore);
            data.readMore.onmousedown = null;
            data.readMore.onclick = null;
        }
    }
    disposeTemplate(templateData) {
        templateData.disposables.dispose();
    }
};
exports.ItemRenderer = ItemRenderer;
exports.ItemRenderer = ItemRenderer = __decorate([
    __param(1, model_js_1.IModelService),
    __param(2, language_js_1.ILanguageService),
    __param(3, themeService_js_1.IThemeService)
], ItemRenderer);
function stripNewLines(str) {
    return str.replace(/\r\n|\r|\n/g, '');
}
