// @theia/monaco-editor-core/esm/vs/editor/contrib/fontZoom/browser/fontZoom.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const editorExtensions_js_1 = require("../../../browser/editorExtensions.js");
const editorZoom_js_1 = require("../../../common/config/editorZoom.js");
const nls = require("../../../../nls.js");
class EditorFontZoomIn extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.fontZoomIn',
            label: nls.localize2('EditorFontZoomIn.label', "Increase Editor Font Size"),
            precondition: undefined
        });
    }
    run(accessor, editor) {
        editorZoom_js_1.EditorZoom.setZoomLevel(editorZoom_js_1.EditorZoom.getZoomLevel() + 1);
    }
}
class EditorFontZoomOut extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.fontZoomOut',
            label: nls.localize2('EditorFontZoomOut.label', "Decrease Editor Font Size"),
            precondition: undefined
        });
    }
    run(accessor, editor) {
        editorZoom_js_1.EditorZoom.setZoomLevel(editorZoom_js_1.EditorZoom.getZoomLevel() - 1);
    }
}
class EditorFontZoomReset extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.fontZoomReset',
            label: nls.localize2('EditorFontZoomReset.label', "Reset Editor Font Size"),
            precondition: undefined
        });
    }
    run(accessor, editor) {
        editorZoom_js_1.EditorZoom.setZoomLevel(0);
    }
}
(0, editorExtensions_js_1.registerEditorAction)(EditorFontZoomIn);
(0, editorExtensions_js_1.registerEditorAction)(EditorFontZoomOut);
(0, editorExtensions_js_1.registerEditorAction)(EditorFontZoomReset);
