// @theia/monaco-editor-core/esm/vs/editor/contrib/caretOperations/browser/caretOperations.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const editorExtensions_js_1 = require("../../../browser/editorExtensions.js");
const editorContextKeys_js_1 = require("../../../common/editorContextKeys.js");
const moveCaretCommand_js_1 = require("./moveCaretCommand.js");
const nls = require("../../../../nls.js");
class MoveCaretAction extends editorExtensions_js_1.EditorAction {
    constructor(left, opts) {
        super(opts);
        this.left = left;
    }
    run(accessor, editor) {
        if (!editor.hasModel()) {
            return;
        }
        const commands = [];
        const selections = editor.getSelections();
        for (const selection of selections) {
            commands.push(new moveCaretCommand_js_1.MoveCaretCommand(selection, this.left));
        }
        editor.pushUndoStop();
        editor.executeCommands(this.id, commands);
        editor.pushUndoStop();
    }
}
class MoveCaretLeftAction extends MoveCaretAction {
    constructor() {
        super(true, {
            id: 'editor.action.moveCarretLeftAction',
            label: nls.localize2('caret.moveLeft', "Move Selected Text Left"),
            precondition: editorContextKeys_js_1.EditorContextKeys.writable
        });
    }
}
class MoveCaretRightAction extends MoveCaretAction {
    constructor() {
        super(false, {
            id: 'editor.action.moveCarretRightAction',
            label: nls.localize2('caret.moveRight', "Move Selected Text Right"),
            precondition: editorContextKeys_js_1.EditorContextKeys.writable
        });
    }
}
(0, editorExtensions_js_1.registerEditorAction)(MoveCaretLeftAction);
(0, editorExtensions_js_1.registerEditorAction)(MoveCaretRightAction);
