// @theia/monaco-editor-core/esm/vs/editor/contrib/comment/browser/comment.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const keyCodes_js_1 = require("../../../../base/common/keyCodes.js");
const editorExtensions_js_1 = require("../../../browser/editorExtensions.js");
const range_js_1 = require("../../../common/core/range.js");
const editorContextKeys_js_1 = require("../../../common/editorContextKeys.js");
const languageConfigurationRegistry_js_1 = require("../../../common/languages/languageConfigurationRegistry.js");
const blockCommentCommand_js_1 = require("./blockCommentCommand.js");
const lineCommentCommand_js_1 = require("./lineCommentCommand.js");
const nls = require("../../../../nls.js");
const actions_js_1 = require("../../../../platform/actions/common/actions.js");
class CommentLineAction extends editorExtensions_js_1.EditorAction {
    constructor(type, opts) {
        super(opts);
        this._type = type;
    }
    run(accessor, editor) {
        const languageConfigurationService = accessor.get(languageConfigurationRegistry_js_1.ILanguageConfigurationService);
        if (!editor.hasModel()) {
            return;
        }
        const model = editor.getModel();
        const commands = [];
        const modelOptions = model.getOptions();
        const commentsOptions = editor.getOption(23 /* EditorOption.comments */);
        const selections = editor.getSelections().map((selection, index) => ({ selection, index, ignoreFirstLine: false }));
        selections.sort((a, b) => range_js_1.Range.compareRangesUsingStarts(a.selection, b.selection));
        // Remove selections that would result in copying the same line
        let prev = selections[0];
        for (let i = 1; i < selections.length; i++) {
            const curr = selections[i];
            if (prev.selection.endLineNumber === curr.selection.startLineNumber) {
                // these two selections would copy the same line
                if (prev.index < curr.index) {
                    // prev wins
                    curr.ignoreFirstLine = true;
                }
                else {
                    // curr wins
                    prev.ignoreFirstLine = true;
                    prev = curr;
                }
            }
        }
        for (const selection of selections) {
            commands.push(new lineCommentCommand_js_1.LineCommentCommand(languageConfigurationService, selection.selection, modelOptions.indentSize, this._type, commentsOptions.insertSpace, commentsOptions.ignoreEmptyLines, selection.ignoreFirstLine));
        }
        editor.pushUndoStop();
        editor.executeCommands(this.id, commands);
        editor.pushUndoStop();
    }
}
class ToggleCommentLineAction extends CommentLineAction {
    constructor() {
        super(0 /* Type.Toggle */, {
            id: 'editor.action.commentLine',
            label: nls.localize2('comment.line', "Toggle Line Comment"),
            precondition: editorContextKeys_js_1.EditorContextKeys.writable,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.editorTextFocus,
                primary: 2048 /* KeyMod.CtrlCmd */ | 90 /* KeyCode.Slash */,
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            menuOpts: {
                menuId: actions_js_1.MenuId.MenubarEditMenu,
                group: '5_insert',
                title: nls.localize({ key: 'miToggleLineComment', comment: ['&& denotes a mnemonic'] }, "&&Toggle Line Comment"),
                order: 1
            }
        });
    }
}
class AddLineCommentAction extends CommentLineAction {
    constructor() {
        super(1 /* Type.ForceAdd */, {
            id: 'editor.action.addCommentLine',
            label: nls.localize2('comment.line.add', "Add Line Comment"),
            precondition: editorContextKeys_js_1.EditorContextKeys.writable,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.editorTextFocus,
                primary: (0, keyCodes_js_1.KeyChord)(2048 /* KeyMod.CtrlCmd */ | 41 /* KeyCode.KeyK */, 2048 /* KeyMod.CtrlCmd */ | 33 /* KeyCode.KeyC */),
                weight: 100 /* KeybindingWeight.EditorContrib */
            }
        });
    }
}
class RemoveLineCommentAction extends CommentLineAction {
    constructor() {
        super(2 /* Type.ForceRemove */, {
            id: 'editor.action.removeCommentLine',
            label: nls.localize2('comment.line.remove', "Remove Line Comment"),
            precondition: editorContextKeys_js_1.EditorContextKeys.writable,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.editorTextFocus,
                primary: (0, keyCodes_js_1.KeyChord)(2048 /* KeyMod.CtrlCmd */ | 41 /* KeyCode.KeyK */, 2048 /* KeyMod.CtrlCmd */ | 51 /* KeyCode.KeyU */),
                weight: 100 /* KeybindingWeight.EditorContrib */
            }
        });
    }
}
class BlockCommentAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.blockComment',
            label: nls.localize2('comment.block', "Toggle Block Comment"),
            precondition: editorContextKeys_js_1.EditorContextKeys.writable,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.editorTextFocus,
                primary: 1024 /* KeyMod.Shift */ | 512 /* KeyMod.Alt */ | 31 /* KeyCode.KeyA */,
                linux: { primary: 2048 /* KeyMod.CtrlCmd */ | 1024 /* KeyMod.Shift */ | 31 /* KeyCode.KeyA */ },
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            menuOpts: {
                menuId: actions_js_1.MenuId.MenubarEditMenu,
                group: '5_insert',
                title: nls.localize({ key: 'miToggleBlockComment', comment: ['&& denotes a mnemonic'] }, "Toggle &&Block Comment"),
                order: 2
            }
        });
    }
    run(accessor, editor) {
        const languageConfigurationService = accessor.get(languageConfigurationRegistry_js_1.ILanguageConfigurationService);
        if (!editor.hasModel()) {
            return;
        }
        const commentsOptions = editor.getOption(23 /* EditorOption.comments */);
        const commands = [];
        const selections = editor.getSelections();
        for (const selection of selections) {
            commands.push(new blockCommentCommand_js_1.BlockCommentCommand(selection, commentsOptions.insertSpace, languageConfigurationService));
        }
        editor.pushUndoStop();
        editor.executeCommands(this.id, commands);
        editor.pushUndoStop();
    }
}
(0, editorExtensions_js_1.registerEditorAction)(ToggleCommentLineAction);
(0, editorExtensions_js_1.registerEditorAction)(AddLineCommentAction);
(0, editorExtensions_js_1.registerEditorAction)(RemoveLineCommentAction);
(0, editorExtensions_js_1.registerEditorAction)(BlockCommentAction);
