// @theia/monaco-editor-core/esm/vs/platform/policy/common/policy.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.NullPolicyService = exports.AbstractPolicyService = exports.IPolicyService = void 0;
const event_js_1 = require("../../../base/common/event.js");
const iterator_js_1 = require("../../../base/common/iterator.js");
const lifecycle_js_1 = require("../../../base/common/lifecycle.js");
const instantiation_js_1 = require("../../instantiation/common/instantiation.js");
exports.IPolicyService = (0, instantiation_js_1.createDecorator)('policy');
class AbstractPolicyService extends lifecycle_js_1.Disposable {
    constructor() {
        super(...arguments);
        this.policyDefinitions = {};
        this.policies = new Map();
        this._onDidChange = this._register(new event_js_1.Emitter());
        this.onDidChange = this._onDidChange.event;
    }
    async updatePolicyDefinitions(policyDefinitions) {
        const size = Object.keys(this.policyDefinitions).length;
        this.policyDefinitions = { ...policyDefinitions, ...this.policyDefinitions };
        if (size !== Object.keys(this.policyDefinitions).length) {
            await this._updatePolicyDefinitions(policyDefinitions);
        }
        return iterator_js_1.Iterable.reduce(this.policies.entries(), (r, [name, value]) => ({ ...r, [name]: value }), {});
    }
    getPolicyValue(name) {
        return this.policies.get(name);
    }
    serialize() {
        return iterator_js_1.Iterable.reduce(Object.entries(this.policyDefinitions), (r, [name, definition]) => ({ ...r, [name]: { definition, value: this.policies.get(name) } }), {});
    }
}
exports.AbstractPolicyService = AbstractPolicyService;
class NullPolicyService {
    constructor() {
        this.onDidChange = event_js_1.Event.None;
    }
    async updatePolicyDefinitions() { return {}; }
    getPolicyValue() { return undefined; }
    serialize() { return undefined; }
}
exports.NullPolicyService = NullPolicyService;
