// @theia/monaco-editor-core/esm/vs/editor/standalone/browser/standaloneLayoutService.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditorScopedLayoutService = void 0;
const dom = require("../../../base/browser/dom.js");
const window_js_1 = require("../../../base/browser/window.js");
const arrays_js_1 = require("../../../base/common/arrays.js");
const event_js_1 = require("../../../base/common/event.js");
const codeEditorService_js_1 = require("../../browser/services/codeEditorService.js");
const extensions_js_1 = require("../../../platform/instantiation/common/extensions.js");
const layoutService_js_1 = require("../../../platform/layout/browser/layoutService.js");
let StandaloneLayoutService = class StandaloneLayoutService {
    get mainContainer() {
        return this._codeEditorService.listCodeEditors().at(0)?.getContainerDomNode() ?? window_js_1.mainWindow.document.body;
    }
    get activeContainer() {
        const activeCodeEditor = this._codeEditorService.getFocusedCodeEditor() ?? this._codeEditorService.getActiveCodeEditor();
        return activeCodeEditor?.getContainerDomNode() ?? this.mainContainer;
    }
    get mainContainerDimension() {
        return dom.getClientArea(this.mainContainer);
    }
    get activeContainerDimension() {
        return dom.getClientArea(this.activeContainer);
    }
    get containers() {
        return (0, arrays_js_1.coalesce)(this._codeEditorService.listCodeEditors().map(codeEditor => codeEditor.getContainerDomNode()));
    }
    getContainer() {
        return this.activeContainer;
    }
    whenContainerStylesLoaded() { return undefined; }
    focus() {
        this._codeEditorService.getFocusedCodeEditor()?.focus();
    }
    constructor(_codeEditorService) {
        this._codeEditorService = _codeEditorService;
        this.onDidLayoutMainContainer = event_js_1.Event.None;
        this.onDidLayoutActiveContainer = event_js_1.Event.None;
        this.onDidLayoutContainer = event_js_1.Event.None;
        this.onDidChangeActiveContainer = event_js_1.Event.None;
        this.onDidAddContainer = event_js_1.Event.None;
        this.mainContainerOffset = { top: 0, quickPickTop: 0 };
        this.activeContainerOffset = { top: 0, quickPickTop: 0 };
    }
};
StandaloneLayoutService = __decorate([
    __param(0, codeEditorService_js_1.ICodeEditorService)
], StandaloneLayoutService);
let EditorScopedLayoutService = class EditorScopedLayoutService extends StandaloneLayoutService {
    get mainContainer() {
        return this._container;
    }
    constructor(_container, codeEditorService) {
        super(codeEditorService);
        this._container = _container;
    }
};
exports.EditorScopedLayoutService = EditorScopedLayoutService;
exports.EditorScopedLayoutService = EditorScopedLayoutService = __decorate([
    __param(1, codeEditorService_js_1.ICodeEditorService)
], EditorScopedLayoutService);
(0, extensions_js_1.registerSingleton)(layoutService_js_1.ILayoutService, StandaloneLayoutService, 1 /* InstantiationType.Delayed */);
