// @theia/monaco-editor-core/esm/vs/editor/standalone/browser/standaloneEditor.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
exports.onDidCreateEditor = onDidCreateEditor;
exports.onDidCreateDiffEditor = onDidCreateDiffEditor;
exports.getEditors = getEditors;
exports.getDiffEditors = getDiffEditors;
exports.createDiffEditor = createDiffEditor;
exports.createMultiFileDiffEditor = createMultiFileDiffEditor;
exports.addCommand = addCommand;
exports.addEditorAction = addEditorAction;
exports.addKeybindingRule = addKeybindingRule;
exports.addKeybindingRules = addKeybindingRules;
exports.createModel = createModel;
exports.setModelLanguage = setModelLanguage;
exports.setModelMarkers = setModelMarkers;
exports.removeAllMarkers = removeAllMarkers;
exports.getModelMarkers = getModelMarkers;
exports.onDidChangeMarkers = onDidChangeMarkers;
exports.getModel = getModel;
exports.getModels = getModels;
exports.onDidCreateModel = onDidCreateModel;
exports.onWillDisposeModel = onWillDisposeModel;
exports.onDidChangeModelLanguage = onDidChangeModelLanguage;
exports.createWebWorker = createWebWorker;
exports.colorizeElement = colorizeElement;
exports.colorize = colorize;
exports.colorizeModelLine = colorizeModelLine;
exports.tokenize = tokenize;
exports.defineTheme = defineTheme;
exports.setTheme = setTheme;
exports.remeasureFonts = remeasureFonts;
exports.registerCommand = registerCommand;
exports.registerLinkOpener = registerLinkOpener;
exports.registerEditorOpener = registerEditorOpener;
exports.createMonacoEditorAPI = createMonacoEditorAPI;
const window_js_1 = require("../../../base/browser/window.js");
const lifecycle_js_1 = require("../../../base/common/lifecycle.js");
const strings_js_1 = require("../../../base/common/strings.js");
const uri_js_1 = require("../../../base/common/uri.js");
require("./standalone-tokens.css");
const fontMeasurements_js_1 = require("../../browser/config/fontMeasurements.js");
const editorExtensions_js_1 = require("../../browser/editorExtensions.js");
const codeEditorService_js_1 = require("../../browser/services/codeEditorService.js");
const standaloneWebWorker_js_1 = require("./standaloneWebWorker.js");
const editorOptions_js_1 = require("../../common/config/editorOptions.js");
const editorZoom_js_1 = require("../../common/config/editorZoom.js");
const fontInfo_js_1 = require("../../common/config/fontInfo.js");
const editorCommon_js_1 = require("../../common/editorCommon.js");
const languages = require("../../common/languages.js");
const language_js_1 = require("../../common/languages/language.js");
const modesRegistry_js_1 = require("../../common/languages/modesRegistry.js");
const nullTokenize_js_1 = require("../../common/languages/nullTokenize.js");
const model_js_1 = require("../../common/model.js");
const model_js_2 = require("../../common/services/model.js");
const standaloneEnums = require("../../common/standalone/standaloneEnums.js");
const colorizer_js_1 = require("./colorizer.js");
const standaloneCodeEditor_js_1 = require("./standaloneCodeEditor.js");
const standaloneServices_js_1 = require("./standaloneServices.js");
const standaloneTheme_js_1 = require("../common/standaloneTheme.js");
const actions_js_1 = require("../../../platform/actions/common/actions.js");
const commands_js_1 = require("../../../platform/commands/common/commands.js");
const contextkey_js_1 = require("../../../platform/contextkey/common/contextkey.js");
const keybinding_js_1 = require("../../../platform/keybinding/common/keybinding.js");
const markers_js_1 = require("../../../platform/markers/common/markers.js");
const opener_js_1 = require("../../../platform/opener/common/opener.js");
const multiDiffEditorWidget_js_1 = require("../../browser/widget/multiDiffEditor/multiDiffEditorWidget.js");
/**
 * Create a new editor under `domElement`.
 * `domElement` should be empty (not contain other dom nodes).
 * The editor will read the size of `domElement`.
 */
function create(domElement, options, override) {
    const instantiationService = standaloneServices_js_1.StandaloneServices.initialize(override || {});
    return instantiationService.createInstance(standaloneCodeEditor_js_1.StandaloneEditor, domElement, options);
}
/**
 * Emitted when an editor is created.
 * Creating a diff editor might cause this listener to be invoked with the two editors.
 * @event
 */
function onDidCreateEditor(listener) {
    const codeEditorService = standaloneServices_js_1.StandaloneServices.get(codeEditorService_js_1.ICodeEditorService);
    return codeEditorService.onCodeEditorAdd((editor) => {
        listener(editor);
    });
}
/**
 * Emitted when an diff editor is created.
 * @event
 */
function onDidCreateDiffEditor(listener) {
    const codeEditorService = standaloneServices_js_1.StandaloneServices.get(codeEditorService_js_1.ICodeEditorService);
    return codeEditorService.onDiffEditorAdd((editor) => {
        listener(editor);
    });
}
/**
 * Get all the created editors.
 */
function getEditors() {
    const codeEditorService = standaloneServices_js_1.StandaloneServices.get(codeEditorService_js_1.ICodeEditorService);
    return codeEditorService.listCodeEditors();
}
/**
 * Get all the created diff editors.
 */
function getDiffEditors() {
    const codeEditorService = standaloneServices_js_1.StandaloneServices.get(codeEditorService_js_1.ICodeEditorService);
    return codeEditorService.listDiffEditors();
}
/**
 * Create a new diff editor under `domElement`.
 * `domElement` should be empty (not contain other dom nodes).
 * The editor will read the size of `domElement`.
 */
function createDiffEditor(domElement, options, override) {
    const instantiationService = standaloneServices_js_1.StandaloneServices.initialize(override || {});
    return instantiationService.createInstance(standaloneCodeEditor_js_1.StandaloneDiffEditor2, domElement, options);
}
function createMultiFileDiffEditor(domElement, override) {
    const instantiationService = standaloneServices_js_1.StandaloneServices.initialize(override || {});
    return new multiDiffEditorWidget_js_1.MultiDiffEditorWidget(domElement, {}, instantiationService);
}
/**
 * Add a command.
 */
function addCommand(descriptor) {
    if ((typeof descriptor.id !== 'string') || (typeof descriptor.run !== 'function')) {
        throw new Error('Invalid command descriptor, `id` and `run` are required properties!');
    }
    return commands_js_1.CommandsRegistry.registerCommand(descriptor.id, descriptor.run);
}
/**
 * Add an action to all editors.
 */
function addEditorAction(descriptor) {
    if ((typeof descriptor.id !== 'string') || (typeof descriptor.label !== 'string') || (typeof descriptor.run !== 'function')) {
        throw new Error('Invalid action descriptor, `id`, `label` and `run` are required properties!');
    }
    const precondition = contextkey_js_1.ContextKeyExpr.deserialize(descriptor.precondition);
    const run = (accessor, ...args) => {
        return editorExtensions_js_1.EditorCommand.runEditorCommand(accessor, args, precondition, (accessor, editor, args) => Promise.resolve(descriptor.run(editor, ...args)));
    };
    const toDispose = new lifecycle_js_1.DisposableStore();
    // Register the command
    toDispose.add(commands_js_1.CommandsRegistry.registerCommand(descriptor.id, run));
    // Register the context menu item
    if (descriptor.contextMenuGroupId) {
        const menuItem = {
            command: {
                id: descriptor.id,
                title: descriptor.label
            },
            when: precondition,
            group: descriptor.contextMenuGroupId,
            order: descriptor.contextMenuOrder || 0
        };
        toDispose.add(actions_js_1.MenuRegistry.appendMenuItem(actions_js_1.MenuId.EditorContext, menuItem));
    }
    // Register the keybindings
    if (Array.isArray(descriptor.keybindings)) {
        const keybindingService = standaloneServices_js_1.StandaloneServices.get(keybinding_js_1.IKeybindingService);
        if (!(keybindingService instanceof standaloneServices_js_1.StandaloneKeybindingService)) {
            console.warn('Cannot add keybinding because the editor is configured with an unrecognized KeybindingService');
        }
        else {
            const keybindingsWhen = contextkey_js_1.ContextKeyExpr.and(precondition, contextkey_js_1.ContextKeyExpr.deserialize(descriptor.keybindingContext));
            toDispose.add(keybindingService.addDynamicKeybindings(descriptor.keybindings.map((keybinding) => {
                return {
                    keybinding,
                    command: descriptor.id,
                    when: keybindingsWhen
                };
            })));
        }
    }
    return toDispose;
}
/**
 * Add a keybinding rule.
 */
function addKeybindingRule(rule) {
    return addKeybindingRules([rule]);
}
/**
 * Add keybinding rules.
 */
function addKeybindingRules(rules) {
    const keybindingService = standaloneServices_js_1.StandaloneServices.get(keybinding_js_1.IKeybindingService);
    if (!(keybindingService instanceof standaloneServices_js_1.StandaloneKeybindingService)) {
        console.warn('Cannot add keybinding because the editor is configured with an unrecognized KeybindingService');
        return lifecycle_js_1.Disposable.None;
    }
    return keybindingService.addDynamicKeybindings(rules.map((rule) => {
        return {
            keybinding: rule.keybinding,
            command: rule.command,
            commandArgs: rule.commandArgs,
            when: contextkey_js_1.ContextKeyExpr.deserialize(rule.when),
        };
    }));
}
/**
 * Create a new editor model.
 * You can specify the language that should be set for this model or let the language be inferred from the `uri`.
 */
function createModel(value, language, uri) {
    const languageService = standaloneServices_js_1.StandaloneServices.get(language_js_1.ILanguageService);
    const languageId = languageService.getLanguageIdByMimeType(language) || language;
    return (0, standaloneCodeEditor_js_1.createTextModel)(standaloneServices_js_1.StandaloneServices.get(model_js_2.IModelService), languageService, value, languageId, uri);
}
/**
 * Change the language for a model.
 */
function setModelLanguage(model, mimeTypeOrLanguageId) {
    const languageService = standaloneServices_js_1.StandaloneServices.get(language_js_1.ILanguageService);
    const languageId = languageService.getLanguageIdByMimeType(mimeTypeOrLanguageId) || mimeTypeOrLanguageId || modesRegistry_js_1.PLAINTEXT_LANGUAGE_ID;
    model.setLanguage(languageService.createById(languageId));
}
/**
 * Set the markers for a model.
 */
function setModelMarkers(model, owner, markers) {
    if (model) {
        const markerService = standaloneServices_js_1.StandaloneServices.get(markers_js_1.IMarkerService);
        markerService.changeOne(owner, model.uri, markers);
    }
}
/**
 * Remove all markers of an owner.
 */
function removeAllMarkers(owner) {
    const markerService = standaloneServices_js_1.StandaloneServices.get(markers_js_1.IMarkerService);
    markerService.changeAll(owner, []);
}
/**
 * Get markers for owner and/or resource
 *
 * @returns list of markers
 */
function getModelMarkers(filter) {
    const markerService = standaloneServices_js_1.StandaloneServices.get(markers_js_1.IMarkerService);
    return markerService.read(filter);
}
/**
 * Emitted when markers change for a model.
 * @event
 */
function onDidChangeMarkers(listener) {
    const markerService = standaloneServices_js_1.StandaloneServices.get(markers_js_1.IMarkerService);
    return markerService.onMarkerChanged(listener);
}
/**
 * Get the model that has `uri` if it exists.
 */
function getModel(uri) {
    const modelService = standaloneServices_js_1.StandaloneServices.get(model_js_2.IModelService);
    return modelService.getModel(uri);
}
/**
 * Get all the created models.
 */
function getModels() {
    const modelService = standaloneServices_js_1.StandaloneServices.get(model_js_2.IModelService);
    return modelService.getModels();
}
/**
 * Emitted when a model is created.
 * @event
 */
function onDidCreateModel(listener) {
    const modelService = standaloneServices_js_1.StandaloneServices.get(model_js_2.IModelService);
    return modelService.onModelAdded(listener);
}
/**
 * Emitted right before a model is disposed.
 * @event
 */
function onWillDisposeModel(listener) {
    const modelService = standaloneServices_js_1.StandaloneServices.get(model_js_2.IModelService);
    return modelService.onModelRemoved(listener);
}
/**
 * Emitted when a different language is set to a model.
 * @event
 */
function onDidChangeModelLanguage(listener) {
    const modelService = standaloneServices_js_1.StandaloneServices.get(model_js_2.IModelService);
    return modelService.onModelLanguageChanged((e) => {
        listener({
            model: e.model,
            oldLanguage: e.oldLanguageId
        });
    });
}
/**
 * Create a new web worker that has model syncing capabilities built in.
 * Specify an AMD module to load that will `create` an object that will be proxied.
 */
function createWebWorker(opts) {
    return (0, standaloneWebWorker_js_1.createWebWorker)(standaloneServices_js_1.StandaloneServices.get(model_js_2.IModelService), opts);
}
/**
 * Colorize the contents of `domNode` using attribute `data-lang`.
 */
function colorizeElement(domNode, options) {
    const languageService = standaloneServices_js_1.StandaloneServices.get(language_js_1.ILanguageService);
    const themeService = standaloneServices_js_1.StandaloneServices.get(standaloneTheme_js_1.IStandaloneThemeService);
    return colorizer_js_1.Colorizer.colorizeElement(themeService, languageService, domNode, options).then(() => {
        themeService.registerEditorContainer(domNode);
    });
}
/**
 * Colorize `text` using language `languageId`.
 */
function colorize(text, languageId, options) {
    const languageService = standaloneServices_js_1.StandaloneServices.get(language_js_1.ILanguageService);
    const themeService = standaloneServices_js_1.StandaloneServices.get(standaloneTheme_js_1.IStandaloneThemeService);
    themeService.registerEditorContainer(window_js_1.mainWindow.document.body);
    return colorizer_js_1.Colorizer.colorize(languageService, text, languageId, options);
}
/**
 * Colorize a line in a model.
 */
function colorizeModelLine(model, lineNumber, tabSize = 4) {
    const themeService = standaloneServices_js_1.StandaloneServices.get(standaloneTheme_js_1.IStandaloneThemeService);
    themeService.registerEditorContainer(window_js_1.mainWindow.document.body);
    return colorizer_js_1.Colorizer.colorizeModelLine(model, lineNumber, tabSize);
}
/**
 * @internal
 */
function getSafeTokenizationSupport(language) {
    const tokenizationSupport = languages.TokenizationRegistry.get(language);
    if (tokenizationSupport) {
        return tokenizationSupport;
    }
    return {
        getInitialState: () => nullTokenize_js_1.NullState,
        tokenize: (line, hasEOL, state) => (0, nullTokenize_js_1.nullTokenize)(language, state)
    };
}
/**
 * Tokenize `text` using language `languageId`
 */
function tokenize(text, languageId) {
    // Needed in order to get the mode registered for subsequent look-ups
    languages.TokenizationRegistry.getOrCreate(languageId);
    const tokenizationSupport = getSafeTokenizationSupport(languageId);
    const lines = (0, strings_js_1.splitLines)(text);
    const result = [];
    let state = tokenizationSupport.getInitialState();
    for (let i = 0, len = lines.length; i < len; i++) {
        const line = lines[i];
        const tokenizationResult = tokenizationSupport.tokenize(line, true, state);
        result[i] = tokenizationResult.tokens;
        state = tokenizationResult.endState;
    }
    return result;
}
/**
 * Define a new theme or update an existing theme.
 */
function defineTheme(themeName, themeData) {
    const standaloneThemeService = standaloneServices_js_1.StandaloneServices.get(standaloneTheme_js_1.IStandaloneThemeService);
    standaloneThemeService.defineTheme(themeName, themeData);
}
/**
 * Switches to a theme.
 */
function setTheme(themeName) {
    const standaloneThemeService = standaloneServices_js_1.StandaloneServices.get(standaloneTheme_js_1.IStandaloneThemeService);
    standaloneThemeService.setTheme(themeName);
}
/**
 * Clears all cached font measurements and triggers re-measurement.
 */
function remeasureFonts() {
    fontMeasurements_js_1.FontMeasurements.clearAllFontInfos();
}
/**
 * Register a command.
 */
function registerCommand(id, handler) {
    return commands_js_1.CommandsRegistry.registerCommand({ id, handler });
}
/**
 * Registers a handler that is called when a link is opened in any editor. The handler callback should return `true` if the link was handled and `false` otherwise.
 * The handler that was registered last will be called first when a link is opened.
 *
 * Returns a disposable that can unregister the opener again.
 */
function registerLinkOpener(opener) {
    const openerService = standaloneServices_js_1.StandaloneServices.get(opener_js_1.IOpenerService);
    return openerService.registerOpener({
        async open(resource) {
            if (typeof resource === 'string') {
                resource = uri_js_1.URI.parse(resource);
            }
            return opener.open(resource);
        }
    });
}
/**
 * Registers a handler that is called when a resource other than the current model should be opened in the editor (e.g. "go to definition").
 * The handler callback should return `true` if the request was handled and `false` otherwise.
 *
 * Returns a disposable that can unregister the opener again.
 *
 * If no handler is registered the default behavior is to do nothing for models other than the currently attached one.
 */
function registerEditorOpener(opener) {
    const codeEditorService = standaloneServices_js_1.StandaloneServices.get(codeEditorService_js_1.ICodeEditorService);
    return codeEditorService.registerCodeEditorOpenHandler(async (input, source, sideBySide) => {
        if (!source) {
            return null;
        }
        const selection = input.options?.selection;
        let selectionOrPosition;
        if (selection && typeof selection.endLineNumber === 'number' && typeof selection.endColumn === 'number') {
            selectionOrPosition = selection;
        }
        else if (selection) {
            selectionOrPosition = { lineNumber: selection.startLineNumber, column: selection.startColumn };
        }
        if (await opener.openCodeEditor(source, input.resource, selectionOrPosition)) {
            return source; // return source editor to indicate that this handler has successfully handled the opening
        }
        return null; // fallback to other registered handlers
    });
}
/**
 * @internal
 */
function createMonacoEditorAPI() {
    return {
        // methods
        create: create,
        getEditors: getEditors,
        getDiffEditors: getDiffEditors,
        onDidCreateEditor: onDidCreateEditor,
        onDidCreateDiffEditor: onDidCreateDiffEditor,
        createDiffEditor: createDiffEditor,
        addCommand: addCommand,
        addEditorAction: addEditorAction,
        addKeybindingRule: addKeybindingRule,
        addKeybindingRules: addKeybindingRules,
        createModel: createModel,
        setModelLanguage: setModelLanguage,
        setModelMarkers: setModelMarkers,
        getModelMarkers: getModelMarkers,
        removeAllMarkers: removeAllMarkers,
        onDidChangeMarkers: onDidChangeMarkers,
        getModels: getModels,
        getModel: getModel,
        onDidCreateModel: onDidCreateModel,
        onWillDisposeModel: onWillDisposeModel,
        onDidChangeModelLanguage: onDidChangeModelLanguage,
        createWebWorker: createWebWorker,
        colorizeElement: colorizeElement,
        colorize: colorize,
        colorizeModelLine: colorizeModelLine,
        tokenize: tokenize,
        defineTheme: defineTheme,
        setTheme: setTheme,
        remeasureFonts: remeasureFonts,
        registerCommand: registerCommand,
        registerLinkOpener: registerLinkOpener,
        registerEditorOpener: registerEditorOpener,
        // enums
        AccessibilitySupport: standaloneEnums.AccessibilitySupport,
        ContentWidgetPositionPreference: standaloneEnums.ContentWidgetPositionPreference,
        CursorChangeReason: standaloneEnums.CursorChangeReason,
        DefaultEndOfLine: standaloneEnums.DefaultEndOfLine,
        EditorAutoIndentStrategy: standaloneEnums.EditorAutoIndentStrategy,
        EditorOption: standaloneEnums.EditorOption,
        EndOfLinePreference: standaloneEnums.EndOfLinePreference,
        EndOfLineSequence: standaloneEnums.EndOfLineSequence,
        MinimapPosition: standaloneEnums.MinimapPosition,
        MinimapSectionHeaderStyle: standaloneEnums.MinimapSectionHeaderStyle,
        MouseTargetType: standaloneEnums.MouseTargetType,
        OverlayWidgetPositionPreference: standaloneEnums.OverlayWidgetPositionPreference,
        OverviewRulerLane: standaloneEnums.OverviewRulerLane,
        GlyphMarginLane: standaloneEnums.GlyphMarginLane,
        RenderLineNumbersType: standaloneEnums.RenderLineNumbersType,
        RenderMinimap: standaloneEnums.RenderMinimap,
        ScrollbarVisibility: standaloneEnums.ScrollbarVisibility,
        ScrollType: standaloneEnums.ScrollType,
        TextEditorCursorBlinkingStyle: standaloneEnums.TextEditorCursorBlinkingStyle,
        TextEditorCursorStyle: standaloneEnums.TextEditorCursorStyle,
        TrackedRangeStickiness: standaloneEnums.TrackedRangeStickiness,
        WrappingIndent: standaloneEnums.WrappingIndent,
        InjectedTextCursorStops: standaloneEnums.InjectedTextCursorStops,
        PositionAffinity: standaloneEnums.PositionAffinity,
        ShowLightbulbIconMode: standaloneEnums.ShowLightbulbIconMode,
        // classes
        ConfigurationChangedEvent: editorOptions_js_1.ConfigurationChangedEvent,
        BareFontInfo: fontInfo_js_1.BareFontInfo,
        FontInfo: fontInfo_js_1.FontInfo,
        TextModelResolvedOptions: model_js_1.TextModelResolvedOptions,
        FindMatch: model_js_1.FindMatch,
        ApplyUpdateResult: editorOptions_js_1.ApplyUpdateResult,
        EditorZoom: editorZoom_js_1.EditorZoom,
        createMultiFileDiffEditor: createMultiFileDiffEditor,
        // vars
        EditorType: editorCommon_js_1.EditorType,
        EditorOptions: editorOptions_js_1.EditorOptions
    };
}
