// @theia/monaco-editor-core/esm/vs/editor/contrib/toggleTabFocusMode/browser/toggleTabFocusMode.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToggleTabFocusModeAction = void 0;
const aria_js_1 = require("../../../../base/browser/ui/aria/aria.js");
const tabFocus_js_1 = require("../../../browser/config/tabFocus.js");
const nls = require("../../../../nls.js");
const actions_js_1 = require("../../../../platform/actions/common/actions.js");
class ToggleTabFocusModeAction extends actions_js_1.Action2 {
    static { this.ID = 'editor.action.toggleTabFocusMode'; }
    constructor() {
        super({
            id: ToggleTabFocusModeAction.ID,
            title: nls.localize2({ key: 'toggle.tabMovesFocus', comment: ['Turn on/off use of tab key for moving focus around VS Code'] }, 'Toggle Tab Key Moves Focus'),
            precondition: undefined,
            keybinding: {
                primary: 2048 /* KeyMod.CtrlCmd */ | 43 /* KeyCode.KeyM */,
                mac: { primary: 256 /* KeyMod.WinCtrl */ | 1024 /* KeyMod.Shift */ | 43 /* KeyCode.KeyM */ },
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            metadata: {
                description: nls.localize2('tabMovesFocusDescriptions', "Determines whether the tab key moves focus around the workbench or inserts the tab character in the current editor. This is also called tab trapping, tab navigation, or tab focus mode."),
            },
            f1: true
        });
    }
    run() {
        const oldValue = tabFocus_js_1.TabFocus.getTabFocusMode();
        const newValue = !oldValue;
        tabFocus_js_1.TabFocus.setTabFocusMode(newValue);
        if (newValue) {
            (0, aria_js_1.alert)(nls.localize('toggle.tabMovesFocus.on', "Pressing Tab will now move focus to the next focusable element"));
        }
        else {
            (0, aria_js_1.alert)(nls.localize('toggle.tabMovesFocus.off', "Pressing Tab will now insert the tab character"));
        }
    }
}
exports.ToggleTabFocusModeAction = ToggleTabFocusModeAction;
(0, actions_js_1.registerAction2)(ToggleTabFocusModeAction);
