// @theia/monaco-editor-core/esm/vs/editor/contrib/gotoSymbol/browser/peek/referencesTree.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var FileReferencesRenderer_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessibilityProvider = exports.OneReferenceRenderer = exports.FileReferencesRenderer = exports.IdentityProvider = exports.StringRepresentationProvider = exports.Delegate = exports.DataSource = void 0;
const dom = require("../../../../../base/browser/dom.js");
const countBadge_js_1 = require("../../../../../base/browser/ui/countBadge/countBadge.js");
const highlightedLabel_js_1 = require("../../../../../base/browser/ui/highlightedlabel/highlightedLabel.js");
const iconLabel_js_1 = require("../../../../../base/browser/ui/iconLabel/iconLabel.js");
const filters_js_1 = require("../../../../../base/common/filters.js");
const lifecycle_js_1 = require("../../../../../base/common/lifecycle.js");
const resources_js_1 = require("../../../../../base/common/resources.js");
const resolverService_js_1 = require("../../../../common/services/resolverService.js");
const nls_js_1 = require("../../../../../nls.js");
const instantiation_js_1 = require("../../../../../platform/instantiation/common/instantiation.js");
const keybinding_js_1 = require("../../../../../platform/keybinding/common/keybinding.js");
const label_js_1 = require("../../../../../platform/label/common/label.js");
const defaultStyles_js_1 = require("../../../../../platform/theme/browser/defaultStyles.js");
const referencesModel_js_1 = require("../referencesModel.js");
let DataSource = class DataSource {
    constructor(_resolverService) {
        this._resolverService = _resolverService;
    }
    hasChildren(element) {
        if (element instanceof referencesModel_js_1.ReferencesModel) {
            return true;
        }
        if (element instanceof referencesModel_js_1.FileReferences) {
            return true;
        }
        return false;
    }
    getChildren(element) {
        if (element instanceof referencesModel_js_1.ReferencesModel) {
            return element.groups;
        }
        if (element instanceof referencesModel_js_1.FileReferences) {
            return element.resolve(this._resolverService).then(val => {
                // if (element.failure) {
                // 	// refresh the element on failure so that
                // 	// we can update its rendering
                // 	return tree.refresh(element).then(() => val.children);
                // }
                return val.children;
            });
        }
        throw new Error('bad tree');
    }
};
exports.DataSource = DataSource;
exports.DataSource = DataSource = __decorate([
    __param(0, resolverService_js_1.ITextModelService)
], DataSource);
//#endregion
class Delegate {
    getHeight() {
        return 23;
    }
    getTemplateId(element) {
        if (element instanceof referencesModel_js_1.FileReferences) {
            return FileReferencesRenderer.id;
        }
        else {
            return OneReferenceRenderer.id;
        }
    }
}
exports.Delegate = Delegate;
let StringRepresentationProvider = class StringRepresentationProvider {
    constructor(_keybindingService) {
        this._keybindingService = _keybindingService;
    }
    getKeyboardNavigationLabel(element) {
        if (element instanceof referencesModel_js_1.OneReference) {
            const parts = element.parent.getPreview(element)?.preview(element.range);
            if (parts) {
                return parts.value;
            }
        }
        // FileReferences or unresolved OneReference
        return (0, resources_js_1.basename)(element.uri);
    }
    mightProducePrintableCharacter(event) {
        return this._keybindingService.mightProducePrintableCharacter(event);
    }
};
exports.StringRepresentationProvider = StringRepresentationProvider;
exports.StringRepresentationProvider = StringRepresentationProvider = __decorate([
    __param(0, keybinding_js_1.IKeybindingService)
], StringRepresentationProvider);
class IdentityProvider {
    getId(element) {
        return element instanceof referencesModel_js_1.OneReference ? element.id : element.uri;
    }
}
exports.IdentityProvider = IdentityProvider;
//#region render: File
let FileReferencesTemplate = class FileReferencesTemplate extends lifecycle_js_1.Disposable {
    constructor(container, _labelService) {
        super();
        this._labelService = _labelService;
        const parent = document.createElement('div');
        parent.classList.add('reference-file');
        this.file = this._register(new iconLabel_js_1.IconLabel(parent, { supportHighlights: true }));
        this.badge = this._register(new countBadge_js_1.CountBadge(dom.append(parent, dom.$('.count')), {}, defaultStyles_js_1.defaultCountBadgeStyles));
        container.appendChild(parent);
    }
    set(element, matches) {
        const parent = (0, resources_js_1.dirname)(element.uri);
        this.file.setLabel(this._labelService.getUriBasenameLabel(element.uri), this._labelService.getUriLabel(parent, { relative: true }), { title: this._labelService.getUriLabel(element.uri), matches });
        const len = element.children.length;
        this.badge.setCount(len);
        if (len > 1) {
            this.badge.setTitleFormat((0, nls_js_1.localize)('referencesCount', "{0} references", len));
        }
        else {
            this.badge.setTitleFormat((0, nls_js_1.localize)('referenceCount', "{0} reference", len));
        }
    }
};
FileReferencesTemplate = __decorate([
    __param(1, label_js_1.ILabelService)
], FileReferencesTemplate);
let FileReferencesRenderer = class FileReferencesRenderer {
    static { FileReferencesRenderer_1 = this; }
    static { this.id = 'FileReferencesRenderer'; }
    constructor(_instantiationService) {
        this._instantiationService = _instantiationService;
        this.templateId = FileReferencesRenderer_1.id;
    }
    renderTemplate(container) {
        return this._instantiationService.createInstance(FileReferencesTemplate, container);
    }
    renderElement(node, index, template) {
        template.set(node.element, (0, filters_js_1.createMatches)(node.filterData));
    }
    disposeTemplate(templateData) {
        templateData.dispose();
    }
};
exports.FileReferencesRenderer = FileReferencesRenderer;
exports.FileReferencesRenderer = FileReferencesRenderer = FileReferencesRenderer_1 = __decorate([
    __param(0, instantiation_js_1.IInstantiationService)
], FileReferencesRenderer);
//#endregion
//#region render: Reference
class OneReferenceTemplate extends lifecycle_js_1.Disposable {
    constructor(container) {
        super();
        this.label = this._register(new highlightedLabel_js_1.HighlightedLabel(container));
    }
    set(element, score) {
        const preview = element.parent.getPreview(element)?.preview(element.range);
        if (!preview || !preview.value) {
            // this means we FAILED to resolve the document or the value is the empty string
            this.label.set(`${(0, resources_js_1.basename)(element.uri)}:${element.range.startLineNumber + 1}:${element.range.startColumn + 1}`);
        }
        else {
            // render search match as highlight unless
            // we have score, then render the score
            const { value, highlight } = preview;
            if (score && !filters_js_1.FuzzyScore.isDefault(score)) {
                this.label.element.classList.toggle('referenceMatch', false);
                this.label.set(value, (0, filters_js_1.createMatches)(score));
            }
            else {
                this.label.element.classList.toggle('referenceMatch', true);
                this.label.set(value, [highlight]);
            }
        }
    }
}
class OneReferenceRenderer {
    constructor() {
        this.templateId = OneReferenceRenderer.id;
    }
    static { this.id = 'OneReferenceRenderer'; }
    renderTemplate(container) {
        return new OneReferenceTemplate(container);
    }
    renderElement(node, index, templateData) {
        templateData.set(node.element, node.filterData);
    }
    disposeTemplate(templateData) {
        templateData.dispose();
    }
}
exports.OneReferenceRenderer = OneReferenceRenderer;
//#endregion
class AccessibilityProvider {
    getWidgetAriaLabel() {
        return (0, nls_js_1.localize)('treeAriaLabel', "References");
    }
    getAriaLabel(element) {
        return element.ariaMessage;
    }
}
exports.AccessibilityProvider = AccessibilityProvider;
