// @theia/monaco-editor-core/esm/vs/platform/history/browser/contextScopedHistoryWidget.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextScopedReplaceInput = exports.ContextScopedFindInput = exports.ContextScopedHistoryInputBox = exports.historyNavigationVisible = void 0;
exports.registerAndCreateHistoryNavigationContext = registerAndCreateHistoryNavigationContext;
const findInput_js_1 = require("../../../base/browser/ui/findinput/findInput.js");
const replaceInput_js_1 = require("../../../base/browser/ui/findinput/replaceInput.js");
const inputBox_js_1 = require("../../../base/browser/ui/inputbox/inputBox.js");
const contextkey_js_1 = require("../../contextkey/common/contextkey.js");
const keybindingsRegistry_js_1 = require("../../keybinding/common/keybindingsRegistry.js");
const nls_js_1 = require("../../../nls.js");
const lifecycle_js_1 = require("../../../base/common/lifecycle.js");
const dom_js_1 = require("../../../base/browser/dom.js");
exports.historyNavigationVisible = new contextkey_js_1.RawContextKey('suggestWidgetVisible', false, (0, nls_js_1.localize)('suggestWidgetVisible', "Whether suggestion are visible"));
const HistoryNavigationWidgetFocusContext = 'historyNavigationWidgetFocus';
const HistoryNavigationForwardsEnablementContext = 'historyNavigationForwardsEnabled';
const HistoryNavigationBackwardsEnablementContext = 'historyNavigationBackwardsEnabled';
let lastFocusedWidget = undefined;
const widgets = [];
function registerAndCreateHistoryNavigationContext(scopedContextKeyService, widget) {
    if (widgets.includes(widget)) {
        throw new Error('Cannot register the same widget multiple times');
    }
    widgets.push(widget);
    const disposableStore = new lifecycle_js_1.DisposableStore();
    const historyNavigationWidgetFocus = new contextkey_js_1.RawContextKey(HistoryNavigationWidgetFocusContext, false).bindTo(scopedContextKeyService);
    const historyNavigationForwardsEnablement = new contextkey_js_1.RawContextKey(HistoryNavigationForwardsEnablementContext, true).bindTo(scopedContextKeyService);
    const historyNavigationBackwardsEnablement = new contextkey_js_1.RawContextKey(HistoryNavigationBackwardsEnablementContext, true).bindTo(scopedContextKeyService);
    const onDidFocus = () => {
        historyNavigationWidgetFocus.set(true);
        lastFocusedWidget = widget;
    };
    const onDidBlur = () => {
        historyNavigationWidgetFocus.set(false);
        if (lastFocusedWidget === widget) {
            lastFocusedWidget = undefined;
        }
    };
    // Check for currently being focused
    if ((0, dom_js_1.isActiveElement)(widget.element)) {
        onDidFocus();
    }
    disposableStore.add(widget.onDidFocus(() => onDidFocus()));
    disposableStore.add(widget.onDidBlur(() => onDidBlur()));
    disposableStore.add((0, lifecycle_js_1.toDisposable)(() => {
        widgets.splice(widgets.indexOf(widget), 1);
        onDidBlur();
    }));
    return {
        historyNavigationForwardsEnablement,
        historyNavigationBackwardsEnablement,
        dispose() {
            disposableStore.dispose();
        }
    };
}
let ContextScopedHistoryInputBox = class ContextScopedHistoryInputBox extends inputBox_js_1.HistoryInputBox {
    constructor(container, contextViewProvider, options, contextKeyService) {
        super(container, contextViewProvider, options);
        const scopedContextKeyService = this._register(contextKeyService.createScoped(this.element));
        this._register(registerAndCreateHistoryNavigationContext(scopedContextKeyService, this));
    }
};
exports.ContextScopedHistoryInputBox = ContextScopedHistoryInputBox;
exports.ContextScopedHistoryInputBox = ContextScopedHistoryInputBox = __decorate([
    __param(3, contextkey_js_1.IContextKeyService)
], ContextScopedHistoryInputBox);
let ContextScopedFindInput = class ContextScopedFindInput extends findInput_js_1.FindInput {
    constructor(container, contextViewProvider, options, contextKeyService) {
        super(container, contextViewProvider, options);
        const scopedContextKeyService = this._register(contextKeyService.createScoped(this.inputBox.element));
        this._register(registerAndCreateHistoryNavigationContext(scopedContextKeyService, this.inputBox));
    }
};
exports.ContextScopedFindInput = ContextScopedFindInput;
exports.ContextScopedFindInput = ContextScopedFindInput = __decorate([
    __param(3, contextkey_js_1.IContextKeyService)
], ContextScopedFindInput);
let ContextScopedReplaceInput = class ContextScopedReplaceInput extends replaceInput_js_1.ReplaceInput {
    constructor(container, contextViewProvider, options, contextKeyService, showReplaceOptions = false) {
        super(container, contextViewProvider, showReplaceOptions, options);
        const scopedContextKeyService = this._register(contextKeyService.createScoped(this.inputBox.element));
        this._register(registerAndCreateHistoryNavigationContext(scopedContextKeyService, this.inputBox));
    }
};
exports.ContextScopedReplaceInput = ContextScopedReplaceInput;
exports.ContextScopedReplaceInput = ContextScopedReplaceInput = __decorate([
    __param(3, contextkey_js_1.IContextKeyService)
], ContextScopedReplaceInput);
keybindingsRegistry_js_1.KeybindingsRegistry.registerCommandAndKeybindingRule({
    id: 'history.showPrevious',
    weight: 200 /* KeybindingWeight.WorkbenchContrib */,
    when: contextkey_js_1.ContextKeyExpr.and(contextkey_js_1.ContextKeyExpr.has(HistoryNavigationWidgetFocusContext), contextkey_js_1.ContextKeyExpr.equals(HistoryNavigationBackwardsEnablementContext, true), contextkey_js_1.ContextKeyExpr.not('isComposing'), exports.historyNavigationVisible.isEqualTo(false)),
    primary: 16 /* KeyCode.UpArrow */,
    secondary: [512 /* KeyMod.Alt */ | 16 /* KeyCode.UpArrow */],
    handler: (accessor) => {
        lastFocusedWidget?.showPreviousValue();
    }
});
keybindingsRegistry_js_1.KeybindingsRegistry.registerCommandAndKeybindingRule({
    id: 'history.showNext',
    weight: 200 /* KeybindingWeight.WorkbenchContrib */,
    when: contextkey_js_1.ContextKeyExpr.and(contextkey_js_1.ContextKeyExpr.has(HistoryNavigationWidgetFocusContext), contextkey_js_1.ContextKeyExpr.equals(HistoryNavigationForwardsEnablementContext, true), contextkey_js_1.ContextKeyExpr.not('isComposing'), exports.historyNavigationVisible.isEqualTo(false)),
    primary: 18 /* KeyCode.DownArrow */,
    secondary: [512 /* KeyMod.Alt */ | 18 /* KeyCode.DownArrow */],
    handler: (accessor) => {
        lastFocusedWidget?.showNextValue();
    }
});
