// @theia/monaco-editor-core/esm/vs/base/browser/ui/dropdown/dropdown.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.DropdownMenu = void 0;
const dom_js_1 = require("../../dom.js");
const keyboardEvent_js_1 = require("../../keyboardEvent.js");
const touch_js_1 = require("../../touch.js");
const hoverDelegate2_js_1 = require("../hover/hoverDelegate2.js");
const hoverDelegateFactory_js_1 = require("../hover/hoverDelegateFactory.js");
const actions_js_1 = require("../../../common/actions.js");
const event_js_1 = require("../../../common/event.js");
require("./dropdown.css");
class BaseDropdown extends actions_js_1.ActionRunner {
    constructor(container, options) {
        super();
        this._onDidChangeVisibility = this._register(new event_js_1.Emitter());
        this.onDidChangeVisibility = this._onDidChangeVisibility.event;
        this._element = (0, dom_js_1.append)(container, (0, dom_js_1.$)('.monaco-dropdown'));
        this._label = (0, dom_js_1.append)(this._element, (0, dom_js_1.$)('.dropdown-label'));
        let labelRenderer = options.labelRenderer;
        if (!labelRenderer) {
            labelRenderer = (container) => {
                container.textContent = options.label || '';
                return null;
            };
        }
        for (const event of [dom_js_1.EventType.CLICK, dom_js_1.EventType.MOUSE_DOWN, touch_js_1.EventType.Tap]) {
            this._register((0, dom_js_1.addDisposableListener)(this.element, event, e => dom_js_1.EventHelper.stop(e, true))); // prevent default click behaviour to trigger
        }
        for (const event of [dom_js_1.EventType.MOUSE_DOWN, touch_js_1.EventType.Tap]) {
            this._register((0, dom_js_1.addDisposableListener)(this._label, event, e => {
                if ((0, dom_js_1.isMouseEvent)(e) && (e.detail > 1 || e.button !== 0)) {
                    // prevent right click trigger to allow separate context menu (https://github.com/microsoft/vscode/issues/151064)
                    // prevent multiple clicks to open multiple context menus (https://github.com/microsoft/vscode/issues/41363)
                    return;
                }
                if (this.visible) {
                    this.hide();
                }
                else {
                    this.show();
                }
            }));
        }
        this._register((0, dom_js_1.addDisposableListener)(this._label, dom_js_1.EventType.KEY_UP, e => {
            const event = new keyboardEvent_js_1.StandardKeyboardEvent(e);
            if (event.equals(3 /* KeyCode.Enter */) || event.equals(10 /* KeyCode.Space */)) {
                dom_js_1.EventHelper.stop(e, true); // https://github.com/microsoft/vscode/issues/57997
                if (this.visible) {
                    this.hide();
                }
                else {
                    this.show();
                }
            }
        }));
        const cleanupFn = labelRenderer(this._label);
        if (cleanupFn) {
            this._register(cleanupFn);
        }
        this._register(touch_js_1.Gesture.addTarget(this._label));
    }
    get element() {
        return this._element;
    }
    get label() {
        return this._label;
    }
    set tooltip(tooltip) {
        if (this._label) {
            if (!this.hover && tooltip !== '') {
                this.hover = this._register((0, hoverDelegate2_js_1.getBaseLayerHoverDelegate)().setupManagedHover((0, hoverDelegateFactory_js_1.getDefaultHoverDelegate)('mouse'), this._label, tooltip));
            }
            else if (this.hover) {
                this.hover.update(tooltip);
            }
        }
    }
    show() {
        if (!this.visible) {
            this.visible = true;
            this._onDidChangeVisibility.fire(true);
        }
    }
    hide() {
        if (this.visible) {
            this.visible = false;
            this._onDidChangeVisibility.fire(false);
        }
    }
    isVisible() {
        return !!this.visible;
    }
    onEvent(_e, activeElement) {
        this.hide();
    }
    dispose() {
        super.dispose();
        this.hide();
        if (this.boxContainer) {
            this.boxContainer.remove();
            this.boxContainer = undefined;
        }
        if (this.contents) {
            this.contents.remove();
            this.contents = undefined;
        }
        if (this._label) {
            this._label.remove();
            this._label = undefined;
        }
    }
}
class DropdownMenu extends BaseDropdown {
    constructor(container, _options) {
        super(container, _options);
        this._options = _options;
        this._actions = [];
        this.actions = _options.actions || [];
    }
    set menuOptions(options) {
        this._menuOptions = options;
    }
    get menuOptions() {
        return this._menuOptions;
    }
    get actions() {
        if (this._options.actionProvider) {
            return this._options.actionProvider.getActions();
        }
        return this._actions;
    }
    set actions(actions) {
        this._actions = actions;
    }
    show() {
        super.show();
        this.element.classList.add('active');
        this._options.contextMenuProvider.showContextMenu({
            getAnchor: () => this.element,
            getActions: () => this.actions,
            getActionsContext: () => this.menuOptions ? this.menuOptions.context : null,
            getActionViewItem: (action, options) => this.menuOptions && this.menuOptions.actionViewItemProvider ? this.menuOptions.actionViewItemProvider(action, options) : undefined,
            getKeyBinding: action => this.menuOptions && this.menuOptions.getKeyBinding ? this.menuOptions.getKeyBinding(action) : undefined,
            getMenuClassName: () => this._options.menuClassName || '',
            onHide: () => this.onHide(),
            actionRunner: this.menuOptions ? this.menuOptions.actionRunner : undefined,
            anchorAlignment: this.menuOptions ? this.menuOptions.anchorAlignment : 0 /* AnchorAlignment.LEFT */,
            domForShadowRoot: this._options.menuAsChild ? this.element : undefined,
            skipTelemetry: this._options.skipTelemetry
        });
    }
    hide() {
        super.hide();
    }
    onHide() {
        this.hide();
        this.element.classList.remove('active');
    }
}
exports.DropdownMenu = DropdownMenu;
