// @theia/core/lib/browser/opener-service.js
"use strict";
// *****************************************************************************
// Copyright (C) 2017 TypeFox and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0.
//
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License v. 2.0 are satisfied: GNU General Public License, version 2
// with the GNU Classpath Exception which is available at
// https://www.gnu.org/software/classpath/license.html.
//
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
// *****************************************************************************
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultOpenerService = exports.defaultHandlerPriority = exports.getDefaultHandler = exports.open = exports.OpenerService = exports.OpenHandler = void 0;
const tslib_1 = require("tslib");
const inversify_1 = require("inversify");
const common_1 = require("../common");
const glob_1 = require("../common/glob");
exports.OpenHandler = Symbol('OpenHandler');
exports.OpenerService = Symbol('OpenerService');
async function open(openerService, uri, options) {
    const opener = await openerService.getOpener(uri, options);
    return opener.open(uri, options);
}
exports.open = open;
function getDefaultHandler(uri, preferenceService) {
    var _a;
    const associations = preferenceService.get('workbench.editorAssociations', {});
    const defaultHandler = (_a = Object.entries(associations).find(([key]) => (0, glob_1.match)(key, uri.path.base))) === null || _a === void 0 ? void 0 : _a[1];
    if (typeof defaultHandler === 'string') {
        return defaultHandler;
    }
    return undefined;
}
exports.getDefaultHandler = getDefaultHandler;
exports.defaultHandlerPriority = 100000;
let DefaultOpenerService = class DefaultOpenerService {
    constructor(handlersProvider) {
        this.handlersProvider = handlersProvider;
        // Collection of open-handlers for custom-editor contributions.
        this.customEditorOpenHandlers = [];
        this.onDidChangeOpenersEmitter = new common_1.Emitter();
        this.onDidChangeOpeners = this.onDidChangeOpenersEmitter.event;
    }
    addHandler(openHandler) {
        this.customEditorOpenHandlers.push(openHandler);
        this.onDidChangeOpenersEmitter.fire();
        return common_1.Disposable.create(() => {
            this.removeHandler(openHandler);
        });
    }
    removeHandler(openHandler) {
        this.customEditorOpenHandlers.splice(this.customEditorOpenHandlers.indexOf(openHandler), 1);
        this.onDidChangeOpenersEmitter.fire();
    }
    async getOpener(uri, options) {
        const handlers = await this.prioritize(uri, options);
        if (handlers.length >= 1) {
            return handlers[0];
        }
        return Promise.reject(new Error(`There is no opener for ${uri}.`));
    }
    async getOpeners(uri, options) {
        return uri ? this.prioritize(uri, options) : this.getHandlers();
    }
    async prioritize(uri, options) {
        const prioritized = await common_1.Prioritizeable.prioritizeAll(this.getHandlers(), async (handler) => {
            try {
                return await handler.canHandle(uri, options);
            }
            catch {
                return 0;
            }
        });
        return prioritized.map(p => p.value);
    }
    getHandlers() {
        return [
            ...this.handlersProvider.getContributions(),
            ...this.customEditorOpenHandlers
        ];
    }
};
exports.DefaultOpenerService = DefaultOpenerService;
exports.DefaultOpenerService = DefaultOpenerService = tslib_1.__decorate([
    (0, inversify_1.injectable)(),
    tslib_1.__param(0, (0, inversify_1.inject)(common_1.ContributionProvider)),
    tslib_1.__param(0, (0, inversify_1.named)(exports.OpenHandler)),
    tslib_1.__metadata("design:paramtypes", [Object])
], DefaultOpenerService);
