// @theia/monaco-editor-core/esm/vs/platform/actions/browser/actionViewItemService.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.NullActionViewItemService = exports.IActionViewItemService = void 0;
const event_js_1 = require("../../../base/common/event.js");
const lifecycle_js_1 = require("../../../base/common/lifecycle.js");
const extensions_js_1 = require("../../instantiation/common/extensions.js");
const instantiation_js_1 = require("../../instantiation/common/instantiation.js");
const actions_js_1 = require("../common/actions.js");
exports.IActionViewItemService = (0, instantiation_js_1.createDecorator)('IActionViewItemService');
class NullActionViewItemService {
    constructor() {
        this.onDidChange = event_js_1.Event.None;
    }
    register(menu, commandId, provider, event) {
        return lifecycle_js_1.Disposable.None;
    }
    lookUp(menu, commandId) {
        return undefined;
    }
}
exports.NullActionViewItemService = NullActionViewItemService;
class ActionViewItemService {
    constructor() {
        this._providers = new Map();
        this._onDidChange = new event_js_1.Emitter();
        this.onDidChange = this._onDidChange.event;
    }
    dispose() {
        this._onDidChange.dispose();
    }
    register(menu, commandOrSubmenuId, provider, event) {
        const id = this._makeKey(menu, commandOrSubmenuId);
        if (this._providers.has(id)) {
            throw new Error(`A provider for the command ${commandOrSubmenuId} and menu ${menu} is already registered.`);
        }
        this._providers.set(id, provider);
        const listener = event?.(() => {
            this._onDidChange.fire(menu);
        });
        return (0, lifecycle_js_1.toDisposable)(() => {
            listener?.dispose();
            this._providers.delete(id);
        });
    }
    lookUp(menu, commandOrMenuId) {
        return this._providers.get(this._makeKey(menu, commandOrMenuId));
    }
    _makeKey(menu, commandOrMenuId) {
        return `${menu.id}/${(commandOrMenuId instanceof actions_js_1.MenuId ? commandOrMenuId.id : commandOrMenuId)}`;
    }
}
(0, extensions_js_1.registerSingleton)(exports.IActionViewItemService, ActionViewItemService, 1 /* InstantiationType.Delayed */);
