// @theia/monaco-editor-core/esm/vs/platform/quickinput/browser/quickInputService.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuickInputService = void 0;
const cancellation_js_1 = require("../../../base/common/cancellation.js");
const event_js_1 = require("../../../base/common/event.js");
const contextkey_js_1 = require("../../contextkey/common/contextkey.js");
const instantiation_js_1 = require("../../instantiation/common/instantiation.js");
const layoutService_js_1 = require("../../layout/browser/layoutService.js");
const opener_js_1 = require("../../opener/common/opener.js");
const quickAccess_js_1 = require("./quickAccess.js");
const defaultStyles_js_1 = require("../../theme/browser/defaultStyles.js");
const colorRegistry_js_1 = require("../../theme/common/colorRegistry.js");
const themeService_js_1 = require("../../theme/common/themeService.js");
const quickInput_js_1 = require("./quickInput.js");
const quickInputController_js_1 = require("./quickInputController.js");
const configuration_js_1 = require("../../configuration/common/configuration.js");
const dom_js_1 = require("../../../base/browser/dom.js");
let QuickInputService = class QuickInputService extends themeService_js_1.Themable {
    get backButton() { return this.controller.backButton; }
    get controller() {
        if (!this._controller) {
            this._controller = this._register(this.createController());
        }
        return this._controller;
    }
    get hasController() { return !!this._controller; }
    get currentQuickInput() { return this.controller.currentQuickInput; }
    get quickAccess() {
        if (!this._quickAccess) {
            this._quickAccess = this._register(this.instantiationService.createInstance(quickAccess_js_1.QuickAccessController));
        }
        return this._quickAccess;
    }
    constructor(instantiationService, contextKeyService, themeService, layoutService, configurationService) {
        super(themeService);
        this.instantiationService = instantiationService;
        this.contextKeyService = contextKeyService;
        this.layoutService = layoutService;
        this.configurationService = configurationService;
        this._onShow = this._register(new event_js_1.Emitter());
        this.onShow = this._onShow.event;
        this._onHide = this._register(new event_js_1.Emitter());
        this.onHide = this._onHide.event;
        this.contexts = new Map();
    }
    createController(host = this.layoutService, options) {
        const defaultOptions = {
            idPrefix: 'quickInput_',
            container: host.activeContainer,
            ignoreFocusOut: () => false,
            backKeybindingLabel: () => undefined,
            setContextKey: (id) => this.setContextKey(id),
            linkOpenerDelegate: (content) => {
                // HACK: https://github.com/microsoft/vscode/issues/173691
                this.instantiationService.invokeFunction(accessor => {
                    const openerService = accessor.get(opener_js_1.IOpenerService);
                    openerService.open(content, { allowCommands: true, fromUserGesture: true });
                });
            },
            returnFocus: () => host.focus(),
            styles: this.computeStyles(),
            hoverDelegate: this._register(this.instantiationService.createInstance(quickInput_js_1.QuickInputHoverDelegate))
        };
        const controller = this._register(this.instantiationService.createInstance(quickInputController_js_1.QuickInputController, {
            ...defaultOptions,
            ...options
        }));
        controller.layout(host.activeContainerDimension, host.activeContainerOffset.quickPickTop);
        // Layout changes
        this._register(host.onDidLayoutActiveContainer(dimension => {
            if ((0, dom_js_1.getWindow)(host.activeContainer) === (0, dom_js_1.getWindow)(controller.container)) {
                controller.layout(dimension, host.activeContainerOffset.quickPickTop);
            }
        }));
        this._register(host.onDidChangeActiveContainer(() => {
            if (controller.isVisible()) {
                return;
            }
            controller.layout(host.activeContainerDimension, host.activeContainerOffset.quickPickTop);
        }));
        // Context keys
        this._register(controller.onShow(() => {
            this.resetContextKeys();
            this._onShow.fire();
        }));
        this._register(controller.onHide(() => {
            this.resetContextKeys();
            this._onHide.fire();
        }));
        return controller;
    }
    setContextKey(id) {
        let key;
        if (id) {
            key = this.contexts.get(id);
            if (!key) {
                key = new contextkey_js_1.RawContextKey(id, false)
                    .bindTo(this.contextKeyService);
                this.contexts.set(id, key);
            }
        }
        if (key && key.get()) {
            return; // already active context
        }
        this.resetContextKeys();
        key?.set(true);
    }
    resetContextKeys() {
        this.contexts.forEach(context => {
            if (context.get()) {
                context.reset();
            }
        });
    }
    pick(picks, options, token = cancellation_js_1.CancellationToken.None) {
        return this.controller.pick(picks, options, token);
    }
    input(options = {}, token = cancellation_js_1.CancellationToken.None) {
        return this.controller.input(options, token);
    }
    createQuickPick(options = { useSeparators: false }) {
        return this.controller.createQuickPick(options);
    }
    createInputBox() {
        return this.controller.createInputBox();
    }
    createQuickWidget() {
        return this.controller.createQuickWidget();
    }
    focus() {
        this.controller.focus();
    }
    toggle() {
        this.controller.toggle();
    }
    navigate(next, quickNavigate) {
        this.controller.navigate(next, quickNavigate);
    }
    accept(keyMods) {
        return this.controller.accept(keyMods);
    }
    back() {
        return this.controller.back();
    }
    cancel() {
        return this.controller.cancel();
    }
    updateStyles() {
        if (this.hasController) {
            this.controller.applyStyles(this.computeStyles());
        }
    }
    computeStyles() {
        return {
            widget: {
                quickInputBackground: (0, colorRegistry_js_1.asCssVariable)(colorRegistry_js_1.quickInputBackground),
                quickInputForeground: (0, colorRegistry_js_1.asCssVariable)(colorRegistry_js_1.quickInputForeground),
                quickInputTitleBackground: (0, colorRegistry_js_1.asCssVariable)(colorRegistry_js_1.quickInputTitleBackground),
                widgetBorder: (0, colorRegistry_js_1.asCssVariable)(colorRegistry_js_1.widgetBorder),
                widgetShadow: (0, colorRegistry_js_1.asCssVariable)(colorRegistry_js_1.widgetShadow),
            },
            inputBox: defaultStyles_js_1.defaultInputBoxStyles,
            toggle: defaultStyles_js_1.defaultToggleStyles,
            countBadge: defaultStyles_js_1.defaultCountBadgeStyles,
            button: defaultStyles_js_1.defaultButtonStyles,
            progressBar: defaultStyles_js_1.defaultProgressBarStyles,
            keybindingLabel: defaultStyles_js_1.defaultKeybindingLabelStyles,
            list: (0, defaultStyles_js_1.getListStyles)({
                listBackground: colorRegistry_js_1.quickInputBackground,
                listFocusBackground: colorRegistry_js_1.quickInputListFocusBackground,
                listFocusForeground: colorRegistry_js_1.quickInputListFocusForeground,
                // Look like focused when inactive.
                listInactiveFocusForeground: colorRegistry_js_1.quickInputListFocusForeground,
                listInactiveSelectionIconForeground: colorRegistry_js_1.quickInputListFocusIconForeground,
                listInactiveFocusBackground: colorRegistry_js_1.quickInputListFocusBackground,
                listFocusOutline: colorRegistry_js_1.activeContrastBorder,
                listInactiveFocusOutline: colorRegistry_js_1.activeContrastBorder,
            }),
            pickerGroup: {
                pickerGroupBorder: (0, colorRegistry_js_1.asCssVariable)(colorRegistry_js_1.pickerGroupBorder),
                pickerGroupForeground: (0, colorRegistry_js_1.asCssVariable)(colorRegistry_js_1.pickerGroupForeground),
            }
        };
    }
};
exports.QuickInputService = QuickInputService;
exports.QuickInputService = QuickInputService = __decorate([
    __param(0, instantiation_js_1.IInstantiationService),
    __param(1, contextkey_js_1.IContextKeyService),
    __param(2, themeService_js_1.IThemeService),
    __param(3, layoutService_js_1.ILayoutService),
    __param(4, configuration_js_1.IConfigurationService)
], QuickInputService);
