// @theia/git/lib/node/git-locator/git-locator-client.js
"use strict";
// *****************************************************************************
// Copyright (C) 2017 TypeFox and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0.
//
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License v. 2.0 are satisfied: GNU General Public License, version 2
// with the GNU Classpath Exception which is available at
// https://www.gnu.org/software/classpath/license.html.
//
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
// *****************************************************************************
Object.defineProperty(exports, "__esModule", { value: true });
exports.GitLocatorClient = void 0;
const tslib_1 = require("tslib");
const paths = tslib_1.__importStar(require("path"));
const inversify_1 = require("@theia/core/shared/inversify");
const core_1 = require("@theia/core");
const node_1 = require("@theia/core/lib/node");
let GitLocatorClient = class GitLocatorClient {
    constructor() {
        this.toDispose = new core_1.DisposableCollection();
    }
    dispose() {
        this.toDispose.dispose();
    }
    locate(path, options) {
        return new Promise((resolve, reject) => {
            const toStop = this.ipcConnectionProvider.listen({
                serverName: 'git-locator',
                entryPoint: paths.join(__dirname, 'git-locator-host'),
            }, async (connection) => {
                const proxyFactory = new core_1.RpcProxyFactory();
                const remote = proxyFactory.createProxy();
                proxyFactory.listen(connection);
                try {
                    resolve(await remote.locate(path, options));
                }
                catch (e) {
                    reject(e);
                }
                finally {
                    toStop.dispose();
                }
            });
            this.toDispose.push(toStop);
        });
    }
};
exports.GitLocatorClient = GitLocatorClient;
tslib_1.__decorate([
    (0, inversify_1.inject)(node_1.IPCConnectionProvider),
    tslib_1.__metadata("design:type", node_1.IPCConnectionProvider)
], GitLocatorClient.prototype, "ipcConnectionProvider", void 0);
exports.GitLocatorClient = GitLocatorClient = tslib_1.__decorate([
    (0, inversify_1.injectable)()
], GitLocatorClient);
