// @theia/monaco-editor-core/esm/vs/editor/editor.api.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.languages = exports.editor = exports.Token = exports.Uri = exports.MarkerTag = exports.MarkerSeverity = exports.SelectionDirection = exports.Selection = exports.Range = exports.Position = exports.KeyMod = exports.KeyCode = exports.Emitter = exports.CancellationTokenSource = void 0;
const editorOptions_js_1 = require("./common/config/editorOptions.js");
const editorBaseApi_js_1 = require("./common/services/editorBaseApi.js");
const standaloneEditor_js_1 = require("./standalone/browser/standaloneEditor.js");
const standaloneLanguages_js_1 = require("./standalone/browser/standaloneLanguages.js");
const format_js_1 = require("./contrib/format/browser/format.js");
// Set defaults for standalone editor
editorOptions_js_1.EditorOptions.wrappingIndent.defaultValue = 0 /* WrappingIndent.None */;
editorOptions_js_1.EditorOptions.glyphMargin.defaultValue = false;
editorOptions_js_1.EditorOptions.autoIndent.defaultValue = 3 /* EditorAutoIndentStrategy.Advanced */;
editorOptions_js_1.EditorOptions.overviewRulerLanes.defaultValue = 2;
// We need to register a formatter selector which simply picks the first available formatter.
// See https://github.com/microsoft/monaco-editor/issues/2327
format_js_1.FormattingConflicts.setFormatterSelector((formatter, document, mode) => Promise.resolve(formatter[0]));
const api = (0, editorBaseApi_js_1.createMonacoBaseAPI)();
api.editor = (0, standaloneEditor_js_1.createMonacoEditorAPI)();
api.languages = (0, standaloneLanguages_js_1.createMonacoLanguagesAPI)();
exports.CancellationTokenSource = api.CancellationTokenSource;
exports.Emitter = api.Emitter;
exports.KeyCode = api.KeyCode;
exports.KeyMod = api.KeyMod;
exports.Position = api.Position;
exports.Range = api.Range;
exports.Selection = api.Selection;
exports.SelectionDirection = api.SelectionDirection;
exports.MarkerSeverity = api.MarkerSeverity;
exports.MarkerTag = api.MarkerTag;
exports.Uri = api.Uri;
exports.Token = api.Token;
exports.editor = api.editor;
exports.languages = api.languages;
const monacoEnvironment = globalThis.MonacoEnvironment;
if (monacoEnvironment?.globalAPI || (typeof globalThis.define === 'function' && (globalThis.define).amd)) {
    globalThis.monaco = api;
}
if (typeof globalThis.require !== 'undefined' && typeof globalThis.require.config === 'function') {
    globalThis.require.config({
        ignoreDuplicateModules: [
            'vscode-languageserver-types',
            'vscode-languageserver-types/main',
            'vscode-languageserver-textdocument',
            'vscode-languageserver-textdocument/main',
            'vscode-nls',
            'vscode-nls/vscode-nls',
            'jsonc-parser',
            'jsonc-parser/main',
            'vscode-uri',
            'vscode-uri/index',
            'vs/basic-languages/typescript/typescript'
        ]
    });
}
