// @theia/monaco-editor-core/esm/vs/editor/standalone/browser/quickAccess/standaloneGotoLineQuickAccess.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GotoLineAction = exports.StandaloneGotoLineQuickAccessProvider = void 0;
const gotoLineQuickAccess_js_1 = require("../../../contrib/quickAccess/browser/gotoLineQuickAccess.js");
const platform_js_1 = require("../../../../platform/registry/common/platform.js");
const quickAccess_js_1 = require("../../../../platform/quickinput/common/quickAccess.js");
const codeEditorService_js_1 = require("../../../browser/services/codeEditorService.js");
const standaloneStrings_js_1 = require("../../../common/standaloneStrings.js");
const event_js_1 = require("../../../../base/common/event.js");
const editorExtensions_js_1 = require("../../../browser/editorExtensions.js");
const editorContextKeys_js_1 = require("../../../common/editorContextKeys.js");
const quickInput_js_1 = require("../../../../platform/quickinput/common/quickInput.js");
let StandaloneGotoLineQuickAccessProvider = class StandaloneGotoLineQuickAccessProvider extends gotoLineQuickAccess_js_1.AbstractGotoLineQuickAccessProvider {
    constructor(editorService) {
        super();
        this.editorService = editorService;
        this.onDidActiveTextEditorControlChange = event_js_1.Event.None;
    }
    get activeTextEditorControl() {
        return this.editorService.getFocusedCodeEditor() ?? undefined;
    }
};
exports.StandaloneGotoLineQuickAccessProvider = StandaloneGotoLineQuickAccessProvider;
exports.StandaloneGotoLineQuickAccessProvider = StandaloneGotoLineQuickAccessProvider = __decorate([
    __param(0, codeEditorService_js_1.ICodeEditorService)
], StandaloneGotoLineQuickAccessProvider);
class GotoLineAction extends editorExtensions_js_1.EditorAction {
    static { this.ID = 'editor.action.gotoLine'; }
    constructor() {
        super({
            id: GotoLineAction.ID,
            label: standaloneStrings_js_1.GoToLineNLS.gotoLineActionLabel,
            alias: 'Go to Line/Column...',
            precondition: undefined,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.focus,
                primary: 2048 /* KeyMod.CtrlCmd */ | 37 /* KeyCode.KeyG */,
                mac: { primary: 256 /* KeyMod.WinCtrl */ | 37 /* KeyCode.KeyG */ },
                weight: 100 /* KeybindingWeight.EditorContrib */
            }
        });
    }
    run(accessor) {
        accessor.get(quickInput_js_1.IQuickInputService).quickAccess.show(StandaloneGotoLineQuickAccessProvider.PREFIX);
    }
}
exports.GotoLineAction = GotoLineAction;
(0, editorExtensions_js_1.registerEditorAction)(GotoLineAction);
platform_js_1.Registry.as(quickAccess_js_1.Extensions.Quickaccess).registerQuickAccessProvider({
    ctor: StandaloneGotoLineQuickAccessProvider,
    prefix: StandaloneGotoLineQuickAccessProvider.PREFIX,
    helpEntries: [{ description: standaloneStrings_js_1.GoToLineNLS.gotoLineActionLabel, commandId: GotoLineAction.ID }]
});
