// @theia/monaco-editor-core/esm/vs/platform/workspace/common/workspaceTrust.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.IWorkspaceTrustRequestService = exports.IWorkspaceTrustManagementService = exports.IWorkspaceTrustEnablementService = exports.WorkspaceTrustScope = void 0;
const instantiation_js_1 = require("../../instantiation/common/instantiation.js");
var WorkspaceTrustScope;
(function (WorkspaceTrustScope) {
    WorkspaceTrustScope[WorkspaceTrustScope["Local"] = 0] = "Local";
    WorkspaceTrustScope[WorkspaceTrustScope["Remote"] = 1] = "Remote";
})(WorkspaceTrustScope || (exports.WorkspaceTrustScope = WorkspaceTrustScope = {}));
exports.IWorkspaceTrustEnablementService = (0, instantiation_js_1.createDecorator)('workspaceTrustEnablementService');
exports.IWorkspaceTrustManagementService = (0, instantiation_js_1.createDecorator)('workspaceTrustManagementService');
exports.IWorkspaceTrustRequestService = (0, instantiation_js_1.createDecorator)('workspaceTrustRequestService');
