// @theia/monaco-editor-core/esm/vs/base/browser/ui/hover/hoverDelegate2.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.setBaseLayerHoverDelegate = setBaseLayerHoverDelegate;
exports.getBaseLayerHoverDelegate = getBaseLayerHoverDelegate;
const lifecycle_js_1 = require("../../../common/lifecycle.js");
let baseHoverDelegate = {
    showHover: () => undefined,
    showDelayedHover: () => undefined,
    setupDelayedHover: () => lifecycle_js_1.Disposable.None,
    setupDelayedHoverAtMouse: () => lifecycle_js_1.Disposable.None,
    hideHover: () => undefined,
    showAndFocusLastHover: () => undefined,
    setupManagedHover: () => null,
    showManagedHover: () => undefined
};
/**
 * Sets the hover delegate for use **only in the `base/` layer**.
 */
function setBaseLayerHoverDelegate(hoverDelegate) {
    baseHoverDelegate = hoverDelegate;
}
/**
 * Gets the hover delegate for use **only in the `base/` layer**.
 *
 * Since the hover service depends on various platform services, this delegate essentially bypasses
 * the standard dependency injection mechanism by injecting a global hover service at start up. The
 * only reason this should be used is if `IHoverService` is not available.
 */
function getBaseLayerHoverDelegate() {
    return baseHoverDelegate;
}
