// @theia/monaco-editor-core/esm/vs/editor/contrib/dropOrPasteInto/browser/defaultProviders.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultPasteProvidersFeature = exports.DefaultDropProvidersFeature = exports.DefaultTextPasteOrDropEditProvider = void 0;
const arrays_js_1 = require("../../../../base/common/arrays.js");
const dataTransfer_js_1 = require("../../../../base/common/dataTransfer.js");
const hierarchicalKind_js_1 = require("../../../../base/common/hierarchicalKind.js");
const lifecycle_js_1 = require("../../../../base/common/lifecycle.js");
const mime_js_1 = require("../../../../base/common/mime.js");
const network_js_1 = require("../../../../base/common/network.js");
const resources_js_1 = require("../../../../base/common/resources.js");
const uri_js_1 = require("../../../../base/common/uri.js");
const nls_js_1 = require("../../../../nls.js");
const workspace_js_1 = require("../../../../platform/workspace/common/workspace.js");
const languages_js_1 = require("../../../common/languages.js");
const languageFeatures_js_1 = require("../../../common/services/languageFeatures.js");
class SimplePasteAndDropProvider {
    constructor(kind) {
        this.copyMimeTypes = [];
        this.kind = kind;
        this.providedDropEditKinds = [this.kind];
        this.providedPasteEditKinds = [this.kind];
    }
    async provideDocumentPasteEdits(_model, _ranges, dataTransfer, context, token) {
        const edit = await this.getEdit(dataTransfer, token);
        if (!edit) {
            return undefined;
        }
        return {
            edits: [{ insertText: edit.insertText, title: edit.title, kind: edit.kind, handledMimeType: edit.handledMimeType, yieldTo: edit.yieldTo }],
            dispose() { },
        };
    }
    async provideDocumentDropEdits(_model, _position, dataTransfer, token) {
        const edit = await this.getEdit(dataTransfer, token);
        if (!edit) {
            return;
        }
        return {
            edits: [{ insertText: edit.insertText, title: edit.title, kind: edit.kind, handledMimeType: edit.handledMimeType, yieldTo: edit.yieldTo }],
            dispose() { },
        };
    }
}
class DefaultTextPasteOrDropEditProvider extends SimplePasteAndDropProvider {
    static { this.id = 'text'; }
    constructor() {
        super(hierarchicalKind_js_1.HierarchicalKind.Empty.append('text', 'plain'));
        this.id = DefaultTextPasteOrDropEditProvider.id;
        this.dropMimeTypes = [mime_js_1.Mimes.text];
        this.pasteMimeTypes = [mime_js_1.Mimes.text];
    }
    async getEdit(dataTransfer, _token) {
        const textEntry = dataTransfer.get(mime_js_1.Mimes.text);
        if (!textEntry) {
            return;
        }
        // Suppress if there's also a uriList entry.
        // Typically the uri-list contains the same text as the text entry so showing both is confusing.
        if (dataTransfer.has(mime_js_1.Mimes.uriList)) {
            return;
        }
        const insertText = await textEntry.asString();
        return {
            handledMimeType: mime_js_1.Mimes.text,
            title: (0, nls_js_1.localize)('text.label', "Insert Plain Text"),
            insertText,
            kind: this.kind,
        };
    }
}
exports.DefaultTextPasteOrDropEditProvider = DefaultTextPasteOrDropEditProvider;
class PathProvider extends SimplePasteAndDropProvider {
    constructor() {
        super(hierarchicalKind_js_1.HierarchicalKind.Empty.append('uri', 'path', 'absolute'));
        this.dropMimeTypes = [mime_js_1.Mimes.uriList];
        this.pasteMimeTypes = [mime_js_1.Mimes.uriList];
    }
    async getEdit(dataTransfer, token) {
        const entries = await extractUriList(dataTransfer);
        if (!entries.length || token.isCancellationRequested) {
            return;
        }
        let uriCount = 0;
        const insertText = entries
            .map(({ uri, originalText }) => {
            if (uri.scheme === network_js_1.Schemas.file) {
                return uri.fsPath;
            }
            else {
                uriCount++;
                return originalText;
            }
        })
            .join(' ');
        let label;
        if (uriCount > 0) {
            // Dropping at least one generic uri (such as https) so use most generic label
            label = entries.length > 1
                ? (0, nls_js_1.localize)('defaultDropProvider.uriList.uris', "Insert Uris")
                : (0, nls_js_1.localize)('defaultDropProvider.uriList.uri', "Insert Uri");
        }
        else {
            // All the paths are file paths
            label = entries.length > 1
                ? (0, nls_js_1.localize)('defaultDropProvider.uriList.paths', "Insert Paths")
                : (0, nls_js_1.localize)('defaultDropProvider.uriList.path', "Insert Path");
        }
        return {
            handledMimeType: mime_js_1.Mimes.uriList,
            insertText,
            title: label,
            kind: this.kind,
        };
    }
}
let RelativePathProvider = class RelativePathProvider extends SimplePasteAndDropProvider {
    constructor(_workspaceContextService) {
        super(hierarchicalKind_js_1.HierarchicalKind.Empty.append('uri', 'path', 'relative'));
        this._workspaceContextService = _workspaceContextService;
        this.dropMimeTypes = [mime_js_1.Mimes.uriList];
        this.pasteMimeTypes = [mime_js_1.Mimes.uriList];
    }
    async getEdit(dataTransfer, token) {
        const entries = await extractUriList(dataTransfer);
        if (!entries.length || token.isCancellationRequested) {
            return;
        }
        const relativeUris = (0, arrays_js_1.coalesce)(entries.map(({ uri }) => {
            const root = this._workspaceContextService.getWorkspaceFolder(uri);
            return root ? (0, resources_js_1.relativePath)(root.uri, uri) : undefined;
        }));
        if (!relativeUris.length) {
            return;
        }
        return {
            handledMimeType: mime_js_1.Mimes.uriList,
            insertText: relativeUris.join(' '),
            title: entries.length > 1
                ? (0, nls_js_1.localize)('defaultDropProvider.uriList.relativePaths', "Insert Relative Paths")
                : (0, nls_js_1.localize)('defaultDropProvider.uriList.relativePath', "Insert Relative Path"),
            kind: this.kind,
        };
    }
};
RelativePathProvider = __decorate([
    __param(0, workspace_js_1.IWorkspaceContextService)
], RelativePathProvider);
class PasteHtmlProvider {
    constructor() {
        this.kind = new hierarchicalKind_js_1.HierarchicalKind('html');
        this.providedPasteEditKinds = [this.kind];
        this.copyMimeTypes = [];
        this.pasteMimeTypes = ['text/html'];
        this._yieldTo = [{ mimeType: mime_js_1.Mimes.text }];
    }
    async provideDocumentPasteEdits(_model, _ranges, dataTransfer, context, token) {
        if (context.triggerKind !== languages_js_1.DocumentPasteTriggerKind.PasteAs && !context.only?.contains(this.kind)) {
            return;
        }
        const entry = dataTransfer.get('text/html');
        const htmlText = await entry?.asString();
        if (!htmlText || token.isCancellationRequested) {
            return;
        }
        return {
            dispose() { },
            edits: [{
                    insertText: htmlText,
                    yieldTo: this._yieldTo,
                    title: (0, nls_js_1.localize)('pasteHtmlLabel', 'Insert HTML'),
                    kind: this.kind,
                }],
        };
    }
}
async function extractUriList(dataTransfer) {
    const urlListEntry = dataTransfer.get(mime_js_1.Mimes.uriList);
    if (!urlListEntry) {
        return [];
    }
    const strUriList = await urlListEntry.asString();
    const entries = [];
    for (const entry of dataTransfer_js_1.UriList.parse(strUriList)) {
        try {
            entries.push({ uri: uri_js_1.URI.parse(entry), originalText: entry });
        }
        catch {
            // noop
        }
    }
    return entries;
}
const genericLanguageSelector = { scheme: '*', hasAccessToAllModels: true };
let DefaultDropProvidersFeature = class DefaultDropProvidersFeature extends lifecycle_js_1.Disposable {
    constructor(languageFeaturesService, workspaceContextService) {
        super();
        this._register(languageFeaturesService.documentDropEditProvider.register(genericLanguageSelector, new DefaultTextPasteOrDropEditProvider()));
        this._register(languageFeaturesService.documentDropEditProvider.register(genericLanguageSelector, new PathProvider()));
        this._register(languageFeaturesService.documentDropEditProvider.register(genericLanguageSelector, new RelativePathProvider(workspaceContextService)));
    }
};
exports.DefaultDropProvidersFeature = DefaultDropProvidersFeature;
exports.DefaultDropProvidersFeature = DefaultDropProvidersFeature = __decorate([
    __param(0, languageFeatures_js_1.ILanguageFeaturesService),
    __param(1, workspace_js_1.IWorkspaceContextService)
], DefaultDropProvidersFeature);
let DefaultPasteProvidersFeature = class DefaultPasteProvidersFeature extends lifecycle_js_1.Disposable {
    constructor(languageFeaturesService, workspaceContextService) {
        super();
        this._register(languageFeaturesService.documentPasteEditProvider.register(genericLanguageSelector, new DefaultTextPasteOrDropEditProvider()));
        this._register(languageFeaturesService.documentPasteEditProvider.register(genericLanguageSelector, new PathProvider()));
        this._register(languageFeaturesService.documentPasteEditProvider.register(genericLanguageSelector, new RelativePathProvider(workspaceContextService)));
        this._register(languageFeaturesService.documentPasteEditProvider.register(genericLanguageSelector, new PasteHtmlProvider()));
    }
};
exports.DefaultPasteProvidersFeature = DefaultPasteProvidersFeature;
exports.DefaultPasteProvidersFeature = DefaultPasteProvidersFeature = __decorate([
    __param(0, languageFeatures_js_1.ILanguageFeaturesService),
    __param(1, workspace_js_1.IWorkspaceContextService)
], DefaultPasteProvidersFeature);
