// @theia/monaco-editor-core/esm/vs/editor/contrib/dropOrPasteInto/browser/dropIntoEditorContribution.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const editorExtensions_js_1 = require("../../../browser/editorExtensions.js");
const editorFeatures_js_1 = require("../../../common/editorFeatures.js");
const defaultProviders_js_1 = require("./defaultProviders.js");
const dropIntoEditorController_js_1 = require("./dropIntoEditorController.js");
(0, editorExtensions_js_1.registerEditorContribution)(dropIntoEditorController_js_1.DropIntoEditorController.ID, dropIntoEditorController_js_1.DropIntoEditorController, 2 /* EditorContributionInstantiation.BeforeFirstInteraction */);
(0, editorFeatures_js_1.registerEditorFeature)(defaultProviders_js_1.DefaultDropProvidersFeature);
(0, editorExtensions_js_1.registerEditorCommand)(new class extends editorExtensions_js_1.EditorCommand {
    constructor() {
        super({
            id: dropIntoEditorController_js_1.changeDropTypeCommandId,
            precondition: dropIntoEditorController_js_1.dropWidgetVisibleCtx,
            kbOpts: {
                weight: 100 /* KeybindingWeight.EditorContrib */,
                primary: 2048 /* KeyMod.CtrlCmd */ | 89 /* KeyCode.Period */,
            }
        });
    }
    runEditorCommand(_accessor, editor, _args) {
        dropIntoEditorController_js_1.DropIntoEditorController.get(editor)?.changeDropType();
    }
});
(0, editorExtensions_js_1.registerEditorCommand)(new class extends editorExtensions_js_1.EditorCommand {
    constructor() {
        super({
            id: 'editor.hideDropWidget',
            precondition: dropIntoEditorController_js_1.dropWidgetVisibleCtx,
            kbOpts: {
                weight: 100 /* KeybindingWeight.EditorContrib */,
                primary: 9 /* KeyCode.Escape */,
            }
        });
    }
    runEditorCommand(_accessor, editor, _args) {
        dropIntoEditorController_js_1.DropIntoEditorController.get(editor)?.clearWidgets();
    }
});
