// @theia/monaco-editor-core/esm/vs/editor/contrib/dropOrPasteInto/browser/dropIntoEditorController.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var DropIntoEditorController_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DropIntoEditorController = exports.dropWidgetVisibleCtx = exports.changeDropTypeCommandId = exports.dropAsPreferenceConfig = void 0;
const arrays_js_1 = require("../../../../base/common/arrays.js");
const async_js_1 = require("../../../../base/common/async.js");
const dataTransfer_js_1 = require("../../../../base/common/dataTransfer.js");
const errors_js_1 = require("../../../../base/common/errors.js");
const hierarchicalKind_js_1 = require("../../../../base/common/hierarchicalKind.js");
const lifecycle_js_1 = require("../../../../base/common/lifecycle.js");
const nls_js_1 = require("../../../../nls.js");
const configuration_js_1 = require("../../../../platform/configuration/common/configuration.js");
const contextkey_js_1 = require("../../../../platform/contextkey/common/contextkey.js");
const dnd_js_1 = require("../../../../platform/dnd/browser/dnd.js");
const instantiation_js_1 = require("../../../../platform/instantiation/common/instantiation.js");
const dnd_js_2 = require("../../../browser/dnd.js");
const range_js_1 = require("../../../common/core/range.js");
const languageFeatures_js_1 = require("../../../common/services/languageFeatures.js");
const treeViewsDnd_js_1 = require("../../../common/services/treeViewsDnd.js");
const treeViewsDndService_js_1 = require("../../../common/services/treeViewsDndService.js");
const editorState_js_1 = require("../../editorState/browser/editorState.js");
const inlineProgress_js_1 = require("../../inlineProgress/browser/inlineProgress.js");
const edit_js_1 = require("./edit.js");
const postEditWidget_js_1 = require("./postEditWidget.js");
exports.dropAsPreferenceConfig = 'editor.dropIntoEditor.preferences';
exports.changeDropTypeCommandId = 'editor.changeDropType';
exports.dropWidgetVisibleCtx = new contextkey_js_1.RawContextKey('dropWidgetVisible', false, (0, nls_js_1.localize)('dropWidgetVisible', "Whether the drop widget is showing"));
let DropIntoEditorController = class DropIntoEditorController extends lifecycle_js_1.Disposable {
    static { DropIntoEditorController_1 = this; }
    static { this.ID = 'editor.contrib.dropIntoEditorController'; }
    static get(editor) {
        return editor.getContribution(DropIntoEditorController_1.ID);
    }
    static setConfigureDefaultAction(action) {
        this._configureDefaultAction = action;
    }
    constructor(editor, instantiationService, _configService, _languageFeaturesService, _treeViewsDragAndDropService) {
        super();
        this._configService = _configService;
        this._languageFeaturesService = _languageFeaturesService;
        this._treeViewsDragAndDropService = _treeViewsDragAndDropService;
        this.treeItemsTransfer = dnd_js_1.LocalSelectionTransfer.getInstance();
        this._dropProgressManager = this._register(instantiationService.createInstance(inlineProgress_js_1.InlineProgressManager, 'dropIntoEditor', editor));
        this._postDropWidgetManager = this._register(instantiationService.createInstance(postEditWidget_js_1.PostEditWidgetManager, 'dropIntoEditor', editor, exports.dropWidgetVisibleCtx, { id: exports.changeDropTypeCommandId, label: (0, nls_js_1.localize)('postDropWidgetTitle', "Show drop options...") }, () => DropIntoEditorController_1._configureDefaultAction ? [DropIntoEditorController_1._configureDefaultAction] : []));
        this._register(editor.onDropIntoEditor(e => this.onDropIntoEditor(editor, e.position, e.event)));
    }
    clearWidgets() {
        this._postDropWidgetManager.clear();
    }
    changeDropType() {
        this._postDropWidgetManager.tryShowSelector();
    }
    async onDropIntoEditor(editor, position, dragEvent) {
        if (!dragEvent.dataTransfer || !editor.hasModel()) {
            return;
        }
        DropIntoEditorController_1._currentDropOperation?.cancel();
        editor.focus();
        editor.setPosition(position);
        const p = (0, async_js_1.createCancelablePromise)(async (token) => {
            const disposables = new lifecycle_js_1.DisposableStore();
            const tokenSource = disposables.add(new editorState_js_1.EditorStateCancellationTokenSource(editor, 1 /* CodeEditorStateFlag.Value */, undefined, token));
            try {
                const ourDataTransfer = await this.extractDataTransferData(dragEvent);
                if (ourDataTransfer.size === 0 || tokenSource.token.isCancellationRequested) {
                    return;
                }
                const model = editor.getModel();
                if (!model) {
                    return;
                }
                const providers = this._languageFeaturesService.documentDropEditProvider
                    .ordered(model)
                    .filter(provider => {
                    if (!provider.dropMimeTypes) {
                        // Keep all providers that don't specify mime types
                        return true;
                    }
                    return provider.dropMimeTypes.some(mime => ourDataTransfer.matches(mime));
                });
                const editSession = disposables.add(await this.getDropEdits(providers, model, position, ourDataTransfer, tokenSource.token));
                if (tokenSource.token.isCancellationRequested) {
                    return;
                }
                if (editSession.edits.length) {
                    const activeEditIndex = this.getInitialActiveEditIndex(model, editSession.edits);
                    const canShowWidget = editor.getOption(36 /* EditorOption.dropIntoEditor */).showDropSelector === 'afterDrop';
                    // Pass in the parent token here as it tracks cancelling the entire drop operation
                    await this._postDropWidgetManager.applyEditAndShowIfNeeded([range_js_1.Range.fromPositions(position)], { activeEditIndex, allEdits: editSession.edits }, canShowWidget, async (edit) => edit, token);
                }
            }
            finally {
                disposables.dispose();
                if (DropIntoEditorController_1._currentDropOperation === p) {
                    DropIntoEditorController_1._currentDropOperation = undefined;
                }
            }
        });
        this._dropProgressManager.showWhile(position, (0, nls_js_1.localize)('dropIntoEditorProgress', "Running drop handlers. Click to cancel"), p, { cancel: () => p.cancel() });
        DropIntoEditorController_1._currentDropOperation = p;
    }
    async getDropEdits(providers, model, position, dataTransfer, token) {
        const disposables = new lifecycle_js_1.DisposableStore();
        const results = await (0, async_js_1.raceCancellation)(Promise.all(providers.map(async (provider) => {
            try {
                const edits = await provider.provideDocumentDropEdits(model, position, dataTransfer, token);
                if (edits) {
                    disposables.add(edits);
                }
                return edits?.edits.map(edit => ({ ...edit, providerId: provider.id }));
            }
            catch (err) {
                if (!(0, errors_js_1.isCancellationError)(err)) {
                    console.error(err);
                }
                console.error(err);
            }
            return undefined;
        })), token);
        const edits = (0, arrays_js_1.coalesce)(results ?? []).flat();
        return {
            edits: (0, edit_js_1.sortEditsByYieldTo)(edits),
            dispose: () => disposables.dispose()
        };
    }
    getInitialActiveEditIndex(model, edits) {
        const preferredProviders = this._configService.getValue(exports.dropAsPreferenceConfig, { resource: model.uri });
        for (const config of Array.isArray(preferredProviders) ? preferredProviders : []) {
            const desiredKind = new hierarchicalKind_js_1.HierarchicalKind(config);
            const editIndex = edits.findIndex(edit => edit.kind && desiredKind.contains(edit.kind));
            if (editIndex >= 0) {
                return editIndex;
            }
        }
        return 0;
    }
    async extractDataTransferData(dragEvent) {
        if (!dragEvent.dataTransfer) {
            return new dataTransfer_js_1.VSDataTransfer();
        }
        const dataTransfer = (0, dnd_js_2.toExternalVSDataTransfer)(dragEvent.dataTransfer);
        if (this.treeItemsTransfer.hasData(treeViewsDnd_js_1.DraggedTreeItemsIdentifier.prototype)) {
            const data = this.treeItemsTransfer.getData(treeViewsDnd_js_1.DraggedTreeItemsIdentifier.prototype);
            if (Array.isArray(data)) {
                for (const id of data) {
                    const treeDataTransfer = await this._treeViewsDragAndDropService.removeDragOperationTransfer(id.identifier);
                    if (treeDataTransfer) {
                        for (const [type, value] of treeDataTransfer) {
                            dataTransfer.replace(type, value);
                        }
                    }
                }
            }
        }
        return dataTransfer;
    }
};
exports.DropIntoEditorController = DropIntoEditorController;
exports.DropIntoEditorController = DropIntoEditorController = DropIntoEditorController_1 = __decorate([
    __param(1, instantiation_js_1.IInstantiationService),
    __param(2, configuration_js_1.IConfigurationService),
    __param(3, languageFeatures_js_1.ILanguageFeaturesService),
    __param(4, treeViewsDndService_js_1.ITreeViewsDnDService)
], DropIntoEditorController);
