// @theia/monaco-editor-core/esm/vs/editor/browser/controller/editContext/native/nativeEditContextUtils.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.FocusTracker = void 0;
exports.editContextAddDisposableListener = editContextAddDisposableListener;
const dom_js_1 = require("../../../../../base/browser/dom.js");
const lifecycle_js_1 = require("../../../../../base/common/lifecycle.js");
class FocusTracker extends lifecycle_js_1.Disposable {
    constructor(_domNode, _onFocusChange) {
        super();
        this._domNode = _domNode;
        this._onFocusChange = _onFocusChange;
        this._isFocused = false;
        this._register((0, dom_js_1.addDisposableListener)(this._domNode, 'focus', () => this._handleFocusedChanged(true)));
        this._register((0, dom_js_1.addDisposableListener)(this._domNode, 'blur', () => this._handleFocusedChanged(false)));
    }
    _handleFocusedChanged(focused) {
        if (this._isFocused === focused) {
            return;
        }
        this._isFocused = focused;
        this._onFocusChange(this._isFocused);
    }
    focus() {
        // fixes: https://github.com/microsoft/vscode/issues/228147
        // Immediately call this method in order to directly set the field isFocused to true so the textInputFocus context key is evaluated correctly
        this._handleFocusedChanged(true);
        this._domNode.focus();
    }
    refreshFocusState() {
        const focused = this._domNode === (0, dom_js_1.getActiveWindow)().document.activeElement;
        this._handleFocusedChanged(focused);
    }
    get isFocused() {
        return this._isFocused;
    }
}
exports.FocusTracker = FocusTracker;
function editContextAddDisposableListener(target, type, listener, options) {
    target.addEventListener(type, listener, options);
    return {
        dispose() {
            target.removeEventListener(type, listener);
        }
    };
}
