// @theia/monaco-editor-core/esm/vs/platform/theme/common/themeService.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.Themable = exports.Extensions = exports.FolderThemeIcon = exports.FileThemeIcon = exports.IThemeService = void 0;
exports.themeColorFromId = themeColorFromId;
exports.getThemeTypeSelector = getThemeTypeSelector;
exports.registerThemingParticipant = registerThemingParticipant;
const codicons_js_1 = require("../../../base/common/codicons.js");
const event_js_1 = require("../../../base/common/event.js");
const lifecycle_js_1 = require("../../../base/common/lifecycle.js");
const instantiation_js_1 = require("../../instantiation/common/instantiation.js");
const platform = require("../../registry/common/platform.js");
const theme_js_1 = require("./theme.js");
exports.IThemeService = (0, instantiation_js_1.createDecorator)('themeService');
function themeColorFromId(id) {
    return { id };
}
exports.FileThemeIcon = codicons_js_1.Codicon.file;
exports.FolderThemeIcon = codicons_js_1.Codicon.folder;
function getThemeTypeSelector(type) {
    switch (type) {
        case theme_js_1.ColorScheme.DARK: return 'vs-dark';
        case theme_js_1.ColorScheme.HIGH_CONTRAST_DARK: return 'hc-black';
        case theme_js_1.ColorScheme.HIGH_CONTRAST_LIGHT: return 'hc-light';
        default: return 'vs';
    }
}
// static theming participant
exports.Extensions = {
    ThemingContribution: 'base.contributions.theming'
};
class ThemingRegistry {
    constructor() {
        this.themingParticipants = [];
        this.themingParticipants = [];
        this.onThemingParticipantAddedEmitter = new event_js_1.Emitter();
    }
    onColorThemeChange(participant) {
        this.themingParticipants.push(participant);
        this.onThemingParticipantAddedEmitter.fire(participant);
        return (0, lifecycle_js_1.toDisposable)(() => {
            const idx = this.themingParticipants.indexOf(participant);
            this.themingParticipants.splice(idx, 1);
        });
    }
    get onThemingParticipantAdded() {
        return this.onThemingParticipantAddedEmitter.event;
    }
    getThemingParticipants() {
        return this.themingParticipants;
    }
}
const themingRegistry = new ThemingRegistry();
platform.Registry.add(exports.Extensions.ThemingContribution, themingRegistry);
function registerThemingParticipant(participant) {
    return themingRegistry.onColorThemeChange(participant);
}
/**
 * Utility base class for all themable components.
 */
class Themable extends lifecycle_js_1.Disposable {
    constructor(themeService) {
        super();
        this.themeService = themeService;
        this.theme = themeService.getColorTheme();
        // Hook up to theme changes
        this._register(this.themeService.onDidColorThemeChange(theme => this.onThemeChange(theme)));
    }
    onThemeChange(theme) {
        this.theme = theme;
        this.updateStyles();
    }
    updateStyles() {
        // Subclasses to override
    }
    getColor(id, modify) {
        let color = this.theme.getColor(id);
        if (color && modify) {
            color = modify(color, this.theme);
        }
        return color ? color.toString() : null;
    }
}
exports.Themable = Themable;
