// @theia/preferences/lib/browser/util/preference-layout.js
"use strict";
// *****************************************************************************
// Copyright (C) 2024 TypeFox and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0.
//
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License v. 2.0 are satisfied: GNU General Public License, version 2
// with the GNU Classpath Exception which is available at
// https://www.gnu.org/software/classpath/license.html.
//
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
// *****************************************************************************
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreferenceLayoutProvider = exports.DEFAULT_LAYOUT = exports.COMMONLY_USED_LAYOUT = exports.COMMONLY_USED_SECTION_PREFIX = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@theia/core");
const inversify_1 = require("@theia/core/shared/inversify");
exports.COMMONLY_USED_SECTION_PREFIX = 'commonly-used';
exports.COMMONLY_USED_LAYOUT = {
    id: exports.COMMONLY_USED_SECTION_PREFIX,
    label: core_1.nls.localizeByDefault('Commonly Used'),
    settings: [
        'files.autoSave',
        'editor.fontSize',
        'editor.fontFamily',
        'editor.tabSize',
        'editor.renderWhitespace',
        'editor.cursorStyle',
        'editor.multiCursorModifier',
        'editor.insertSpaces',
        'editor.wordWrap',
        'files.exclude',
        'files.associations'
    ]
};
exports.DEFAULT_LAYOUT = [
    {
        id: 'editor',
        label: core_1.nls.localizeByDefault('Text Editor'),
        settings: ['editor.*'],
        children: [
            {
                id: 'editor.cursor',
                label: core_1.nls.localizeByDefault('Cursor'),
                settings: ['editor.cursor*']
            },
            {
                id: 'editor.find',
                label: core_1.nls.localizeByDefault('Find'),
                settings: ['editor.find.*']
            },
            {
                id: 'editor.font',
                label: core_1.nls.localizeByDefault('Font'),
                settings: ['editor.font*']
            },
            {
                id: 'editor.format',
                label: core_1.nls.localizeByDefault('Formatting'),
                settings: ['editor.format*']
            },
            {
                id: 'editor.diffEditor',
                label: core_1.nls.localizeByDefault('Diff Editor'),
                settings: ['diffEditor.*']
            },
            {
                id: 'editor.multiDiffEditor',
                label: core_1.nls.localizeByDefault('Multi-File Diff Editor'),
                settings: ['multiDiffEditor.*']
            },
            {
                id: 'editor.minimap',
                label: core_1.nls.localizeByDefault('Minimap'),
                settings: ['editor.minimap.*']
            },
            {
                id: 'editor.suggestions',
                label: core_1.nls.localizeByDefault('Suggestions'),
                settings: ['editor.*suggest*']
            },
            {
                id: 'editor.files',
                label: core_1.nls.localizeByDefault('Files'),
                settings: ['files.*']
            }
        ]
    },
    {
        id: 'workbench',
        label: core_1.nls.localizeByDefault('Workbench'),
        settings: ['workbench.*', 'workspace.*'],
        children: [
            {
                id: 'workbench.appearance',
                label: core_1.nls.localizeByDefault('Appearance'),
                settings: [
                    'workbench.activityBar.*', 'workbench.*color*', 'workbench.fontAliasing', 'workbench.iconTheme', 'workbench.sidebar.location',
                    'workbench.*.visible', 'workbench.tips.enabled', 'workbench.tree.*', 'workbench.view.*'
                ]
            },
            {
                id: 'workbench.breadcrumbs',
                label: core_1.nls.localizeByDefault('Breadcrumbs'),
                settings: ['breadcrumbs.*']
            },
            {
                id: 'workbench.editor',
                label: core_1.nls.localizeByDefault('Editor Management'),
                settings: ['workbench.editor.*']
            },
            {
                id: 'workbench.settings',
                label: core_1.nls.localizeByDefault('Settings Editor'),
                settings: ['workbench.settings.*']
            },
            {
                id: 'workbench.zenmode',
                label: core_1.nls.localizeByDefault('Zen Mode'),
                settings: ['zenmode.*']
            },
            {
                id: 'workbench.screencastmode',
                label: core_1.nls.localizeByDefault('Screencast Mode'),
                settings: ['screencastMode.*']
            }
        ]
    },
    {
        id: 'window',
        label: core_1.nls.localizeByDefault('Window'),
        settings: ['window.*'],
        children: [
            {
                id: 'window.newWindow',
                label: core_1.nls.localizeByDefault('New Window'),
                settings: ['window.*newwindow*']
            }
        ]
    },
    {
        id: 'features',
        label: core_1.nls.localizeByDefault('Features'),
        children: [
            {
                id: 'features.accessibilitySignals',
                label: core_1.nls.localizeByDefault('Accessibility Signals'),
                settings: ['accessibility.signal*']
            },
            {
                id: 'features.accessibility',
                label: core_1.nls.localizeByDefault('Accessibility'),
                settings: ['accessibility.*']
            },
            {
                id: 'features.explorer',
                label: core_1.nls.localizeByDefault('Explorer'),
                settings: ['explorer.*', 'outline.*']
            },
            {
                id: 'features.search',
                label: core_1.nls.localizeByDefault('Search'),
                settings: ['search.*']
            },
            {
                id: 'features.debug',
                label: core_1.nls.localizeByDefault('Debug'),
                settings: ['debug.*', 'launch']
            },
            {
                id: 'features.testing',
                label: core_1.nls.localizeByDefault('Testing'),
                settings: ['testing.*']
            },
            {
                id: 'features.scm',
                label: core_1.nls.localizeByDefault('Source Control'),
                settings: ['scm.*']
            },
            {
                id: 'features.extensions',
                label: core_1.nls.localizeByDefault('Extensions'),
                settings: ['extensions.*']
            },
            {
                id: 'features.terminal',
                label: core_1.nls.localizeByDefault('Terminal'),
                settings: ['terminal.*']
            },
            {
                id: 'features.task',
                label: core_1.nls.localizeByDefault('Task'),
                settings: ['task.*']
            },
            {
                id: 'features.problems',
                label: core_1.nls.localizeByDefault('Problems'),
                settings: ['problems.*']
            },
            {
                id: 'features.output',
                label: core_1.nls.localizeByDefault('Output'),
                settings: ['output.*']
            },
            {
                id: 'features.comments',
                label: core_1.nls.localizeByDefault('Comments'),
                settings: ['comments.*']
            },
            {
                id: 'features.remote',
                label: core_1.nls.localizeByDefault('Remote'),
                settings: ['remote.*']
            },
            {
                id: 'features.timeline',
                label: core_1.nls.localizeByDefault('Timeline'),
                settings: ['timeline.*']
            },
            {
                id: 'features.toolbar',
                label: core_1.nls.localize('theia/preferences/toolbar', 'Toolbar'),
                settings: ['toolbar.*']
            },
            {
                id: 'features.notebook',
                label: core_1.nls.localizeByDefault('Notebook'),
                settings: ['notebook.*', 'interactiveWindow.*']
            },
            {
                id: 'features.mergeEditor',
                label: core_1.nls.localizeByDefault('Merge Editor'),
                settings: ['mergeEditor.*']
            },
            {
                id: 'features.chat',
                label: core_1.nls.localizeByDefault('Chat'),
                settings: ['chat.*', 'inlineChat.*']
            }
        ]
    },
    {
        id: 'application',
        label: core_1.nls.localizeByDefault('Application'),
        children: [
            {
                id: 'application.http',
                label: core_1.nls.localizeByDefault('HTTP'),
                settings: ['http.*']
            },
            {
                id: 'application.keyboard',
                label: core_1.nls.localizeByDefault('Keyboard'),
                settings: ['keyboard.*']
            },
            {
                id: 'application.update',
                label: core_1.nls.localizeByDefault('Update'),
                settings: ['update.*']
            },
            {
                id: 'application.telemetry',
                label: core_1.nls.localizeByDefault('Telemetry'),
                settings: ['telemetry.*']
            },
            {
                id: 'application.settingsSync',
                label: core_1.nls.localizeByDefault('Settings Sync'),
                settings: ['settingsSync.*']
            },
            {
                id: 'application.experimental',
                label: core_1.nls.localizeByDefault('Experimental'),
                settings: ['application.experimental.*']
            },
            {
                id: 'application.other',
                label: core_1.nls.localizeByDefault('Other'),
                settings: ['application.*']
            }
        ]
    },
    {
        id: 'security',
        label: core_1.nls.localizeByDefault('Security'),
        settings: ['security.*'],
        children: [
            {
                id: 'security.workspace',
                label: core_1.nls.localizeByDefault('Workspace'),
                settings: ['security.workspace.*']
            }
        ]
    },
    {
        id: 'ai-features',
        label: 'AI Features', // TODO localize
    },
    {
        id: 'extensions',
        label: core_1.nls.localizeByDefault('Extensions'),
        children: [
            {
                id: 'extensions.hosted-plugin',
                label: core_1.nls.localize('theia/preferences/hostedPlugin', 'Hosted Plugin'),
                settings: ['hosted-plugin.*']
            }
        ]
    }
];
let PreferenceLayoutProvider = class PreferenceLayoutProvider {
    getLayout() {
        return exports.DEFAULT_LAYOUT;
    }
    getCommonlyUsedLayout() {
        return exports.COMMONLY_USED_LAYOUT;
    }
    hasCategory(id) {
        return [...this.getLayout(), this.getCommonlyUsedLayout()].some(e => e.id === id);
    }
    getLayoutForPreference(preferenceId) {
        const layout = this.getLayout();
        for (const section of layout) {
            const item = this.findItemInSection(section, preferenceId);
            if (item) {
                return item;
            }
        }
        return undefined;
    }
    findItemInSection(section, preferenceId) {
        // First check whether any of its children match the preferenceId.
        if (section.children) {
            for (const child of section.children) {
                const item = this.findItemInSection(child, preferenceId);
                if (item) {
                    return item;
                }
            }
        }
        // Then check whether the section itself matches the preferenceId.
        if (section.settings) {
            for (const setting of section.settings) {
                if (this.matchesSetting(preferenceId, setting)) {
                    return section;
                }
            }
        }
        return undefined;
    }
    matchesSetting(preferenceId, setting) {
        if (setting.includes('*')) {
            return this.createRegExp(setting).test(preferenceId);
        }
        return preferenceId === setting;
    }
    createRegExp(setting) {
        return new RegExp(`^${setting.replace(/\./g, '\\.').replace(/\*/g, '.*')}$`);
    }
};
exports.PreferenceLayoutProvider = PreferenceLayoutProvider;
exports.PreferenceLayoutProvider = PreferenceLayoutProvider = tslib_1.__decorate([
    (0, inversify_1.injectable)()
], PreferenceLayoutProvider);
