// @theia/monaco-editor-core/esm/vs/editor/contrib/editorState/browser/keybindingCancellation.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditorKeybindingCancellationTokenSource = void 0;
const editorExtensions_js_1 = require("../../../browser/editorExtensions.js");
const contextkey_js_1 = require("../../../../platform/contextkey/common/contextkey.js");
const cancellation_js_1 = require("../../../../base/common/cancellation.js");
const linkedList_js_1 = require("../../../../base/common/linkedList.js");
const instantiation_js_1 = require("../../../../platform/instantiation/common/instantiation.js");
const extensions_js_1 = require("../../../../platform/instantiation/common/extensions.js");
const nls_js_1 = require("../../../../nls.js");
const IEditorCancellationTokens = (0, instantiation_js_1.createDecorator)('IEditorCancelService');
const ctxCancellableOperation = new contextkey_js_1.RawContextKey('cancellableOperation', false, (0, nls_js_1.localize)('cancellableOperation', 'Whether the editor runs a cancellable operation, e.g. like \'Peek References\''));
(0, extensions_js_1.registerSingleton)(IEditorCancellationTokens, class {
    constructor() {
        this._tokens = new WeakMap();
    }
    add(editor, cts) {
        let data = this._tokens.get(editor);
        if (!data) {
            data = editor.invokeWithinContext(accessor => {
                const key = ctxCancellableOperation.bindTo(accessor.get(contextkey_js_1.IContextKeyService));
                const tokens = new linkedList_js_1.LinkedList();
                return { key, tokens };
            });
            this._tokens.set(editor, data);
        }
        let removeFn;
        data.key.set(true);
        removeFn = data.tokens.push(cts);
        return () => {
            // remove w/o cancellation
            if (removeFn) {
                removeFn();
                data.key.set(!data.tokens.isEmpty());
                removeFn = undefined;
            }
        };
    }
    cancel(editor) {
        const data = this._tokens.get(editor);
        if (!data) {
            return;
        }
        // remove with cancellation
        const cts = data.tokens.pop();
        if (cts) {
            cts.cancel();
            data.key.set(!data.tokens.isEmpty());
        }
    }
}, 1 /* InstantiationType.Delayed */);
class EditorKeybindingCancellationTokenSource extends cancellation_js_1.CancellationTokenSource {
    constructor(editor, parent) {
        super(parent);
        this.editor = editor;
        this._unregister = editor.invokeWithinContext(accessor => accessor.get(IEditorCancellationTokens).add(editor, this));
    }
    dispose() {
        this._unregister();
        super.dispose();
    }
}
exports.EditorKeybindingCancellationTokenSource = EditorKeybindingCancellationTokenSource;
(0, editorExtensions_js_1.registerEditorCommand)(new class extends editorExtensions_js_1.EditorCommand {
    constructor() {
        super({
            id: 'editor.cancelOperation',
            kbOpts: {
                weight: 100 /* KeybindingWeight.EditorContrib */,
                primary: 9 /* KeyCode.Escape */
            },
            precondition: ctxCancellableOperation
        });
    }
    runEditorCommand(accessor, editor) {
        accessor.get(IEditorCancellationTokens).cancel(editor);
    }
});
