// @theia/monaco-editor-core/esm/vs/platform/keybinding/common/baseResolvedKeybinding.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseResolvedKeybinding = void 0;
const errors_js_1 = require("../../../base/common/errors.js");
const keybindingLabels_js_1 = require("../../../base/common/keybindingLabels.js");
const keybindings_js_1 = require("../../../base/common/keybindings.js");
class BaseResolvedKeybinding extends keybindings_js_1.ResolvedKeybinding {
    constructor(os, chords) {
        super();
        if (chords.length === 0) {
            throw (0, errors_js_1.illegalArgument)(`chords`);
        }
        this._os = os;
        this._chords = chords;
    }
    getLabel() {
        return keybindingLabels_js_1.UILabelProvider.toLabel(this._os, this._chords, (keybinding) => this._getLabel(keybinding));
    }
    getAriaLabel() {
        return keybindingLabels_js_1.AriaLabelProvider.toLabel(this._os, this._chords, (keybinding) => this._getAriaLabel(keybinding));
    }
    getElectronAccelerator() {
        if (this._chords.length > 1) {
            // [Electron Accelerators] Electron cannot handle chords
            return null;
        }
        if (this._chords[0].isDuplicateModifierCase()) {
            // [Electron Accelerators] Electron cannot handle modifier only keybindings
            // e.g. "shift shift"
            return null;
        }
        return keybindingLabels_js_1.ElectronAcceleratorLabelProvider.toLabel(this._os, this._chords, (keybinding) => this._getElectronAccelerator(keybinding));
    }
    getUserSettingsLabel() {
        return keybindingLabels_js_1.UserSettingsLabelProvider.toLabel(this._os, this._chords, (keybinding) => this._getUserSettingsLabel(keybinding));
    }
    isWYSIWYG() {
        return this._chords.every((keybinding) => this._isWYSIWYG(keybinding));
    }
    hasMultipleChords() {
        return (this._chords.length > 1);
    }
    getChords() {
        return this._chords.map((keybinding) => this._getChord(keybinding));
    }
    _getChord(keybinding) {
        return new keybindings_js_1.ResolvedChord(keybinding.ctrlKey, keybinding.shiftKey, keybinding.altKey, keybinding.metaKey, this._getLabel(keybinding), this._getAriaLabel(keybinding));
    }
    getDispatchChords() {
        return this._chords.map((keybinding) => this._getChordDispatch(keybinding));
    }
    getSingleModifierDispatchChords() {
        return this._chords.map((keybinding) => this._getSingleModifierChordDispatch(keybinding));
    }
}
exports.BaseResolvedKeybinding = BaseResolvedKeybinding;
