// @theia/terminal/lib/node/shell-process.js
"use strict";
// *****************************************************************************
// Copyright (C) 2017 Ericsson and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0.
//
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License v. 2.0 are satisfied: GNU General Public License, version 2
// with the GNU Classpath Exception which is available at
// https://www.gnu.org/software/classpath/license.html.
//
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
// *****************************************************************************
var ShellProcess_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShellProcess = exports.getRootPath = exports.ShellProcessOptions = exports.ShellProcessFactory = void 0;
const tslib_1 = require("tslib");
const inversify_1 = require("@theia/core/shared/inversify");
const os = require("os");
const logger_1 = require("@theia/core/lib/common/logger");
const node_1 = require("@theia/process/lib/node");
const common_1 = require("@theia/core/lib/common");
const uri_1 = require("@theia/core/lib/common/uri");
const file_uri_1 = require("@theia/core/lib/common/file-uri");
const environment_utils_1 = require("@theia/core/lib/node/environment-utils");
const utils_1 = require("@theia/process/lib/node/utils");
exports.ShellProcessFactory = Symbol('ShellProcessFactory');
exports.ShellProcessOptions = Symbol('ShellProcessOptions');
function getRootPath(rootURI) {
    if (rootURI) {
        const uri = new uri_1.default(rootURI);
        return file_uri_1.FileUri.fsPath(uri);
    }
    else {
        return os.homedir();
    }
}
exports.getRootPath = getRootPath;
let ShellProcess = ShellProcess_1 = class ShellProcess extends node_1.TerminalProcess {
    constructor(options, processManager, ringBuffer, logger, environmentUtils) {
        const env = { 'COLORTERM': 'truecolor' };
        super({
            command: options.shell || ShellProcess_1.getShellExecutablePath(),
            args: options.args || ShellProcess_1.getShellExecutableArgs(),
            options: {
                name: 'xterm-256color',
                cols: options.cols || ShellProcess_1.defaultCols,
                rows: options.rows || ShellProcess_1.defaultRows,
                cwd: getRootPath(options.rootURI),
                env: options.strictEnv !== true ? Object.assign(env, environmentUtils.mergeProcessEnv(options.env)) : Object.assign(env, options.env),
            },
            isPseudo: options.isPseudo,
        }, processManager, ringBuffer, logger);
    }
    static getShellExecutablePath() {
        const shell = process.env.THEIA_SHELL;
        if (shell) {
            return shell;
        }
        if (common_1.isWindows) {
            return 'cmd.exe';
        }
        else {
            return process.env.SHELL;
        }
    }
    static getShellExecutableArgs() {
        const args = process.env.THEIA_SHELL_ARGS;
        if (args) {
            return (0, utils_1.parseArgs)(args);
        }
        if (common_1.isOSX) {
            return ['-l'];
        }
        else {
            return [];
        }
    }
};
exports.ShellProcess = ShellProcess;
ShellProcess.defaultCols = 80;
ShellProcess.defaultRows = 24;
exports.ShellProcess = ShellProcess = ShellProcess_1 = tslib_1.__decorate([
    (0, inversify_1.injectable)(),
    tslib_1.__param(0, (0, inversify_1.inject)(exports.ShellProcessOptions)),
    tslib_1.__param(1, (0, inversify_1.inject)(node_1.ProcessManager)),
    tslib_1.__param(2, (0, inversify_1.inject)(node_1.MultiRingBuffer)),
    tslib_1.__param(3, (0, inversify_1.inject)(logger_1.ILogger)),
    tslib_1.__param(3, (0, inversify_1.named)('terminal')),
    tslib_1.__param(4, (0, inversify_1.inject)(environment_utils_1.EnvironmentUtils)),
    tslib_1.__metadata("design:paramtypes", [Object, node_1.ProcessManager,
        node_1.MultiRingBuffer, Object, environment_utils_1.EnvironmentUtils])
], ShellProcess);
