// @theia/monaco-editor-core/esm/vs/platform/telemetry/common/telemetry.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.TELEMETRY_OLD_SETTING_ID = exports.TELEMETRY_CRASH_REPORTER_SETTING_ID = exports.TELEMETRY_SETTING_ID = exports.TELEMETRY_SECTION_ID = exports.devDeviceIdKey = exports.sqmIdKey = exports.machineIdKey = exports.lastSessionDateStorageKey = exports.firstSessionDateStorageKey = exports.currentSessionDateStorageKey = exports.ICustomEndpointTelemetryService = exports.ITelemetryService = void 0;
const instantiation_js_1 = require("../../instantiation/common/instantiation.js");
exports.ITelemetryService = (0, instantiation_js_1.createDecorator)('telemetryService');
exports.ICustomEndpointTelemetryService = (0, instantiation_js_1.createDecorator)('customEndpointTelemetryService');
// Keys
exports.currentSessionDateStorageKey = 'telemetry.currentSessionDate';
exports.firstSessionDateStorageKey = 'telemetry.firstSessionDate';
exports.lastSessionDateStorageKey = 'telemetry.lastSessionDate';
exports.machineIdKey = 'telemetry.machineId';
exports.sqmIdKey = 'telemetry.sqmId';
exports.devDeviceIdKey = 'telemetry.devDeviceId';
// Configuration Keys
exports.TELEMETRY_SECTION_ID = 'telemetry';
exports.TELEMETRY_SETTING_ID = 'telemetry.telemetryLevel';
exports.TELEMETRY_CRASH_REPORTER_SETTING_ID = 'telemetry.enableCrashReporter';
exports.TELEMETRY_OLD_SETTING_ID = 'telemetry.enableTelemetry';
