// @theia/monaco-editor-core/esm/vs/editor/contrib/codeAction/browser/codeAction.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplyCodeActionReason = exports.fixAllCommandId = exports.organizeImportsCommandId = exports.sourceActionCommandId = exports.refactorPreviewCommandId = exports.refactorCommandId = exports.autoFixCommandId = exports.quickFixCommandId = exports.codeActionCommandId = void 0;
exports.getCodeActions = getCodeActions;
exports.applyCodeAction = applyCodeAction;
const arrays_js_1 = require("../../../../base/common/arrays.js");
const cancellation_js_1 = require("../../../../base/common/cancellation.js");
const errors_js_1 = require("../../../../base/common/errors.js");
const lifecycle_js_1 = require("../../../../base/common/lifecycle.js");
const uri_js_1 = require("../../../../base/common/uri.js");
const bulkEditService_js_1 = require("../../../browser/services/bulkEditService.js");
const range_js_1 = require("../../../common/core/range.js");
const selection_js_1 = require("../../../common/core/selection.js");
const languageFeatures_js_1 = require("../../../common/services/languageFeatures.js");
const model_js_1 = require("../../../common/services/model.js");
const editorState_js_1 = require("../../editorState/browser/editorState.js");
const nls = require("../../../../nls.js");
const commands_js_1 = require("../../../../platform/commands/common/commands.js");
const notification_js_1 = require("../../../../platform/notification/common/notification.js");
const progress_js_1 = require("../../../../platform/progress/common/progress.js");
const telemetry_js_1 = require("../../../../platform/telemetry/common/telemetry.js");
const types_js_1 = require("../common/types.js");
const hierarchicalKind_js_1 = require("../../../../base/common/hierarchicalKind.js");
const accessibilitySignalService_js_1 = require("../../../../platform/accessibilitySignal/browser/accessibilitySignalService.js");
exports.codeActionCommandId = 'editor.action.codeAction';
exports.quickFixCommandId = 'editor.action.quickFix';
exports.autoFixCommandId = 'editor.action.autoFix';
exports.refactorCommandId = 'editor.action.refactor';
exports.refactorPreviewCommandId = 'editor.action.refactor.preview';
exports.sourceActionCommandId = 'editor.action.sourceAction';
exports.organizeImportsCommandId = 'editor.action.organizeImports';
exports.fixAllCommandId = 'editor.action.fixAll';
class ManagedCodeActionSet extends lifecycle_js_1.Disposable {
    static codeActionsPreferredComparator(a, b) {
        if (a.isPreferred && !b.isPreferred) {
            return -1;
        }
        else if (!a.isPreferred && b.isPreferred) {
            return 1;
        }
        else {
            return 0;
        }
    }
    static codeActionsComparator({ action: a }, { action: b }) {
        if (a.isAI && !b.isAI) {
            return 1;
        }
        else if (!a.isAI && b.isAI) {
            return -1;
        }
        if ((0, arrays_js_1.isNonEmptyArray)(a.diagnostics)) {
            return (0, arrays_js_1.isNonEmptyArray)(b.diagnostics) ? ManagedCodeActionSet.codeActionsPreferredComparator(a, b) : -1;
        }
        else if ((0, arrays_js_1.isNonEmptyArray)(b.diagnostics)) {
            return 1;
        }
        else {
            return ManagedCodeActionSet.codeActionsPreferredComparator(a, b); // both have no diagnostics
        }
    }
    constructor(actions, documentation, disposables) {
        super();
        this.documentation = documentation;
        this._register(disposables);
        this.allActions = [...actions].sort(ManagedCodeActionSet.codeActionsComparator);
        this.validActions = this.allActions.filter(({ action }) => !action.disabled);
    }
    get hasAutoFix() {
        return this.validActions.some(({ action: fix }) => !!fix.kind && types_js_1.CodeActionKind.QuickFix.contains(new hierarchicalKind_js_1.HierarchicalKind(fix.kind)) && !!fix.isPreferred);
    }
    get hasAIFix() {
        return this.validActions.some(({ action: fix }) => !!fix.isAI);
    }
    get allAIFixes() {
        return this.validActions.every(({ action: fix }) => !!fix.isAI);
    }
}
const emptyCodeActionsResponse = { actions: [], documentation: undefined };
async function getCodeActions(registry, model, rangeOrSelection, trigger, progress, token) {
    const filter = trigger.filter || {};
    const notebookFilter = {
        ...filter,
        excludes: [...(filter.excludes || []), types_js_1.CodeActionKind.Notebook],
    };
    const codeActionContext = {
        only: filter.include?.value,
        trigger: trigger.type,
    };
    const cts = new editorState_js_1.TextModelCancellationTokenSource(model, token);
    // if the trigger is auto (autosave, lightbulb, etc), we should exclude notebook codeActions
    const excludeNotebookCodeActions = (trigger.type === 2 /* languages.CodeActionTriggerType.Auto */);
    const providers = getCodeActionProviders(registry, model, (excludeNotebookCodeActions) ? notebookFilter : filter);
    const disposables = new lifecycle_js_1.DisposableStore();
    const promises = providers.map(async (provider) => {
        const handle = setTimeout(() => progress.report(provider), 1250);
        try {
            const providedCodeActions = await provider.provideCodeActions(model, rangeOrSelection, codeActionContext, cts.token);
            if (providedCodeActions) {
                disposables.add(providedCodeActions);
            }
            if (cts.token.isCancellationRequested) {
                return emptyCodeActionsResponse;
            }
            const filteredActions = (providedCodeActions?.actions || []).filter(action => action && (0, types_js_1.filtersAction)(filter, action));
            const documentation = getDocumentationFromProvider(provider, filteredActions, filter.include);
            return {
                actions: filteredActions.map(action => new types_js_1.CodeActionItem(action, provider)),
                documentation
            };
        }
        catch (err) {
            if ((0, errors_js_1.isCancellationError)(err)) {
                throw err;
            }
            (0, errors_js_1.onUnexpectedExternalError)(err);
            return emptyCodeActionsResponse;
        }
        finally {
            clearTimeout(handle);
        }
    });
    const listener = registry.onDidChange(() => {
        const newProviders = registry.all(model);
        if (!(0, arrays_js_1.equals)(newProviders, providers)) {
            cts.cancel();
        }
    });
    try {
        const actions = await Promise.all(promises);
        const allActions = actions.map(x => x.actions).flat();
        const allDocumentation = [
            ...(0, arrays_js_1.coalesce)(actions.map(x => x.documentation)),
            ...getAdditionalDocumentationForShowingActions(registry, model, trigger, allActions)
        ];
        return new ManagedCodeActionSet(allActions, allDocumentation, disposables);
    }
    finally {
        listener.dispose();
        cts.dispose();
    }
}
function getCodeActionProviders(registry, model, filter) {
    return registry.all(model)
        // Don't include providers that we know will not return code actions of interest
        .filter(provider => {
        if (!provider.providedCodeActionKinds) {
            // We don't know what type of actions this provider will return.
            return true;
        }
        return provider.providedCodeActionKinds.some(kind => (0, types_js_1.mayIncludeActionsOfKind)(filter, new hierarchicalKind_js_1.HierarchicalKind(kind)));
    });
}
function* getAdditionalDocumentationForShowingActions(registry, model, trigger, actionsToShow) {
    if (model && actionsToShow.length) {
        for (const provider of registry.all(model)) {
            if (provider._getAdditionalMenuItems) {
                yield* provider._getAdditionalMenuItems?.({ trigger: trigger.type, only: trigger.filter?.include?.value }, actionsToShow.map(item => item.action));
            }
        }
    }
}
function getDocumentationFromProvider(provider, providedCodeActions, only) {
    if (!provider.documentation) {
        return undefined;
    }
    const documentation = provider.documentation.map(entry => ({ kind: new hierarchicalKind_js_1.HierarchicalKind(entry.kind), command: entry.command }));
    if (only) {
        let currentBest;
        for (const entry of documentation) {
            if (entry.kind.contains(only)) {
                if (!currentBest) {
                    currentBest = entry;
                }
                else {
                    // Take best match
                    if (currentBest.kind.contains(entry.kind)) {
                        currentBest = entry;
                    }
                }
            }
        }
        if (currentBest) {
            return currentBest?.command;
        }
    }
    // Otherwise, check to see if any of the provided actions match.
    for (const action of providedCodeActions) {
        if (!action.kind) {
            continue;
        }
        for (const entry of documentation) {
            if (entry.kind.contains(new hierarchicalKind_js_1.HierarchicalKind(action.kind))) {
                return entry.command;
            }
        }
    }
    return undefined;
}
var ApplyCodeActionReason;
(function (ApplyCodeActionReason) {
    ApplyCodeActionReason["OnSave"] = "onSave";
    ApplyCodeActionReason["FromProblemsView"] = "fromProblemsView";
    ApplyCodeActionReason["FromCodeActions"] = "fromCodeActions";
    ApplyCodeActionReason["FromAILightbulb"] = "fromAILightbulb";
    ApplyCodeActionReason["FromProblemsHover"] = "fromProblemsHover";
})(ApplyCodeActionReason || (exports.ApplyCodeActionReason = ApplyCodeActionReason = {}));
async function applyCodeAction(accessor, item, codeActionReason, options, token = cancellation_js_1.CancellationToken.None) {
    const bulkEditService = accessor.get(bulkEditService_js_1.IBulkEditService);
    const commandService = accessor.get(commands_js_1.ICommandService);
    const telemetryService = accessor.get(telemetry_js_1.ITelemetryService);
    const notificationService = accessor.get(notification_js_1.INotificationService);
    const accessibilitySignalService = accessor.get(accessibilitySignalService_js_1.IAccessibilitySignalService);
    telemetryService.publicLog2('codeAction.applyCodeAction', {
        codeActionTitle: item.action.title,
        codeActionKind: item.action.kind,
        codeActionIsPreferred: !!item.action.isPreferred,
        reason: codeActionReason,
    });
    accessibilitySignalService.playSignal(accessibilitySignalService_js_1.AccessibilitySignal.codeActionTriggered);
    await item.resolve(token);
    if (token.isCancellationRequested) {
        return;
    }
    if (item.action.edit?.edits.length) {
        const result = await bulkEditService.apply(item.action.edit, {
            editor: options?.editor,
            label: item.action.title,
            quotableLabel: item.action.title,
            code: 'undoredo.codeAction',
            respectAutoSaveConfig: codeActionReason !== ApplyCodeActionReason.OnSave,
            showPreview: options?.preview,
        });
        if (!result.isApplied) {
            return;
        }
    }
    if (item.action.command) {
        try {
            await commandService.executeCommand(item.action.command.id, ...(item.action.command.arguments || []));
        }
        catch (err) {
            const message = asMessage(err);
            notificationService.error(typeof message === 'string'
                ? message
                : nls.localize('applyCodeActionFailed', "An unknown error occurred while applying the code action"));
        }
    }
    // ensure the start sound and end sound do not overlap
    setTimeout(() => accessibilitySignalService.playSignal(accessibilitySignalService_js_1.AccessibilitySignal.codeActionApplied), 100);
}
function asMessage(err) {
    if (typeof err === 'string') {
        return err;
    }
    else if (err instanceof Error && typeof err.message === 'string') {
        return err.message;
    }
    else {
        return undefined;
    }
}
commands_js_1.CommandsRegistry.registerCommand('_executeCodeActionProvider', async function (accessor, resource, rangeOrSelection, kind, itemResolveCount) {
    if (!(resource instanceof uri_js_1.URI)) {
        throw (0, errors_js_1.illegalArgument)();
    }
    const { codeActionProvider } = accessor.get(languageFeatures_js_1.ILanguageFeaturesService);
    const model = accessor.get(model_js_1.IModelService).getModel(resource);
    if (!model) {
        throw (0, errors_js_1.illegalArgument)();
    }
    const validatedRangeOrSelection = selection_js_1.Selection.isISelection(rangeOrSelection)
        ? selection_js_1.Selection.liftSelection(rangeOrSelection)
        : range_js_1.Range.isIRange(rangeOrSelection)
            ? model.validateRange(rangeOrSelection)
            : undefined;
    if (!validatedRangeOrSelection) {
        throw (0, errors_js_1.illegalArgument)();
    }
    const include = typeof kind === 'string' ? new hierarchicalKind_js_1.HierarchicalKind(kind) : undefined;
    const codeActionSet = await getCodeActions(codeActionProvider, model, validatedRangeOrSelection, { type: 1 /* languages.CodeActionTriggerType.Invoke */, triggerAction: types_js_1.CodeActionTriggerSource.Default, filter: { includeSourceActions: true, include } }, progress_js_1.Progress.None, cancellation_js_1.CancellationToken.None);
    const resolving = [];
    const resolveCount = Math.min(codeActionSet.validActions.length, typeof itemResolveCount === 'number' ? itemResolveCount : 0);
    for (let i = 0; i < resolveCount; i++) {
        resolving.push(codeActionSet.validActions[i].resolve(cancellation_js_1.CancellationToken.None));
    }
    try {
        await Promise.all(resolving);
        return codeActionSet.validActions.map(item => item.action);
    }
    finally {
        setTimeout(() => codeActionSet.dispose(), 100);
    }
});
