// @theia/monaco-editor-core/esm/vs/editor/contrib/parameterHints/browser/parameterHints.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var ParameterHintsController_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TriggerParameterHintsAction = exports.ParameterHintsController = void 0;
const lazy_js_1 = require("../../../../base/common/lazy.js");
const lifecycle_js_1 = require("../../../../base/common/lifecycle.js");
const editorExtensions_js_1 = require("../../../browser/editorExtensions.js");
const editorContextKeys_js_1 = require("../../../common/editorContextKeys.js");
const languages = require("../../../common/languages.js");
const languageFeatures_js_1 = require("../../../common/services/languageFeatures.js");
const parameterHintsModel_js_1 = require("./parameterHintsModel.js");
const provideSignatureHelp_js_1 = require("./provideSignatureHelp.js");
const nls = require("../../../../nls.js");
const contextkey_js_1 = require("../../../../platform/contextkey/common/contextkey.js");
const instantiation_js_1 = require("../../../../platform/instantiation/common/instantiation.js");
const parameterHintsWidget_js_1 = require("./parameterHintsWidget.js");
let ParameterHintsController = class ParameterHintsController extends lifecycle_js_1.Disposable {
    static { ParameterHintsController_1 = this; }
    static { this.ID = 'editor.controller.parameterHints'; }
    static get(editor) {
        return editor.getContribution(ParameterHintsController_1.ID);
    }
    constructor(editor, instantiationService, languageFeaturesService) {
        super();
        this.editor = editor;
        this.model = this._register(new parameterHintsModel_js_1.ParameterHintsModel(editor, languageFeaturesService.signatureHelpProvider));
        this._register(this.model.onChangedHints(newParameterHints => {
            if (newParameterHints) {
                this.widget.value.show();
                this.widget.value.render(newParameterHints);
            }
            else {
                this.widget.rawValue?.hide();
            }
        }));
        this.widget = new lazy_js_1.Lazy(() => this._register(instantiationService.createInstance(parameterHintsWidget_js_1.ParameterHintsWidget, this.editor, this.model)));
    }
    cancel() {
        this.model.cancel();
    }
    previous() {
        this.widget.rawValue?.previous();
    }
    next() {
        this.widget.rawValue?.next();
    }
    trigger(context) {
        this.model.trigger(context, 0);
    }
};
exports.ParameterHintsController = ParameterHintsController;
exports.ParameterHintsController = ParameterHintsController = ParameterHintsController_1 = __decorate([
    __param(1, instantiation_js_1.IInstantiationService),
    __param(2, languageFeatures_js_1.ILanguageFeaturesService)
], ParameterHintsController);
class TriggerParameterHintsAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.triggerParameterHints',
            label: nls.localize2('parameterHints.trigger.label', "Trigger Parameter Hints"),
            precondition: editorContextKeys_js_1.EditorContextKeys.hasSignatureHelpProvider,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.editorTextFocus,
                primary: 2048 /* KeyMod.CtrlCmd */ | 1024 /* KeyMod.Shift */ | 10 /* KeyCode.Space */,
                weight: 100 /* KeybindingWeight.EditorContrib */
            }
        });
    }
    run(accessor, editor) {
        const controller = ParameterHintsController.get(editor);
        controller?.trigger({
            triggerKind: languages.SignatureHelpTriggerKind.Invoke
        });
    }
}
exports.TriggerParameterHintsAction = TriggerParameterHintsAction;
(0, editorExtensions_js_1.registerEditorContribution)(ParameterHintsController.ID, ParameterHintsController, 2 /* EditorContributionInstantiation.BeforeFirstInteraction */);
(0, editorExtensions_js_1.registerEditorAction)(TriggerParameterHintsAction);
const weight = 100 /* KeybindingWeight.EditorContrib */ + 75;
const ParameterHintsCommand = editorExtensions_js_1.EditorCommand.bindToContribution(ParameterHintsController.get);
(0, editorExtensions_js_1.registerEditorCommand)(new ParameterHintsCommand({
    id: 'closeParameterHints',
    precondition: provideSignatureHelp_js_1.Context.Visible,
    handler: x => x.cancel(),
    kbOpts: {
        weight: weight,
        kbExpr: editorContextKeys_js_1.EditorContextKeys.focus,
        primary: 9 /* KeyCode.Escape */,
        secondary: [1024 /* KeyMod.Shift */ | 9 /* KeyCode.Escape */]
    }
}));
(0, editorExtensions_js_1.registerEditorCommand)(new ParameterHintsCommand({
    id: 'showPrevParameterHint',
    precondition: contextkey_js_1.ContextKeyExpr.and(provideSignatureHelp_js_1.Context.Visible, provideSignatureHelp_js_1.Context.MultipleSignatures),
    handler: x => x.previous(),
    kbOpts: {
        weight: weight,
        kbExpr: editorContextKeys_js_1.EditorContextKeys.focus,
        primary: 16 /* KeyCode.UpArrow */,
        secondary: [512 /* KeyMod.Alt */ | 16 /* KeyCode.UpArrow */],
        mac: { primary: 16 /* KeyCode.UpArrow */, secondary: [512 /* KeyMod.Alt */ | 16 /* KeyCode.UpArrow */, 256 /* KeyMod.WinCtrl */ | 46 /* KeyCode.KeyP */] }
    }
}));
(0, editorExtensions_js_1.registerEditorCommand)(new ParameterHintsCommand({
    id: 'showNextParameterHint',
    precondition: contextkey_js_1.ContextKeyExpr.and(provideSignatureHelp_js_1.Context.Visible, provideSignatureHelp_js_1.Context.MultipleSignatures),
    handler: x => x.next(),
    kbOpts: {
        weight: weight,
        kbExpr: editorContextKeys_js_1.EditorContextKeys.focus,
        primary: 18 /* KeyCode.DownArrow */,
        secondary: [512 /* KeyMod.Alt */ | 18 /* KeyCode.DownArrow */],
        mac: { primary: 18 /* KeyCode.DownArrow */, secondary: [512 /* KeyMod.Alt */ | 18 /* KeyCode.DownArrow */, 256 /* KeyMod.WinCtrl */ | 44 /* KeyCode.KeyN */] }
    }
}));
