// @theia/monaco-editor-core/esm/vs/editor/contrib/parameterHints/browser/provideSignatureHelp.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.Context = void 0;
exports.provideSignatureHelp = provideSignatureHelp;
const cancellation_js_1 = require("../../../../base/common/cancellation.js");
const errors_js_1 = require("../../../../base/common/errors.js");
const types_js_1 = require("../../../../base/common/types.js");
const uri_js_1 = require("../../../../base/common/uri.js");
const position_js_1 = require("../../../common/core/position.js");
const languages = require("../../../common/languages.js");
const languageFeatures_js_1 = require("../../../common/services/languageFeatures.js");
const resolverService_js_1 = require("../../../common/services/resolverService.js");
const commands_js_1 = require("../../../../platform/commands/common/commands.js");
const contextkey_js_1 = require("../../../../platform/contextkey/common/contextkey.js");
exports.Context = {
    Visible: new contextkey_js_1.RawContextKey('parameterHintsVisible', false),
    MultipleSignatures: new contextkey_js_1.RawContextKey('parameterHintsMultipleSignatures', false),
};
async function provideSignatureHelp(registry, model, position, context, token) {
    const supports = registry.ordered(model);
    for (const support of supports) {
        try {
            const result = await support.provideSignatureHelp(model, position, token, context);
            if (result) {
                return result;
            }
        }
        catch (err) {
            (0, errors_js_1.onUnexpectedExternalError)(err);
        }
    }
    return undefined;
}
commands_js_1.CommandsRegistry.registerCommand('_executeSignatureHelpProvider', async (accessor, ...args) => {
    const [uri, position, triggerCharacter] = args;
    (0, types_js_1.assertType)(uri_js_1.URI.isUri(uri));
    (0, types_js_1.assertType)(position_js_1.Position.isIPosition(position));
    (0, types_js_1.assertType)(typeof triggerCharacter === 'string' || !triggerCharacter);
    const languageFeaturesService = accessor.get(languageFeatures_js_1.ILanguageFeaturesService);
    const ref = await accessor.get(resolverService_js_1.ITextModelService).createModelReference(uri);
    try {
        const result = await provideSignatureHelp(languageFeaturesService.signatureHelpProvider, ref.object.textEditorModel, position_js_1.Position.lift(position), {
            triggerKind: languages.SignatureHelpTriggerKind.Invoke,
            isRetrigger: false,
            triggerCharacter,
        }, cancellation_js_1.CancellationToken.None);
        if (!result) {
            return undefined;
        }
        setTimeout(() => result.dispose(), 0);
        return result.value;
    }
    finally {
        ref.dispose();
    }
});
