// @theia/native-webpack-plugin/lib/native-webpack-plugin.js
"use strict";
// *****************************************************************************
// Copyright (C) 2023 TypeFox and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0.
//
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License v. 2.0 are satisfied: GNU General Public License, version 2
// with the GNU Classpath Exception which is available at
// https://www.gnu.org/software/classpath/license.html.
//
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
// *****************************************************************************
Object.defineProperty(exports, "__esModule", { value: true });
exports.NativeWebpackPlugin = void 0;
const path = require("path");
const fs = require("fs");
const os = require("os");
const REQUIRE_RIPGREP = '@vscode/ripgrep';
const REQUIRE_VSCODE_WINDOWS_CA_CERTS = '@vscode/windows-ca-certs';
const REQUIRE_BINDINGS = 'bindings';
const REQUIRE_KEYMAPPING = './build/Release/keymapping';
const REQUIRE_PARCEL_WATCHER = './build/Release/watcher.node';
class NativeWebpackPlugin {
    constructor(options) {
        var _a;
        this.bindings = new Map();
        this.options = options;
        for (const [name, value] of Object.entries((_a = options.nativeBindings) !== null && _a !== void 0 ? _a : {})) {
            this.nativeBinding(name, value);
        }
    }
    nativeBinding(dependency, nodePath) {
        this.bindings.set(dependency, nodePath);
    }
    apply(compiler) {
        let replacements = {};
        let nodePtyIssuer;
        let trashHelperIssuer;
        let ripgrepIssuer;
        compiler.hooks.initialize.tap(NativeWebpackPlugin.name, async () => {
            var _a;
            const directory = path.resolve(compiler.outputPath, 'native-webpack-plugin');
            if (fs.existsSync(directory)) {
                await fs.promises.rm(directory, {
                    force: true,
                    recursive: true
                });
            }
            await fs.promises.mkdir(directory, {
                recursive: true
            });
            const bindingsFile = (issuer) => buildFile(directory, 'bindings.js', bindingsReplacement(issuer, Array.from(this.bindings.entries())));
            const ripgrepFile = () => buildFile(directory, 'ripgrep.js', ripgrepReplacement(this.options.out));
            const keymappingFile = () => Promise.resolve('./build/Release/keymapping.node');
            const windowsCaCertsFile = () => Promise.resolve('@vscode/windows-ca-certs/build/Release/crypt32.node');
            replacements = {
                ...((_a = this.options.replacements) !== null && _a !== void 0 ? _a : {}),
                [REQUIRE_RIPGREP]: ripgrepFile,
                [REQUIRE_BINDINGS]: bindingsFile,
                [REQUIRE_KEYMAPPING]: keymappingFile,
                [REQUIRE_VSCODE_WINDOWS_CA_CERTS]: windowsCaCertsFile,
                [REQUIRE_PARCEL_WATCHER]: issuer => Promise.resolve(findNativeWatcherFile(issuer))
            };
        });
        compiler.hooks.normalModuleFactory.tap(NativeWebpackPlugin.name, nmf => {
            nmf.hooks.beforeResolve.tapPromise(NativeWebpackPlugin.name, async (result) => {
                if (result.request === REQUIRE_RIPGREP) {
                    ripgrepIssuer = result.contextInfo.issuer;
                }
                else if (result.request === 'node-pty') {
                    nodePtyIssuer = result.contextInfo.issuer;
                }
                else if (result.request === 'trash') {
                    trashHelperIssuer = result.contextInfo.issuer;
                }
                for (const [file, replacement] of Object.entries(replacements)) {
                    if (result.request === file) {
                        result.request = await replacement(result.contextInfo.issuer);
                    }
                }
            });
            nmf.hooks.afterResolve.tapPromise(NativeWebpackPlugin.name, async (result) => {
                const createData = result.createData;
                for (const [file, replacement] of Object.entries(replacements)) {
                    if (createData.resource === file) {
                        createData.resource = await replacement(result.contextInfo.issuer);
                    }
                }
            });
        });
        compiler.hooks.afterEmit.tapPromise(NativeWebpackPlugin.name, async () => {
            if (this.options.trash && trashHelperIssuer) {
                await this.copyTrashHelper(trashHelperIssuer, compiler);
            }
            if (this.options.ripgrep && ripgrepIssuer) {
                await this.copyRipgrep(ripgrepIssuer, compiler);
            }
            if (this.options.pty && nodePtyIssuer) {
                await this.copyNodePtySpawnHelper(nodePtyIssuer, compiler);
            }
        });
    }
    async copyRipgrep(issuer, compiler) {
        const suffix = process.platform === 'win32' ? '.exe' : '';
        const sourceFile = require.resolve(`@vscode/ripgrep/bin/rg${suffix}`, { paths: [issuer] });
        const targetFile = path.join(compiler.outputPath, this.options.out, `rg${suffix}`);
        await this.copyExecutable(sourceFile, targetFile);
    }
    async copyNodePtySpawnHelper(issuer, compiler) {
        const targetDirectory = path.resolve(compiler.outputPath, '..', 'build', 'Release');
        if (process.platform === 'win32') {
            const agentFile = require.resolve('node-pty/build/Release/winpty-agent.exe', { paths: [issuer] });
            const targetAgentFile = path.join(targetDirectory, 'winpty-agent.exe');
            await this.copyExecutable(agentFile, targetAgentFile);
            const dllFile = require.resolve('node-pty/build/Release/winpty.dll', { paths: [issuer] });
            const targetDllFile = path.join(targetDirectory, 'winpty.dll');
            await this.copyExecutable(dllFile, targetDllFile);
        }
        else if (process.platform === 'darwin') {
            const sourceFile = require.resolve('node-pty/build/Release/spawn-helper', { paths: [issuer] });
            const targetFile = path.join(targetDirectory, 'spawn-helper');
            await this.copyExecutable(sourceFile, targetFile);
        }
    }
    async copyTrashHelper(issuer, compiler) {
        let sourceFile;
        let targetFile;
        if (process.platform === 'win32') {
            sourceFile = require.resolve('trash/lib/windows-trash.exe', { paths: [issuer] });
            targetFile = path.join(compiler.outputPath, 'windows-trash.exe');
        }
        else if (process.platform === 'darwin') {
            sourceFile = require.resolve('trash/lib/macos-trash', { paths: [issuer] });
            targetFile = path.join(compiler.outputPath, 'macos-trash');
        }
        if (sourceFile && targetFile) {
            await this.copyExecutable(sourceFile, targetFile);
        }
    }
    async copyExecutable(source, target) {
        const targetDirectory = path.dirname(target);
        await fs.promises.mkdir(targetDirectory, { recursive: true });
        await fs.promises.copyFile(source, target);
        await fs.promises.chmod(target, 0o777);
    }
}
exports.NativeWebpackPlugin = NativeWebpackPlugin;
function findNativeWatcherFile(issuer) {
    let name = `@parcel/watcher-${process.platform}-${process.arch}`;
    if (process.platform === 'linux') {
        const { MUSL, family } = require('detect-libc');
        if (family === MUSL) {
            name += '-musl';
        }
        else {
            name += '-glibc';
        }
    }
    return require.resolve(name, {
        paths: [issuer]
    });
}
async function buildFile(root, name, content) {
    const tmpFile = path.join(root, name);
    await fs.promises.writeFile(tmpFile, content);
    return tmpFile;
}
const ripgrepReplacement = (nativePath = '.') => `
const path = require('path');

exports.rgPath = path.join(__dirname, \`./${nativePath}/rg\${process.platform === 'win32' ? '.exe' : ''}\`);
`;
const bindingsReplacement = (issuer, entries) => {
    const cases = [];
    for (const [module, node] of entries) {
        const modulePath = require.resolve(node, {
            paths: [issuer]
        });
        cases.push(`${' '.repeat(8)}case '${module}': return require('${modulePath.replace(/\\/g, '/')}');`);
    }
    return `
module.exports = function (jsModule) {
    switch (jsModule) {
${cases.join(os.EOL)}
    }
    throw new Error(\`unhandled module: "\${jsModule}"\`);
}`.trim();
};
