// @theia/monaco-editor-core/esm/vs/editor/contrib/hover/browser/contentHoverStatusBar.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditorHoverStatusBar = void 0;
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
const dom = require("../../../../base/browser/dom.js");
const hoverWidget_js_1 = require("../../../../base/browser/ui/hover/hoverWidget.js");
const lifecycle_js_1 = require("../../../../base/common/lifecycle.js");
const keybinding_js_1 = require("../../../../platform/keybinding/common/keybinding.js");
const hover_js_1 = require("../../../../platform/hover/browser/hover.js");
const hoverDelegateFactory_js_1 = require("../../../../base/browser/ui/hover/hoverDelegateFactory.js");
const $ = dom.$;
let EditorHoverStatusBar = class EditorHoverStatusBar extends lifecycle_js_1.Disposable {
    get hasContent() {
        return this._hasContent;
    }
    constructor(_keybindingService, _hoverService) {
        super();
        this._keybindingService = _keybindingService;
        this._hoverService = _hoverService;
        this.actions = [];
        this._hasContent = false;
        this.hoverElement = $('div.hover-row.status-bar');
        this.hoverElement.tabIndex = 0;
        this.actionsElement = dom.append(this.hoverElement, $('div.actions'));
    }
    addAction(actionOptions) {
        const keybinding = this._keybindingService.lookupKeybinding(actionOptions.commandId);
        const keybindingLabel = keybinding ? keybinding.getLabel() : null;
        this._hasContent = true;
        const action = this._register(hoverWidget_js_1.HoverAction.render(this.actionsElement, actionOptions, keybindingLabel));
        this._register(this._hoverService.setupManagedHover((0, hoverDelegateFactory_js_1.getDefaultHoverDelegate)('element'), action.actionContainer, action.actionRenderedLabel));
        this.actions.push(action);
        return action;
    }
    append(element) {
        const result = dom.append(this.actionsElement, element);
        this._hasContent = true;
        return result;
    }
};
exports.EditorHoverStatusBar = EditorHoverStatusBar;
exports.EditorHoverStatusBar = EditorHoverStatusBar = __decorate([
    __param(0, keybinding_js_1.IKeybindingService),
    __param(1, hover_js_1.IHoverService)
], EditorHoverStatusBar);
