// @theia/monaco-editor-core/esm/vs/editor/contrib/hover/browser/hoverAccessibleViews.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtHoverAccessibleView = exports.HoverAccessibleViewProvider = exports.HoverAccessibilityHelpProvider = exports.HoverAccessibilityHelp = exports.HoverAccessibleView = void 0;
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
const nls_js_1 = require("../../../../nls.js");
const editorContextKeys_js_1 = require("../../../common/editorContextKeys.js");
const contentHoverController_js_1 = require("./contentHoverController.js");
const accessibleView_js_1 = require("../../../../platform/accessibility/browser/accessibleView.js");
const contextView_js_1 = require("../../../../platform/contextview/browser/contextView.js");
const hover_js_1 = require("../../../../platform/hover/browser/hover.js");
const instantiation_js_1 = require("../../../../platform/instantiation/common/instantiation.js");
const languages_js_1 = require("../../../common/languages.js");
const hoverActionIds_js_1 = require("./hoverActionIds.js");
const codeEditorService_js_1 = require("../../../browser/services/codeEditorService.js");
const actions_js_1 = require("../../../../base/common/actions.js");
const themables_js_1 = require("../../../../base/common/themables.js");
const codicons_js_1 = require("../../../../base/common/codicons.js");
const event_js_1 = require("../../../../base/common/event.js");
const lifecycle_js_1 = require("../../../../base/common/lifecycle.js");
const keybinding_js_1 = require("../../../../platform/keybinding/common/keybinding.js");
const markdownHoverParticipant_js_1 = require("./markdownHoverParticipant.js");
var HoverAccessibilityHelpNLS;
(function (HoverAccessibilityHelpNLS) {
    HoverAccessibilityHelpNLS.increaseVerbosity = (0, nls_js_1.localize)('increaseVerbosity', '- The focused hover part verbosity level can be increased with the Increase Hover Verbosity command.', `<keybinding:${hoverActionIds_js_1.INCREASE_HOVER_VERBOSITY_ACTION_ID}>`);
    HoverAccessibilityHelpNLS.decreaseVerbosity = (0, nls_js_1.localize)('decreaseVerbosity', '- The focused hover part verbosity level can be decreased with the Decrease Hover Verbosity command.', `<keybinding:${hoverActionIds_js_1.DECREASE_HOVER_VERBOSITY_ACTION_ID}>`);
})(HoverAccessibilityHelpNLS || (HoverAccessibilityHelpNLS = {}));
class HoverAccessibleView {
    constructor() {
        this.type = "view" /* AccessibleViewType.View */;
        this.priority = 95;
        this.name = 'hover';
        this.when = editorContextKeys_js_1.EditorContextKeys.hoverFocused;
    }
    getProvider(accessor) {
        const codeEditorService = accessor.get(codeEditorService_js_1.ICodeEditorService);
        const codeEditor = codeEditorService.getActiveCodeEditor() || codeEditorService.getFocusedCodeEditor();
        if (!codeEditor) {
            throw new Error('No active or focused code editor');
        }
        const hoverController = contentHoverController_js_1.ContentHoverController.get(codeEditor);
        if (!hoverController) {
            return;
        }
        const keybindingService = accessor.get(keybinding_js_1.IKeybindingService);
        return accessor.get(instantiation_js_1.IInstantiationService).createInstance(HoverAccessibleViewProvider, keybindingService, codeEditor, hoverController);
    }
}
exports.HoverAccessibleView = HoverAccessibleView;
class HoverAccessibilityHelp {
    constructor() {
        this.priority = 100;
        this.name = 'hover';
        this.type = "help" /* AccessibleViewType.Help */;
        this.when = editorContextKeys_js_1.EditorContextKeys.hoverVisible;
    }
    getProvider(accessor) {
        const codeEditorService = accessor.get(codeEditorService_js_1.ICodeEditorService);
        const codeEditor = codeEditorService.getActiveCodeEditor() || codeEditorService.getFocusedCodeEditor();
        if (!codeEditor) {
            throw new Error('No active or focused code editor');
        }
        const hoverController = contentHoverController_js_1.ContentHoverController.get(codeEditor);
        if (!hoverController) {
            return;
        }
        return accessor.get(instantiation_js_1.IInstantiationService).createInstance(HoverAccessibilityHelpProvider, hoverController);
    }
}
exports.HoverAccessibilityHelp = HoverAccessibilityHelp;
class BaseHoverAccessibleViewProvider extends lifecycle_js_1.Disposable {
    constructor(_hoverController) {
        super();
        this._hoverController = _hoverController;
        this.id = "hover" /* AccessibleViewProviderId.Hover */;
        this.verbositySettingKey = 'accessibility.verbosity.hover';
        this._onDidChangeContent = this._register(new event_js_1.Emitter());
        this.onDidChangeContent = this._onDidChangeContent.event;
        this._focusedHoverPartIndex = -1;
    }
    onOpen() {
        if (!this._hoverController) {
            return;
        }
        this._hoverController.shouldKeepOpenOnEditorMouseMoveOrLeave = true;
        this._focusedHoverPartIndex = this._hoverController.focusedHoverPartIndex();
        this._register(this._hoverController.onHoverContentsChanged(() => {
            this._onDidChangeContent.fire();
        }));
    }
    onClose() {
        if (!this._hoverController) {
            return;
        }
        if (this._focusedHoverPartIndex === -1) {
            this._hoverController.focus();
        }
        else {
            this._hoverController.focusHoverPartWithIndex(this._focusedHoverPartIndex);
        }
        this._focusedHoverPartIndex = -1;
        this._hoverController.shouldKeepOpenOnEditorMouseMoveOrLeave = false;
    }
    provideContentAtIndex(focusedHoverIndex, includeVerbosityActions) {
        if (focusedHoverIndex !== -1) {
            const accessibleContent = this._hoverController.getAccessibleWidgetContentAtIndex(focusedHoverIndex);
            if (accessibleContent === undefined) {
                return '';
            }
            const contents = [];
            if (includeVerbosityActions) {
                contents.push(...this._descriptionsOfVerbosityActionsForIndex(focusedHoverIndex));
            }
            contents.push(accessibleContent);
            return contents.join('\n');
        }
        else {
            const accessibleContent = this._hoverController.getAccessibleWidgetContent();
            if (accessibleContent === undefined) {
                return '';
            }
            const contents = [];
            contents.push(accessibleContent);
            return contents.join('\n');
        }
    }
    _descriptionsOfVerbosityActionsForIndex(index) {
        const content = [];
        const descriptionForIncreaseAction = this._descriptionOfVerbosityActionForIndex(languages_js_1.HoverVerbosityAction.Increase, index);
        if (descriptionForIncreaseAction !== undefined) {
            content.push(descriptionForIncreaseAction);
        }
        const descriptionForDecreaseAction = this._descriptionOfVerbosityActionForIndex(languages_js_1.HoverVerbosityAction.Decrease, index);
        if (descriptionForDecreaseAction !== undefined) {
            content.push(descriptionForDecreaseAction);
        }
        return content;
    }
    _descriptionOfVerbosityActionForIndex(action, index) {
        const isActionSupported = this._hoverController.doesHoverAtIndexSupportVerbosityAction(index, action);
        if (!isActionSupported) {
            return;
        }
        switch (action) {
            case languages_js_1.HoverVerbosityAction.Increase:
                return HoverAccessibilityHelpNLS.increaseVerbosity;
            case languages_js_1.HoverVerbosityAction.Decrease:
                return HoverAccessibilityHelpNLS.decreaseVerbosity;
        }
    }
}
class HoverAccessibilityHelpProvider extends BaseHoverAccessibleViewProvider {
    constructor(hoverController) {
        super(hoverController);
        this.options = { type: "help" /* AccessibleViewType.Help */ };
    }
    provideContent() {
        return this.provideContentAtIndex(this._focusedHoverPartIndex, true);
    }
}
exports.HoverAccessibilityHelpProvider = HoverAccessibilityHelpProvider;
class HoverAccessibleViewProvider extends BaseHoverAccessibleViewProvider {
    constructor(_keybindingService, _editor, hoverController) {
        super(hoverController);
        this._keybindingService = _keybindingService;
        this._editor = _editor;
        this.options = { type: "view" /* AccessibleViewType.View */ };
        this._initializeOptions(this._editor, hoverController);
    }
    provideContent() {
        return this.provideContentAtIndex(this._focusedHoverPartIndex, false);
    }
    get actions() {
        const actions = [];
        actions.push(this._getActionFor(this._editor, languages_js_1.HoverVerbosityAction.Increase));
        actions.push(this._getActionFor(this._editor, languages_js_1.HoverVerbosityAction.Decrease));
        return actions;
    }
    _getActionFor(editor, action) {
        let actionId;
        let accessibleActionId;
        let actionCodicon;
        switch (action) {
            case languages_js_1.HoverVerbosityAction.Increase:
                actionId = hoverActionIds_js_1.INCREASE_HOVER_VERBOSITY_ACTION_ID;
                accessibleActionId = hoverActionIds_js_1.INCREASE_HOVER_VERBOSITY_ACCESSIBLE_ACTION_ID;
                actionCodicon = codicons_js_1.Codicon.add;
                break;
            case languages_js_1.HoverVerbosityAction.Decrease:
                actionId = hoverActionIds_js_1.DECREASE_HOVER_VERBOSITY_ACTION_ID;
                accessibleActionId = hoverActionIds_js_1.DECREASE_HOVER_VERBOSITY_ACCESSIBLE_ACTION_ID;
                actionCodicon = codicons_js_1.Codicon.remove;
                break;
        }
        const actionLabel = (0, markdownHoverParticipant_js_1.labelForHoverVerbosityAction)(this._keybindingService, action);
        const actionEnabled = this._hoverController.doesHoverAtIndexSupportVerbosityAction(this._focusedHoverPartIndex, action);
        return new actions_js_1.Action(accessibleActionId, actionLabel, themables_js_1.ThemeIcon.asClassName(actionCodicon), actionEnabled, () => {
            editor.getAction(actionId)?.run({ index: this._focusedHoverPartIndex, focus: false });
        });
    }
    _initializeOptions(editor, hoverController) {
        const helpProvider = this._register(new HoverAccessibilityHelpProvider(hoverController));
        this.options.language = editor.getModel()?.getLanguageId();
        this.options.customHelp = () => { return helpProvider.provideContentAtIndex(this._focusedHoverPartIndex, true); };
    }
}
exports.HoverAccessibleViewProvider = HoverAccessibleViewProvider;
class ExtHoverAccessibleView {
    constructor() {
        this.type = "view" /* AccessibleViewType.View */;
        this.priority = 90;
        this.name = 'extension-hover';
    }
    getProvider(accessor) {
        const contextViewService = accessor.get(contextView_js_1.IContextViewService);
        const contextViewElement = contextViewService.getContextViewElement();
        const extensionHoverContent = contextViewElement?.textContent ?? undefined;
        const hoverService = accessor.get(hover_js_1.IHoverService);
        if (contextViewElement.classList.contains('accessible-view-container') || !extensionHoverContent) {
            // The accessible view, itself, uses the context view service to display the text. We don't want to read that.
            return;
        }
        return new accessibleView_js_1.AccessibleContentProvider("hover" /* AccessibleViewProviderId.Hover */, { language: 'typescript', type: "view" /* AccessibleViewType.View */ }, () => { return extensionHoverContent; }, () => {
            hoverService.showAndFocusLastHover();
        }, 'accessibility.verbosity.hover');
    }
}
exports.ExtHoverAccessibleView = ExtHoverAccessibleView;
