// @theia/monaco-editor-core/esm/vs/editor/contrib/hover/browser/hoverActions.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecreaseHoverVerbosityLevel = exports.IncreaseHoverVerbosityLevel = exports.GoToBottomHoverAction = exports.GoToTopHoverAction = exports.PageDownHoverAction = exports.PageUpHoverAction = exports.ScrollRightHoverAction = exports.ScrollLeftHoverAction = exports.ScrollDownHoverAction = exports.ScrollUpHoverAction = exports.ShowDefinitionPreviewHoverAction = exports.ShowOrFocusHoverAction = void 0;
const hoverActionIds_js_1 = require("./hoverActionIds.js");
const keyCodes_js_1 = require("../../../../base/common/keyCodes.js");
const editorExtensions_js_1 = require("../../../browser/editorExtensions.js");
const range_js_1 = require("../../../common/core/range.js");
const editorContextKeys_js_1 = require("../../../common/editorContextKeys.js");
const goToDefinitionAtPosition_js_1 = require("../../gotoSymbol/browser/link/goToDefinitionAtPosition.js");
const contentHoverController_js_1 = require("./contentHoverController.js");
const languages_js_1 = require("../../../common/languages.js");
const nls = require("../../../../nls.js");
require("./hover.css");
var HoverFocusBehavior;
(function (HoverFocusBehavior) {
    HoverFocusBehavior["NoAutoFocus"] = "noAutoFocus";
    HoverFocusBehavior["FocusIfVisible"] = "focusIfVisible";
    HoverFocusBehavior["AutoFocusImmediately"] = "autoFocusImmediately";
})(HoverFocusBehavior || (HoverFocusBehavior = {}));
class ShowOrFocusHoverAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.SHOW_OR_FOCUS_HOVER_ACTION_ID,
            label: nls.localize2({
                key: 'showOrFocusHover',
                comment: [
                    'Label for action that will trigger the showing/focusing of a hover in the editor.',
                    'If the hover is not visible, it will show the hover.',
                    'This allows for users to show the hover without using the mouse.'
                ]
            }, "Show or Focus Hover"),
            metadata: {
                description: nls.localize2('showOrFocusHoverDescription', 'Show or focus the editor hover which shows documentation, references, and other content for a symbol at the current cursor position.'),
                args: [{
                        name: 'args',
                        schema: {
                            type: 'object',
                            properties: {
                                'focus': {
                                    description: 'Controls if and when the hover should take focus upon being triggered by this action.',
                                    enum: [HoverFocusBehavior.NoAutoFocus, HoverFocusBehavior.FocusIfVisible, HoverFocusBehavior.AutoFocusImmediately],
                                    enumDescriptions: [
                                        nls.localize('showOrFocusHover.focus.noAutoFocus', 'The hover will not automatically take focus.'),
                                        nls.localize('showOrFocusHover.focus.focusIfVisible', 'The hover will take focus only if it is already visible.'),
                                        nls.localize('showOrFocusHover.focus.autoFocusImmediately', 'The hover will automatically take focus when it appears.'),
                                    ],
                                    default: HoverFocusBehavior.FocusIfVisible,
                                }
                            },
                        }
                    }]
            },
            precondition: undefined,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.editorTextFocus,
                primary: (0, keyCodes_js_1.KeyChord)(2048 /* KeyMod.CtrlCmd */ | 41 /* KeyCode.KeyK */, 2048 /* KeyMod.CtrlCmd */ | 39 /* KeyCode.KeyI */),
                weight: 100 /* KeybindingWeight.EditorContrib */
            }
        });
    }
    run(accessor, editor, args) {
        if (!editor.hasModel()) {
            return;
        }
        const controller = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!controller) {
            return;
        }
        const focusArgument = args?.focus;
        let focusOption = HoverFocusBehavior.FocusIfVisible;
        if (Object.values(HoverFocusBehavior).includes(focusArgument)) {
            focusOption = focusArgument;
        }
        else if (typeof focusArgument === 'boolean' && focusArgument) {
            focusOption = HoverFocusBehavior.AutoFocusImmediately;
        }
        const showContentHover = (focus) => {
            const position = editor.getPosition();
            const range = new range_js_1.Range(position.lineNumber, position.column, position.lineNumber, position.column);
            controller.showContentHover(range, 1 /* HoverStartMode.Immediate */, 2 /* HoverStartSource.Keyboard */, focus);
        };
        const accessibilitySupportEnabled = editor.getOption(2 /* EditorOption.accessibilitySupport */) === 2 /* AccessibilitySupport.Enabled */;
        if (controller.isHoverVisible) {
            if (focusOption !== HoverFocusBehavior.NoAutoFocus) {
                controller.focus();
            }
            else {
                showContentHover(accessibilitySupportEnabled);
            }
        }
        else {
            showContentHover(accessibilitySupportEnabled || focusOption === HoverFocusBehavior.AutoFocusImmediately);
        }
    }
}
exports.ShowOrFocusHoverAction = ShowOrFocusHoverAction;
class ShowDefinitionPreviewHoverAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.SHOW_DEFINITION_PREVIEW_HOVER_ACTION_ID,
            label: nls.localize2({
                key: 'showDefinitionPreviewHover',
                comment: [
                    'Label for action that will trigger the showing of definition preview hover in the editor.',
                    'This allows for users to show the definition preview hover without using the mouse.'
                ]
            }, "Show Definition Preview Hover"),
            precondition: undefined,
            metadata: {
                description: nls.localize2('showDefinitionPreviewHoverDescription', 'Show the definition preview hover in the editor.'),
            },
        });
    }
    run(accessor, editor) {
        const controller = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!controller) {
            return;
        }
        const position = editor.getPosition();
        if (!position) {
            return;
        }
        const range = new range_js_1.Range(position.lineNumber, position.column, position.lineNumber, position.column);
        const goto = goToDefinitionAtPosition_js_1.GotoDefinitionAtPositionEditorContribution.get(editor);
        if (!goto) {
            return;
        }
        const promise = goto.startFindDefinitionFromCursor(position);
        promise.then(() => {
            controller.showContentHover(range, 1 /* HoverStartMode.Immediate */, 2 /* HoverStartSource.Keyboard */, true);
        });
    }
}
exports.ShowDefinitionPreviewHoverAction = ShowDefinitionPreviewHoverAction;
class ScrollUpHoverAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.SCROLL_UP_HOVER_ACTION_ID,
            label: nls.localize2({
                key: 'scrollUpHover',
                comment: [
                    'Action that allows to scroll up in the hover widget with the up arrow when the hover widget is focused.'
                ]
            }, "Scroll Up Hover"),
            precondition: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
                primary: 16 /* KeyCode.UpArrow */,
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            metadata: {
                description: nls.localize2('scrollUpHoverDescription', 'Scroll up the editor hover.')
            },
        });
    }
    run(accessor, editor) {
        const controller = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!controller) {
            return;
        }
        controller.scrollUp();
    }
}
exports.ScrollUpHoverAction = ScrollUpHoverAction;
class ScrollDownHoverAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.SCROLL_DOWN_HOVER_ACTION_ID,
            label: nls.localize2({
                key: 'scrollDownHover',
                comment: [
                    'Action that allows to scroll down in the hover widget with the up arrow when the hover widget is focused.'
                ]
            }, "Scroll Down Hover"),
            precondition: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
                primary: 18 /* KeyCode.DownArrow */,
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            metadata: {
                description: nls.localize2('scrollDownHoverDescription', 'Scroll down the editor hover.'),
            },
        });
    }
    run(accessor, editor) {
        const controller = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!controller) {
            return;
        }
        controller.scrollDown();
    }
}
exports.ScrollDownHoverAction = ScrollDownHoverAction;
class ScrollLeftHoverAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.SCROLL_LEFT_HOVER_ACTION_ID,
            label: nls.localize2({
                key: 'scrollLeftHover',
                comment: [
                    'Action that allows to scroll left in the hover widget with the left arrow when the hover widget is focused.'
                ]
            }, "Scroll Left Hover"),
            precondition: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
                primary: 15 /* KeyCode.LeftArrow */,
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            metadata: {
                description: nls.localize2('scrollLeftHoverDescription', 'Scroll left the editor hover.'),
            },
        });
    }
    run(accessor, editor) {
        const controller = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!controller) {
            return;
        }
        controller.scrollLeft();
    }
}
exports.ScrollLeftHoverAction = ScrollLeftHoverAction;
class ScrollRightHoverAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.SCROLL_RIGHT_HOVER_ACTION_ID,
            label: nls.localize2({
                key: 'scrollRightHover',
                comment: [
                    'Action that allows to scroll right in the hover widget with the right arrow when the hover widget is focused.'
                ]
            }, "Scroll Right Hover"),
            precondition: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
                primary: 17 /* KeyCode.RightArrow */,
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            metadata: {
                description: nls.localize2('scrollRightHoverDescription', 'Scroll right the editor hover.')
            },
        });
    }
    run(accessor, editor) {
        const controller = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!controller) {
            return;
        }
        controller.scrollRight();
    }
}
exports.ScrollRightHoverAction = ScrollRightHoverAction;
class PageUpHoverAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.PAGE_UP_HOVER_ACTION_ID,
            label: nls.localize2({
                key: 'pageUpHover',
                comment: [
                    'Action that allows to page up in the hover widget with the page up command when the hover widget is focused.'
                ]
            }, "Page Up Hover"),
            precondition: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
                primary: 11 /* KeyCode.PageUp */,
                secondary: [512 /* KeyMod.Alt */ | 16 /* KeyCode.UpArrow */],
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            metadata: {
                description: nls.localize2('pageUpHoverDescription', 'Page up the editor hover.'),
            },
        });
    }
    run(accessor, editor) {
        const controller = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!controller) {
            return;
        }
        controller.pageUp();
    }
}
exports.PageUpHoverAction = PageUpHoverAction;
class PageDownHoverAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.PAGE_DOWN_HOVER_ACTION_ID,
            label: nls.localize2({
                key: 'pageDownHover',
                comment: [
                    'Action that allows to page down in the hover widget with the page down command when the hover widget is focused.'
                ]
            }, "Page Down Hover"),
            precondition: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
                primary: 12 /* KeyCode.PageDown */,
                secondary: [512 /* KeyMod.Alt */ | 18 /* KeyCode.DownArrow */],
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            metadata: {
                description: nls.localize2('pageDownHoverDescription', 'Page down the editor hover.'),
            },
        });
    }
    run(accessor, editor) {
        const controller = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!controller) {
            return;
        }
        controller.pageDown();
    }
}
exports.PageDownHoverAction = PageDownHoverAction;
class GoToTopHoverAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.GO_TO_TOP_HOVER_ACTION_ID,
            label: nls.localize2({
                key: 'goToTopHover',
                comment: [
                    'Action that allows to go to the top of the hover widget with the home command when the hover widget is focused.'
                ]
            }, "Go To Top Hover"),
            precondition: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
                primary: 14 /* KeyCode.Home */,
                secondary: [2048 /* KeyMod.CtrlCmd */ | 16 /* KeyCode.UpArrow */],
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            metadata: {
                description: nls.localize2('goToTopHoverDescription', 'Go to the top of the editor hover.'),
            },
        });
    }
    run(accessor, editor) {
        const controller = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!controller) {
            return;
        }
        controller.goToTop();
    }
}
exports.GoToTopHoverAction = GoToTopHoverAction;
class GoToBottomHoverAction extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.GO_TO_BOTTOM_HOVER_ACTION_ID,
            label: nls.localize2({
                key: 'goToBottomHover',
                comment: [
                    'Action that allows to go to the bottom in the hover widget with the end command when the hover widget is focused.'
                ]
            }, "Go To Bottom Hover"),
            precondition: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
            kbOpts: {
                kbExpr: editorContextKeys_js_1.EditorContextKeys.hoverFocused,
                primary: 13 /* KeyCode.End */,
                secondary: [2048 /* KeyMod.CtrlCmd */ | 18 /* KeyCode.DownArrow */],
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            metadata: {
                description: nls.localize2('goToBottomHoverDescription', 'Go to the bottom of the editor hover.')
            },
        });
    }
    run(accessor, editor) {
        const controller = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!controller) {
            return;
        }
        controller.goToBottom();
    }
}
exports.GoToBottomHoverAction = GoToBottomHoverAction;
class IncreaseHoverVerbosityLevel extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.INCREASE_HOVER_VERBOSITY_ACTION_ID,
            label: hoverActionIds_js_1.INCREASE_HOVER_VERBOSITY_ACTION_LABEL,
            alias: 'Increase Hover Verbosity Level',
            precondition: editorContextKeys_js_1.EditorContextKeys.hoverVisible
        });
    }
    run(accessor, editor, args) {
        const hoverController = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!hoverController) {
            return;
        }
        const index = args?.index !== undefined ? args.index : hoverController.focusedHoverPartIndex();
        hoverController.updateHoverVerbosityLevel(languages_js_1.HoverVerbosityAction.Increase, index, args?.focus);
    }
}
exports.IncreaseHoverVerbosityLevel = IncreaseHoverVerbosityLevel;
class DecreaseHoverVerbosityLevel extends editorExtensions_js_1.EditorAction {
    constructor() {
        super({
            id: hoverActionIds_js_1.DECREASE_HOVER_VERBOSITY_ACTION_ID,
            label: hoverActionIds_js_1.DECREASE_HOVER_VERBOSITY_ACTION_LABEL,
            alias: 'Decrease Hover Verbosity Level',
            precondition: editorContextKeys_js_1.EditorContextKeys.hoverVisible
        });
    }
    run(accessor, editor, args) {
        const hoverController = contentHoverController_js_1.ContentHoverController.get(editor);
        if (!hoverController) {
            return;
        }
        const index = args?.index !== undefined ? args.index : hoverController.focusedHoverPartIndex();
        contentHoverController_js_1.ContentHoverController.get(editor)?.updateHoverVerbosityLevel(languages_js_1.HoverVerbosityAction.Decrease, index, args?.focus);
    }
}
exports.DecreaseHoverVerbosityLevel = DecreaseHoverVerbosityLevel;
